{"cells":[{"cell_type":"markdown","metadata":{"id":"fDNqm1NSqk03"},"source":["# Training and testing LSTM50 on the yeast genome "]},{"cell_type":"markdown","metadata":{"id":"I14tydD1qk06"},"source":["This short Jupyter notebook takes you through the main steps we did for our LSTM50 model on yeast.\n","The code was originally written in Python 2.7, but has been amended for Python 3.6. We ran it in Google Colab where you can let a GPU run the training and prediction of the model (it's all set up and just there to use). If you want to run this on your own server/machine, you need to install Keras (we used v 2.1.6) and Tensorflow (we used tensorflow_gpu-1.12.0-cp27-none-linux_x86_64.whl).\n","It may be easiest to create and environment in which you install these. And then invoke that environment \n","with the kernel for this notebook (you'll need to install a few things for having your environment available among the kernels in your Jupyter session --- check the internet, there's lots of help).\n","\n","The commands below call code (functions) in various python modules that we created for our purposes. All \n","these modules are included in the home Github repo of this notebook. So to run the code you must download/clone\n","the repo contents to a folder on your machine/server.\n","\n","If you run the notebook in Google Colab you can place all code in a Google drive and then mount the drive. This makes the code available in the Colab session (and that's what we do below). Output is also placed there, so may may need a few Gb's of free space on your drive!\n","\n","The data needed here is the yeast genome, which you can download from various public sources, e.g. the UCSC \n","genome site (http://www.genome.ucsc.edu/). \n","\n","We first mount a Google drive and then set some paths to where code and genome data are placed, and where to place output (you'll need to create these folders in your Google drive):"]},{"cell_type":"code","source":["from google.colab import drive\n","drive.mount('/content/drive', force_remount=True)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"V0H9j2BdvCPf","executionInfo":{"status":"ok","timestamp":1650141509415,"user_tz":-120,"elapsed":18353,"user":{"displayName":"Christian Grønbæk","userId":"14909394381660051993"}},"outputId":"fa756cab-ecc8-497d-f66a-15274a07e081"},"execution_count":3,"outputs":[{"output_type":"stream","name":"stdout","text":["Mounted at /content/drive\n"]}]},{"cell_type":"code","source":["# Insert the directory\n","import sys\n","sys.path.insert(0,\"/content/drive/MyDrive/DNA_proj/development/\")\n","sys.path.insert(0,\"/content/drive/MyDrive/DNA_proj/data/\")\n","sys.path.insert(0,\"/content/drive/MyDrive/DNA_proj/results/training/\")\n","sys.path.insert(0,\"/content/drive/MyDrive/DNA_proj/results/predictions/\")\n","sys.path.insert(0,\"/content/drive/MyDrive/DNA_proj/results/GCbias/\")\n"],"metadata":{"id":"GVeIE6Fs0AoO","executionInfo":{"status":"ok","timestamp":1650141509415,"user_tz":-120,"elapsed":4,"user":{"displayName":"Christian Grønbæk","userId":"14909394381660051993"}}},"execution_count":4,"outputs":[]},{"cell_type":"code","execution_count":11,"metadata":{"id":"6_wOQcq5qk07","executionInfo":{"status":"ok","timestamp":1650141629980,"user_tz":-120,"elapsed":368,"user":{"displayName":"Christian Grønbæk","userId":"14909394381660051993"}}},"outputs":[],"source":["rootGenome = \"/content/drive/MyDrive/DNA_proj/data/\"\n","fileName = \"S288C_reference_sequence_R64-1-1_20110203.fsa\"\n","fileGenome = rootGenome + fileName\n","chromoList = ['R64_chr1', 'R64_chr2', 'R64_chr3', 'R64_chr4', 'R64_chr5', 'R64_chr6', 'R64_chr7', 'R64_chr8','R64_chr9', 'R64_chr10', 'R64_chr11', 'R64_chr12','R64_chr13', 'R64_chr14', 'R64_chr15', 'R64_chr16','R64_chr17' ]\n"]},{"cell_type":"markdown","source":["For training and validating the model we split the genome in two truly non-overlapping parts (you can skip this and just use e.g. an 80-20 split for training/validation)."],"metadata":{"id":"0Rk7zRStDfPk"}},{"cell_type":"code","source":["import dnaNet_dataGen as dataGen"],"metadata":{"id":"uTJszlexXmtT","executionInfo":{"status":"ok","timestamp":1650129346087,"user_tz":-120,"elapsed":3900,"user":{"displayName":"Christian Grønbæk","userId":"14909394381660051993"}}},"execution_count":4,"outputs":[]},{"cell_type":"code","source":["rootOutput =rootGenome\n","ext = '.txt'\n","chromoFileList = []\n","for ch in chromoList:\n","    chromoFileList.append( rootOutput + ch + ext )\n","fileNamePart1 =  r'/part1.txt'\n","fileNamePart2 =  r'/part2.txt'\n","splitNrPositions = [40000, 10000 ]\n","dataGen.splitChromoStrings(chromoFileList = chromoFileList, splitNrPositions = splitNrPositions,lineLength = 60,  rootOutput = rootOutput, fileNamePart1 = fileNamePart1, fileNamePart2 = fileNamePart2)\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"wvd5m1oxXZjc","executionInfo":{"status":"ok","timestamp":1650129361744,"user_tz":-120,"elapsed":15663,"user":{"displayName":"Christian Grønbæk","userId":"14909394381660051993"}},"outputId":"318e2771-6a85-483e-80fe-68b541bddb7b"},"execution_count":5,"outputs":[{"output_type":"stream","name":"stdout","text":["doneTheseChromos for part 1:  ['/content/drive/MyDrive/DNA_proj/data/R64_chr1.txt', '/content/drive/MyDrive/DNA_proj/data/R64_chr2.txt', '/content/drive/MyDrive/DNA_proj/data/R64_chr3.txt', '/content/drive/MyDrive/DNA_proj/data/R64_chr4.txt', '/content/drive/MyDrive/DNA_proj/data/R64_chr5.txt', '/content/drive/MyDrive/DNA_proj/data/R64_chr6.txt', '/content/drive/MyDrive/DNA_proj/data/R64_chr7.txt', '/content/drive/MyDrive/DNA_proj/data/R64_chr8.txt', '/content/drive/MyDrive/DNA_proj/data/R64_chr9.txt', '/content/drive/MyDrive/DNA_proj/data/R64_chr10.txt', '/content/drive/MyDrive/DNA_proj/data/R64_chr11.txt', '/content/drive/MyDrive/DNA_proj/data/R64_chr12.txt', '/content/drive/MyDrive/DNA_proj/data/R64_chr13.txt', '/content/drive/MyDrive/DNA_proj/data/R64_chr14.txt', '/content/drive/MyDrive/DNA_proj/data/R64_chr15.txt', '/content/drive/MyDrive/DNA_proj/data/R64_chr16.txt', '/content/drive/MyDrive/DNA_proj/data/R64_chr17.txt']\n","Number of chromos in vs out:  17 17\n","doneTheseChromos for part 2:  ['/content/drive/MyDrive/DNA_proj/data/R64_chr1.txt', '/content/drive/MyDrive/DNA_proj/data/R64_chr2.txt', '/content/drive/MyDrive/DNA_proj/data/R64_chr3.txt', '/content/drive/MyDrive/DNA_proj/data/R64_chr4.txt', '/content/drive/MyDrive/DNA_proj/data/R64_chr5.txt', '/content/drive/MyDrive/DNA_proj/data/R64_chr6.txt', '/content/drive/MyDrive/DNA_proj/data/R64_chr7.txt', '/content/drive/MyDrive/DNA_proj/data/R64_chr8.txt', '/content/drive/MyDrive/DNA_proj/data/R64_chr9.txt', '/content/drive/MyDrive/DNA_proj/data/R64_chr10.txt', '/content/drive/MyDrive/DNA_proj/data/R64_chr11.txt', '/content/drive/MyDrive/DNA_proj/data/R64_chr12.txt', '/content/drive/MyDrive/DNA_proj/data/R64_chr13.txt', '/content/drive/MyDrive/DNA_proj/data/R64_chr14.txt', '/content/drive/MyDrive/DNA_proj/data/R64_chr15.txt', '/content/drive/MyDrive/DNA_proj/data/R64_chr16.txt', '/content/drive/MyDrive/DNA_proj/data/R64_chr17.txt']\n","Number of chromos in vs out:  17 17\n"]}]},{"cell_type":"markdown","metadata":{"id":"cu89pR5Rqk08"},"source":["For defining and training the model we will use a function called allInOneWithDynSampling_ConvLSTMmodel, which is placed in \n","our dnaNet_LSTM module. The function is a 'wrapper': It covers the definition of the model and the configuration for the \n","training process (and some more). It has a long list of arguments, here we use some of them. If you want to get deeper into its\n","details, you will be best off by reading the code. If you manage to come to terms with it you will also know how to define new \n","model types and plug them into this wrapper. For now let us first set the model details. \n","\n","The model architecture that we will use is a 'convolutional-LSTM'. We have coded our version in the function makeConv1DLSTMmodel\n","(in dnaNet_LSTM). This function is then called from within allInOneWithDynSampling_ConvLSTMmodel with whatever argument values we\n","have suppplied. We want a model with flank size 50, with an intial convolutional layer using 64 filters of size 4, two bidirectional LSTM-layers where the first is sequence-to-sequence and a final dense layer with 50 untis. So we set (don't worry about the many unexplained arguments):"]},{"cell_type":"code","execution_count":6,"metadata":{"id":"4nb1WDk0qk08","executionInfo":{"status":"ok","timestamp":1650129361744,"user_tz":-120,"elapsed":13,"user":{"displayName":"Christian Grønbæk","userId":"14909394381660051993"}}},"outputs":[],"source":["usedThisModel = 'makeConv1DLSTMmodel'\n","customFlankSize = 50\n","#convo:\n","overlap = 0\n","pool_b = 0\n","poolAt = [1, 3] #not in use\n","maxPooling_b = 0\n","poolStrides = 1\n","lengthWindows = [4]\n","nrFilters = [256] \n","filterStride = 1\n","padding = 'valid'\n","#lstm layers:\n","nrOfParallelLSTMstacks = 1 #parallel LSTMs\n","nrLSTMlayers = 1 #OBS: the run data file will record nrLSTMlayers as this number plus 1 if summarizingLSTMLayer_b == 1\n","summarizingLSTMLayer_b = 1\n","LSTMFiltersByLastConvFilters_b = 1\n","nrLSTMFilters = [-1]  #-1: just placeholder to be recorded in runData file\n","tryAveraging_b = 0\n","#Final dense layers:\n","finalDenseLayers_b = 1\n","hiddenUnits = [50]\n","inclFrqModel_b = 0\n","insertFrqModel_b = 0\n","#Additional\n","exonicInfoBinaryFileName  = ''"]},{"cell_type":"markdown","metadata":{"id":"J8UfRryOqk09"},"source":["Next we set the configuration for the training process. First some sizes: Here we let the training consist of 20 rounds ('repeats') each of 10 'epochs' which in turn each consists of 100 steps of loading in a batch of 500 samples and updating the model parameters on it (the last four parameters are not in use, so just set to placeholder values). Each such round then comprises training on 500000 randomly picked samples; at completion of each round a validation is run on 100000 samples (nrTestSamples): "]},{"cell_type":"code","execution_count":13,"metadata":{"id":"90fggUtXqk09","executionInfo":{"status":"ok","timestamp":1650141656316,"user_tz":-120,"elapsed":264,"user":{"displayName":"Christian Grønbæk","userId":"14909394381660051993"}}},"outputs":[],"source":["#In anger (yeast):\n","nrOuterLoops = 1\n","firstIterNr = 0\n","nrOfRepeats = 20\n","firstRepeatNr = 0 #if = n > 0: loads in model from repeatNr n-1 \n","testDataIntervalIdTotrainDataInterval_b = 1\n","trainTestSplitRatio = 0.8 #no used when we run training in splitExercise mode\n","nrEpochs = 10 \n","batchSize = 500\n","stepsPerEpoch = 100\n","trainDataIntervalStepSize = 0 \n","trainDataInterval = [0,15000000]\n","nrTestSamples = 100000\n","testDataInterval = [10000000,-12000000]"]},{"cell_type":"markdown","metadata":{"id":"rG3JZgx2qk0-"},"source":["\n","Second (still parameters for the training process) some parameters, the most important being what optimizer we use, learning rate at start and whether to include the reverse complement of each sample to the batch or not:\n"]},{"cell_type":"code","execution_count":14,"metadata":{"id":"Bhe6O5sqqk0_","executionInfo":{"status":"ok","timestamp":1650141661291,"user_tz":-120,"elapsed":256,"user":{"displayName":"Christian Grønbæk","userId":"14909394381660051993"}}},"outputs":[],"source":["#!!!:\n","augmentWithRevComplementary_b = 0 \n","optimizer = 'ADAM' \n","learningRate = 0.001 \n","chromoNameBound = 1000 \n","labelsCodetype = 0 #1: base pair type prediction\n","# \n","dynSamplesTransformStyle_b = 0 \n","rootFrq = '' \n","file = \"\" \n","frqModelFileName = rootFrq + file \n","flankSizeFrqModel = -1 \n","exclFrqModelFlanks_b = 0\n","\n","## \n","dropout_b = 0 \n","dropoutVal = 0.0 \n","momentum = 0.1 #but we use Adam here, so the value here isn't used\n","onlyOneRandomChromo_b = 0 \n","avoidChromo = [] \n","on_binf_b = 1"]},{"cell_type":"markdown","metadata":{"id":"2xdjTIauqk0_"},"source":["Finally, set a model name and a path for where to place the output:"]},{"cell_type":"code","execution_count":15,"metadata":{"id":"CUk395lnqk1A","executionInfo":{"status":"ok","timestamp":1650141666859,"user_tz":-120,"elapsed":273,"user":{"displayName":"Christian Grønbæk","userId":"14909394381660051993"}}},"outputs":[],"source":["subStr = '1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00'\n","\n","modelName = 'modelLSTM_' + subStr\n","modelDescr = subStr\n","\n","rootOutput =  r\"/content/drive/MyDrive/DNA_proj/results/training/\"\n"]},{"cell_type":"markdown","metadata":{"id":"LBR9-Aeyqk1A"},"source":["Before we can call the allInOneWithDynSampling_ConvLSTMmodel wrapper, we must   see to that we have a GPU running for us and of course load the python module holding the wrapper:"]},{"cell_type":"code","source":["%tensorflow_version 2.x\n","import tensorflow as tf\n","device_name = tf.test.gpu_device_name()\n","if device_name != '/device:GPU:0':\n","  raise SystemError('GPU device not found')\n","print('Found GPU at: {}'.format(device_name))\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"pA4kWZyu23Dm","executionInfo":{"status":"ok","timestamp":1650129369453,"user_tz":-120,"elapsed":7715,"user":{"displayName":"Christian Grønbæk","userId":"14909394381660051993"}},"outputId":"81573b4d-b57f-461f-8595-5ceac90efef3"},"execution_count":10,"outputs":[{"output_type":"stream","name":"stdout","text":["Found GPU at: /device:GPU:0\n"]}]},{"cell_type":"code","source":["import dnaNet_LSTM_v3_py3 as dnaNet"],"metadata":{"id":"DBkcdkH9vaoJ","executionInfo":{"status":"ok","timestamp":1650141522806,"user_tz":-120,"elapsed":7334,"user":{"displayName":"Christian Grønbæk","userId":"14909394381660051993"}}},"execution_count":5,"outputs":[]},{"cell_type":"code","source":["#This cell is just for the case that we have changed the code and need to reload:\n","import imp\n","imp.reload(dnaNet)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"8tavJIoXKoGk","executionInfo":{"status":"ok","timestamp":1650129370687,"user_tz":-120,"elapsed":11,"user":{"displayName":"Christian Grønbæk","userId":"14909394381660051993"}},"outputId":"2742cd6c-b63f-4cc6-e5af-260509736c44"},"execution_count":12,"outputs":[{"output_type":"execute_result","data":{"text/plain":["<module 'dnaNet_LSTM_v3_py3' from '/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py'>"]},"metadata":{},"execution_count":12}]},{"cell_type":"markdown","source":["So now we are all set to run the wrapper with all the values we have set for its arguments above. The time this training takes depends on the GPU and our access to it (one or two hours is not unrealistic). When running, the wrapper writes files to the /results/training directory, which we set right above as rootOutput (see also top where we mounted the drive); in each round files defining the model as well as some keeping track of the training-results are stored; a 'runData.txt' file records the settings of the run.\n","\n","A note: if you want rather to run a 80-20 split version, you don't need to split the genome and when running the wrapper leave splitExercise_b out (or set it to 0); further let fileGenome point to the genome file and leave fileGenome_forVal out (or set it to ' ')."],"metadata":{"id":"o6ksGewG3m_2"}},{"cell_type":"code","source":["fileGenome = r\"/content/drive/MyDrive/DNA_proj/data/part1.txt\"\n","fileGenome_forVal = r\"/content/drive/MyDrive/DNA_proj/data/part2.txt\"\n","labelsCodetype = 0 #1: base pair type prediction\n","dnaNet.allInOneWithDynSampling_ConvLSTMmodel(splitExercise_b = 1, genomeFileName_forVal = fileGenome_forVal, rootOutput = rootOutput, usedThisModel = usedThisModel, labelsCodetype = labelsCodetype, nrOuterLoops = nrOuterLoops, firstIterNr = firstIterNr, nrOfRepeats = nrOfRepeats,  firstRepeatNr = firstRepeatNr, convLayers_b = 1, overlap = overlap, learningRate = learningRate, momentum = momentum,  genomeFileName = fileGenome, chromoNameBound = chromoNameBound, trainTestSplitRatio = trainTestSplitRatio, customFlankSize = customFlankSize, inclFrqModel_b = inclFrqModel_b, insertFrqModel_b = insertFrqModel_b, exclFrqModelFlanks_b = exclFrqModelFlanks_b, frqModelFileName = frqModelFileName, flankSizeFrqModel = flankSizeFrqModel, modelName = modelName, testDataIntervalIdTotrainDataInterval_b = testDataIntervalIdTotrainDataInterval_b, trainDataIntervalStepSize = trainDataIntervalStepSize, trainDataInterval0 = trainDataInterval , nrTestSamples = nrTestSamples, testDataInterval = testDataInterval,   genSamples_b = 1,  nrOfParallelLSTMstacks = nrOfParallelLSTMstacks, lengthWindows = lengthWindows, nrLSTMlayers = nrLSTMlayers, summarizingLSTMLayer_b = summarizingLSTMLayer_b, LSTMFiltersByLastConvFilters_b = LSTMFiltersByLastConvFilters_b, nrLSTMFilters = nrLSTMFilters, finalDenseLayers_b = finalDenseLayers_b, hiddenUnits = hiddenUnits, nrFilters = nrFilters, padding = padding, filterStride = filterStride, tryAveraging_b= tryAveraging_b, pool_b = pool_b, maxPooling_b = maxPooling_b, poolAt = poolAt, poolStrides = poolStrides, optimizer = optimizer, dropoutVal = dropoutVal, dropout_b = dropout_b, augmentWithRevComplementary_b = augmentWithRevComplementary_b, batchSize = batchSize, nrEpochs = nrEpochs, stepsPerEpoch = stepsPerEpoch, shuffle_b = 0, on_binf_b = on_binf_b) \n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"9oSQNK95Zklv","outputId":"930a3a0d-9608-42b4-95ec-45ce0e8a821f","executionInfo":{"status":"ok","timestamp":1650134766697,"user_tz":-120,"elapsed":5396018,"user":{"displayName":"Christian Grønbæk","userId":"14909394381660051993"}}},"execution_count":13,"outputs":[{"output_type":"stream","name":"stdout","text":["Now at outer iteration:  0\n","trainDataInterval  [0, 15000000]\n","testDataInterval  [0, 15000000]\n","Reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 1000\n","Fast reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 1000\n","OBS: no file containing exonic info was provided, so exonic status is set to 0 from 0 - 15000001\n","Genome data file 1st line:\n","  >ref|NC_001133| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=I]\n","\n","Found data for this chromosome: ref|NC_001133| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=I]\n","1 1 1\n","Found data for this chromosome: ref|NC_001134| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=II]\n","Found data for this chromosome: ref|NC_001135| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=III]\n","Found data for this chromosome: ref|NC_001136| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=IV]\n","Found data for this chromosome: ref|NC_001137| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=V]\n","Found data for this chromosome: ref|NC_001138| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VI]\n","Found data for this chromosome: ref|NC_001139| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VII]\n","Found data for this chromosome: ref|NC_001140| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VIII]\n","Found data for this chromosome: ref|NC_001141| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=IX]\n","Found data for this chromosome: ref|NC_001142| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=X]\n","Found data for this chromosome: ref|NC_001143| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XI]\n","Found data for this chromosome: ref|NC_001144| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XII]\n","Found data for this chromosome: ref|NC_001145| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XIII]\n","Found data for this chromosome: ref|NC_001146| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XIV]\n","Found data for this chromosome: ref|NC_001147| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XV]\n","Found data for this chromosome: ref|NC_001148| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XVI]\n","Found data for this chromosome: ref|NC_001224| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [location=mitochondrion] [top=circular]\n","[['ref|NC_001133| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=I]', 181024], ['ref|NC_001134| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=II]', 645922], ['ref|NC_001135| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=III]', 247748], ['ref|NC_001136| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=IV]', 1220192], ['ref|NC_001137| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=V]', 458807], ['ref|NC_001138| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VI]', 211128], ['ref|NC_001139| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VII]', 865943], ['ref|NC_001140| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VIII]', 444576], ['ref|NC_001141| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=IX]', 351338], ['ref|NC_001142| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=X]', 590416], ['ref|NC_001143| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XI]', 529071], ['ref|NC_001144| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XII]', 861720], ['ref|NC_001145| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XIII]', 737491], ['ref|NC_001146| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XIV]', 626910], ['ref|NC_001147| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XV]', 865943], ['ref|NC_001148| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XVI]', 751287], ['ref|NC_001224| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [location=mitochondrion] [top=circular]', 66102]]\n","[]\n","Length of genome sequence read in:9655618\n","Length of exonic-info sequence read in:9655618\n","ACGTacgt checked 0 tokens\n","Length genome sequence, ACGT's and W's:9655618\n","Of these 268 are W's\n","Length genome sequence, only ACGT's:9655618\n","Read in data for these (chromos, len(chromo)):  [['ref|NC_001133| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=I]', 181024], ['ref|NC_001134| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=II]', 645922], ['ref|NC_001135| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=III]', 247748], ['ref|NC_001136| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=IV]', 1220192], ['ref|NC_001137| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=V]', 458807], ['ref|NC_001138| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VI]', 211128], ['ref|NC_001139| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VII]', 865943], ['ref|NC_001140| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VIII]', 444576], ['ref|NC_001141| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=IX]', 351338], ['ref|NC_001142| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=X]', 590416], ['ref|NC_001143| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XI]', 529071], ['ref|NC_001144| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XII]', 861720], ['ref|NC_001145| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XIII]', 737491], ['ref|NC_001146| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XIV]', 626910], ['ref|NC_001147| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XV]', 865943], ['ref|NC_001148| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XVI]', 751287], ['ref|NC_001224| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [location=mitochondrion] [top=circular]', 66102]]\n","I'm doing the split-exercise, and will be using this part of the splitted genome for validations:  /content/drive/MyDrive/DNA_proj/data/part2.txt\n","Reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 1000\n","Fast reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 1000\n","OBS: no file containing exonic info was provided, so exonic status is set to 0 from 0 - 15000001\n","Genome data file 1st line:\n","  >ref|NC_001133| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=I]\n","\n","Found data for this chromosome: ref|NC_001133| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=I]\n","60 60 60\n","Found data for this chromosome: ref|NC_001134| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=II]\n","Found data for this chromosome: ref|NC_001135| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=III]\n","Found data for this chromosome: ref|NC_001136| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=IV]\n","Found data for this chromosome: ref|NC_001137| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=V]\n","Found data for this chromosome: ref|NC_001138| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VI]\n","Found data for this chromosome: ref|NC_001139| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VII]\n","Found data for this chromosome: ref|NC_001140| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VIII]\n","Found data for this chromosome: ref|NC_001141| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=IX]\n","Found data for this chromosome: ref|NC_001142| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=X]\n","Found data for this chromosome: ref|NC_001143| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XI]\n","Found data for this chromosome: ref|NC_001144| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XII]\n","Found data for this chromosome: ref|NC_001145| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XIII]\n","Found data for this chromosome: ref|NC_001146| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XIV]\n","Found data for this chromosome: ref|NC_001147| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XV]\n","Found data for this chromosome: ref|NC_001148| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XVI]\n","Found data for this chromosome: ref|NC_001224| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [location=mitochondrion] [top=circular]\n","[['ref|NC_001133| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=I]', 49205], ['ref|NC_001134| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=II]', 167297], ['ref|NC_001135| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=III]', 68887], ['ref|NC_001136| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=IV]', 311805], ['ref|NC_001137| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=V]', 118092], ['ref|NC_001138| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VI]', 59046], ['ref|NC_001139| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VII]', 225043], ['ref|NC_001140| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VIII]', 118092], ['ref|NC_001141| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=IX]', 88569], ['ref|NC_001142| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=X]', 155367], ['ref|NC_001143| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XI]', 137774], ['ref|NC_001144| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XII]', 216502], ['ref|NC_001145| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XIII]', 186979], ['ref|NC_001146| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XIV]', 157456], ['ref|NC_001147| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XV]', 225394], ['ref|NC_001148| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XVI]', 196820], ['ref|NC_001224| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [location=mitochondrion] [top=circular]', 19682]]\n","[]\n","Length of genome sequence read in:2502010\n","Length of exonic-info sequence read in:2502010\n","ACGTacgt checked 0 tokens\n","Length genome sequence, ACGT's and W's:2502010\n","Of these 255 are W's\n","Length genome sequence, only ACGT's:2502010\n","Read in validation-data for these (chromos, len(chromo)):  [['ref|NC_001133| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=I]', 49205], ['ref|NC_001134| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=II]', 167297], ['ref|NC_001135| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=III]', 68887], ['ref|NC_001136| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=IV]', 311805], ['ref|NC_001137| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=V]', 118092], ['ref|NC_001138| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VI]', 59046], ['ref|NC_001139| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VII]', 225043], ['ref|NC_001140| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VIII]', 118092], ['ref|NC_001141| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=IX]', 88569], ['ref|NC_001142| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=X]', 155367], ['ref|NC_001143| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XI]', 137774], ['ref|NC_001144| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XII]', 216502], ['ref|NC_001145| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XIII]', 186979], ['ref|NC_001146| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XIV]', 157456], ['ref|NC_001147| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XV]', 225394], ['ref|NC_001148| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XVI]', 196820], ['ref|NC_001224| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [location=mitochondrion] [top=circular]', 19682]]\n","lGenome: 9655618\n","sizeInput is set to:  100\n","Train-test split sizes (nr of positions of genome array):  9655618 2502010\n","Build Conv1d plus LSTM model...\n","First the 1d-convo, ba ...\n","Left-hand shape after 1st convo  (None, 47, 256)\n","Right-hand shape after 1st convo  (None, 47, 256)\n","...by!\n","Left-hand shape after all convo's  (None, 47, 256)\n","Right-hand shape after all convo's  (None, 47, 256)\n","Then the lstm part ...\n","I'm using these filter sizes for the lstms  [256]\n","LSTM layer:  0\n","Left-hand shape after after i'th LSTM (None, 47, 256)\n","Right-hand shape after i'th LSTM  (None, 47, 256)\n","Left-hand shape after 2nd LSTM  (None, 256)\n","Right-hand shape after 2nd LSTM  (None, 256)\n","Shape of concatenated LSTM output  (None, 512)\n","Shape of LSTM-stacks output  (None, 512)\n","Shape after final dense layer  (None, 50)\n","Output shape  (None, 4)\n","... Model's build.\n","Model: \"model\"\n","__________________________________________________________________________________________________\n"," Layer (type)                   Output Shape         Param #     Connected to                     \n","==================================================================================================\n"," input_1 (InputLayer)           [(None, 50, 4)]      0           []                               \n","                                                                                                  \n"," input_2 (InputLayer)           [(None, 50, 4)]      0           []                               \n","                                                                                                  \n"," conv1d (Conv1D)                (None, 47, 256)      4352        ['input_1[0][0]']                \n","                                                                                                  \n"," conv1d_1 (Conv1D)              (None, 47, 256)      4352        ['input_2[0][0]']                \n","                                                                                                  \n"," lstm (LSTM)                    (None, 47, 256)      525312      ['conv1d[0][0]']                 \n","                                                                                                  \n"," lstm_1 (LSTM)                  (None, 47, 256)      525312      ['conv1d_1[0][0]']               \n","                                                                                                  \n"," lstm_2 (LSTM)                  (None, 256)          525312      ['lstm[0][0]']                   \n","                                                                                                  \n"," lstm_3 (LSTM)                  (None, 256)          525312      ['lstm_1[0][0]']                 \n","                                                                                                  \n"," concatenate (Concatenate)      (None, 512)          0           ['lstm_2[0][0]',                 \n","                                                                  'lstm_3[0][0]']                 \n","                                                                                                  \n"," dense (Dense)                  (None, 50)           25650       ['concatenate[0][0]']            \n","                                                                                                  \n"," dense_1 (Dense)                (None, 4)            204         ['dense[0][0]']                  \n","                                                                                                  \n","==================================================================================================\n","Total params: 2,135,806\n","Trainable params: 2,135,806\n","Non-trainable params: 0\n","__________________________________________________________________________________________________\n","Next: compile it ..\n","Compiled model ...\n"]},{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.7/dist-packages/keras/optimizer_v2/adam.py:105: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.\n","  super(Adam, self).__init__(name, **kwargs)\n"]},{"output_type":"stream","name":"stdout","text":["Epoch 2/10\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2559: UserWarning: `Model.fit_generator` is deprecated and will be removed in a future version. Please use `Model.fit`, which supports generators.\n","  history = net.fit_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps_per_epoch= stepsPerEpoch, epochs=nrEpochs, verbose=1, callbacks=None, validation_data=None, validation_steps=None, class_weight=None, max_queue_size=2, workers=1, use_multiprocessing=False,  initial_epoch=1)\n"]},{"output_type":"stream","name":"stdout","text":["100/100 [==============================] - 38s 224ms/step - loss: 1.3564 - accuracy: 0.3290\n","Epoch 3/10\n","100/100 [==============================] - 22s 224ms/step - loss: 1.3456 - accuracy: 0.3523\n","Epoch 4/10\n","100/100 [==============================] - 23s 228ms/step - loss: 1.3399 - accuracy: 0.3602\n","Epoch 5/10\n","100/100 [==============================] - 23s 228ms/step - loss: 1.3422 - accuracy: 0.3568\n","Epoch 6/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3410 - accuracy: 0.3599\n","Epoch 7/10\n","100/100 [==============================] - 23s 228ms/step - loss: 1.3392 - accuracy: 0.3586\n","Epoch 8/10\n","100/100 [==============================] - 23s 228ms/step - loss: 1.3364 - accuracy: 0.3599\n","Epoch 9/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3332 - accuracy: 0.3613\n","Epoch 10/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3345 - accuracy: 0.3613\n","dict_keys(['loss', 'accuracy'])\n","Now testing ...\n","Split-exercise. The data for this test/validation are taken from another source than the train data\n","Train-test split ratio set to: 0.259125\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2728: UserWarning: `Model.evaluate_generator` is deprecated and will be removed in a future version. Please use `Model.evaluate`, which supports generators.\n","  score, acc = net.evaluate_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps = np.int(float(nrTestSamples)/batchSize))\n"]},{"output_type":"stream","name":"stdout","text":["Test score: 1.3357731103897095\n","Test accuracy: 0.3591499924659729\n","I've now reloaded the model from the previous iteration (for test-only: for this repeatNr:  /content/drive/MyDrive/DNA_proj/results/training/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr0\n","Next: compile it ..\n","Compiled model ...\n","Epoch 2/10\n"]},{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.7/dist-packages/keras/optimizer_v2/adam.py:105: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.\n","  super(Adam, self).__init__(name, **kwargs)\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2559: UserWarning: `Model.fit_generator` is deprecated and will be removed in a future version. Please use `Model.fit`, which supports generators.\n","  history = net.fit_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps_per_epoch= stepsPerEpoch, epochs=nrEpochs, verbose=1, callbacks=None, validation_data=None, validation_steps=None, class_weight=None, max_queue_size=2, workers=1, use_multiprocessing=False,  initial_epoch=1)\n"]},{"output_type":"stream","name":"stdout","text":["100/100 [==============================] - 30s 232ms/step - loss: 1.3337 - accuracy: 0.3595\n","Epoch 3/10\n","100/100 [==============================] - 23s 231ms/step - loss: 1.3325 - accuracy: 0.3610\n","Epoch 4/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3334 - accuracy: 0.3585\n","Epoch 5/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3297 - accuracy: 0.3643\n","Epoch 6/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3291 - accuracy: 0.3665\n","Epoch 7/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3290 - accuracy: 0.3670\n","Epoch 8/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3270 - accuracy: 0.3677\n","Epoch 9/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3299 - accuracy: 0.3652\n","Epoch 10/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3261 - accuracy: 0.3685\n","dict_keys(['loss', 'accuracy'])\n","Now testing ...\n","Split-exercise. The data for this test/validation are taken from another source than the train data\n","Train-test split ratio set to: 0.259125\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2728: UserWarning: `Model.evaluate_generator` is deprecated and will be removed in a future version. Please use `Model.evaluate`, which supports generators.\n","  score, acc = net.evaluate_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps = np.int(float(nrTestSamples)/batchSize))\n"]},{"output_type":"stream","name":"stdout","text":["Test score: 1.3285356760025024\n","Test accuracy: 0.365119993686676\n","I've now reloaded the model from the previous iteration (for test-only: for this repeatNr:  /content/drive/MyDrive/DNA_proj/results/training/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr1\n","Next: compile it ..\n","Compiled model ...\n","Epoch 2/10\n"]},{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.7/dist-packages/keras/optimizer_v2/adam.py:105: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.\n","  super(Adam, self).__init__(name, **kwargs)\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2559: UserWarning: `Model.fit_generator` is deprecated and will be removed in a future version. Please use `Model.fit`, which supports generators.\n","  history = net.fit_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps_per_epoch= stepsPerEpoch, epochs=nrEpochs, verbose=1, callbacks=None, validation_data=None, validation_steps=None, class_weight=None, max_queue_size=2, workers=1, use_multiprocessing=False,  initial_epoch=1)\n"]},{"output_type":"stream","name":"stdout","text":["100/100 [==============================] - 30s 231ms/step - loss: 1.3312 - accuracy: 0.3632\n","Epoch 3/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3267 - accuracy: 0.3672\n","Epoch 4/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3312 - accuracy: 0.3627\n","Epoch 5/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3320 - accuracy: 0.3645\n","Epoch 6/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3313 - accuracy: 0.3638\n","Epoch 7/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3300 - accuracy: 0.3681\n","Epoch 8/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3266 - accuracy: 0.3677\n","Epoch 9/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3287 - accuracy: 0.3658\n","Epoch 10/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3272 - accuracy: 0.3691\n","dict_keys(['loss', 'accuracy'])\n","Now testing ...\n","Split-exercise. The data for this test/validation are taken from another source than the train data\n","Train-test split ratio set to: 0.259125\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2728: UserWarning: `Model.evaluate_generator` is deprecated and will be removed in a future version. Please use `Model.evaluate`, which supports generators.\n","  score, acc = net.evaluate_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps = np.int(float(nrTestSamples)/batchSize))\n"]},{"output_type":"stream","name":"stdout","text":["Test score: 1.3267571926116943\n","Test accuracy: 0.3661800026893616\n","I've now reloaded the model from the previous iteration (for test-only: for this repeatNr:  /content/drive/MyDrive/DNA_proj/results/training/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr2\n","Next: compile it ..\n","Compiled model ...\n","Epoch 2/10\n"]},{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.7/dist-packages/keras/optimizer_v2/adam.py:105: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.\n","  super(Adam, self).__init__(name, **kwargs)\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2559: UserWarning: `Model.fit_generator` is deprecated and will be removed in a future version. Please use `Model.fit`, which supports generators.\n","  history = net.fit_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps_per_epoch= stepsPerEpoch, epochs=nrEpochs, verbose=1, callbacks=None, validation_data=None, validation_steps=None, class_weight=None, max_queue_size=2, workers=1, use_multiprocessing=False,  initial_epoch=1)\n"]},{"output_type":"stream","name":"stdout","text":["100/100 [==============================] - 30s 230ms/step - loss: 1.3257 - accuracy: 0.3691\n","Epoch 3/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3256 - accuracy: 0.3682\n","Epoch 4/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3250 - accuracy: 0.3698\n","Epoch 5/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3240 - accuracy: 0.3697\n","Epoch 6/10\n","100/100 [==============================] - 23s 228ms/step - loss: 1.3230 - accuracy: 0.3701\n","Epoch 7/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3210 - accuracy: 0.3717\n","Epoch 8/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3219 - accuracy: 0.3712\n","Epoch 9/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3223 - accuracy: 0.3711\n","Epoch 10/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3242 - accuracy: 0.3708\n","dict_keys(['loss', 'accuracy'])\n","Now testing ...\n","Split-exercise. The data for this test/validation are taken from another source than the train data\n","Train-test split ratio set to: 0.259125\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2728: UserWarning: `Model.evaluate_generator` is deprecated and will be removed in a future version. Please use `Model.evaluate`, which supports generators.\n","  score, acc = net.evaluate_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps = np.int(float(nrTestSamples)/batchSize))\n"]},{"output_type":"stream","name":"stdout","text":["Test score: 1.3219184875488281\n","Test accuracy: 0.3705199956893921\n","I've now reloaded the model from the previous iteration (for test-only: for this repeatNr:  /content/drive/MyDrive/DNA_proj/results/training/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr3\n","Next: compile it ..\n","Compiled model ...\n","Epoch 2/10\n"]},{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.7/dist-packages/keras/optimizer_v2/adam.py:105: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.\n","  super(Adam, self).__init__(name, **kwargs)\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2559: UserWarning: `Model.fit_generator` is deprecated and will be removed in a future version. Please use `Model.fit`, which supports generators.\n","  history = net.fit_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps_per_epoch= stepsPerEpoch, epochs=nrEpochs, verbose=1, callbacks=None, validation_data=None, validation_steps=None, class_weight=None, max_queue_size=2, workers=1, use_multiprocessing=False,  initial_epoch=1)\n"]},{"output_type":"stream","name":"stdout","text":["100/100 [==============================] - 29s 231ms/step - loss: 1.3226 - accuracy: 0.3677\n","Epoch 3/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3215 - accuracy: 0.3699\n","Epoch 4/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3212 - accuracy: 0.3712\n","Epoch 5/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3252 - accuracy: 0.3673\n","Epoch 6/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3308 - accuracy: 0.3634\n","Epoch 7/10\n","100/100 [==============================] - 23s 228ms/step - loss: 1.3289 - accuracy: 0.3651\n","Epoch 8/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3256 - accuracy: 0.3682\n","Epoch 9/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3240 - accuracy: 0.3667\n","Epoch 10/10\n","100/100 [==============================] - 23s 228ms/step - loss: 1.3241 - accuracy: 0.3682\n","dict_keys(['loss', 'accuracy'])\n","Now testing ...\n","Split-exercise. The data for this test/validation are taken from another source than the train data\n","Train-test split ratio set to: 0.259125\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2728: UserWarning: `Model.evaluate_generator` is deprecated and will be removed in a future version. Please use `Model.evaluate`, which supports generators.\n","  score, acc = net.evaluate_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps = np.int(float(nrTestSamples)/batchSize))\n"]},{"output_type":"stream","name":"stdout","text":["Test score: 1.3236052989959717\n","Test accuracy: 0.3694100081920624\n","I've now reloaded the model from the previous iteration (for test-only: for this repeatNr:  /content/drive/MyDrive/DNA_proj/results/training/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr4\n","Next: compile it ..\n","Compiled model ...\n","Epoch 2/10\n"]},{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.7/dist-packages/keras/optimizer_v2/adam.py:105: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.\n","  super(Adam, self).__init__(name, **kwargs)\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2559: UserWarning: `Model.fit_generator` is deprecated and will be removed in a future version. Please use `Model.fit`, which supports generators.\n","  history = net.fit_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps_per_epoch= stepsPerEpoch, epochs=nrEpochs, verbose=1, callbacks=None, validation_data=None, validation_steps=None, class_weight=None, max_queue_size=2, workers=1, use_multiprocessing=False,  initial_epoch=1)\n"]},{"output_type":"stream","name":"stdout","text":["100/100 [==============================] - 30s 231ms/step - loss: 1.3209 - accuracy: 0.3735\n","Epoch 3/10\n","100/100 [==============================] - 23s 231ms/step - loss: 1.3194 - accuracy: 0.3714\n","Epoch 4/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3210 - accuracy: 0.3732\n","Epoch 5/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3242 - accuracy: 0.3669\n","Epoch 6/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3185 - accuracy: 0.3739\n","Epoch 7/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3200 - accuracy: 0.3705\n","Epoch 8/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3204 - accuracy: 0.3707\n","Epoch 9/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3168 - accuracy: 0.3747\n","Epoch 10/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3162 - accuracy: 0.3757\n","dict_keys(['loss', 'accuracy'])\n","Now testing ...\n","Split-exercise. The data for this test/validation are taken from another source than the train data\n","Train-test split ratio set to: 0.259125\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2728: UserWarning: `Model.evaluate_generator` is deprecated and will be removed in a future version. Please use `Model.evaluate`, which supports generators.\n","  score, acc = net.evaluate_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps = np.int(float(nrTestSamples)/batchSize))\n"]},{"output_type":"stream","name":"stdout","text":["Test score: 1.3183401823043823\n","Test accuracy: 0.37060999870300293\n","I've now reloaded the model from the previous iteration (for test-only: for this repeatNr:  /content/drive/MyDrive/DNA_proj/results/training/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr5\n","Next: compile it ..\n","Compiled model ...\n","Epoch 2/10\n"]},{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.7/dist-packages/keras/optimizer_v2/adam.py:105: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.\n","  super(Adam, self).__init__(name, **kwargs)\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2559: UserWarning: `Model.fit_generator` is deprecated and will be removed in a future version. Please use `Model.fit`, which supports generators.\n","  history = net.fit_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps_per_epoch= stepsPerEpoch, epochs=nrEpochs, verbose=1, callbacks=None, validation_data=None, validation_steps=None, class_weight=None, max_queue_size=2, workers=1, use_multiprocessing=False,  initial_epoch=1)\n"]},{"output_type":"stream","name":"stdout","text":["100/100 [==============================] - 30s 231ms/step - loss: 1.3176 - accuracy: 0.3747\n","Epoch 3/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3222 - accuracy: 0.3730\n","Epoch 4/10\n","100/100 [==============================] - 23s 231ms/step - loss: 1.3223 - accuracy: 0.3695\n","Epoch 5/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3213 - accuracy: 0.3699\n","Epoch 6/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3203 - accuracy: 0.3704\n","Epoch 7/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3182 - accuracy: 0.3722\n","Epoch 8/10\n","100/100 [==============================] - 23s 231ms/step - loss: 1.3164 - accuracy: 0.3753\n","Epoch 9/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3202 - accuracy: 0.3724\n","Epoch 10/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3192 - accuracy: 0.3733\n","dict_keys(['loss', 'accuracy'])\n","Now testing ...\n","Split-exercise. The data for this test/validation are taken from another source than the train data\n","Train-test split ratio set to: 0.259125\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2728: UserWarning: `Model.evaluate_generator` is deprecated and will be removed in a future version. Please use `Model.evaluate`, which supports generators.\n","  score, acc = net.evaluate_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps = np.int(float(nrTestSamples)/batchSize))\n"]},{"output_type":"stream","name":"stdout","text":["Test score: 1.3176774978637695\n","Test accuracy: 0.3746500015258789\n","I've now reloaded the model from the previous iteration (for test-only: for this repeatNr:  /content/drive/MyDrive/DNA_proj/results/training/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr6\n","Next: compile it ..\n","Compiled model ...\n","Epoch 2/10\n"]},{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.7/dist-packages/keras/optimizer_v2/adam.py:105: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.\n","  super(Adam, self).__init__(name, **kwargs)\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2559: UserWarning: `Model.fit_generator` is deprecated and will be removed in a future version. Please use `Model.fit`, which supports generators.\n","  history = net.fit_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps_per_epoch= stepsPerEpoch, epochs=nrEpochs, verbose=1, callbacks=None, validation_data=None, validation_steps=None, class_weight=None, max_queue_size=2, workers=1, use_multiprocessing=False,  initial_epoch=1)\n"]},{"output_type":"stream","name":"stdout","text":["100/100 [==============================] - 30s 231ms/step - loss: 1.3192 - accuracy: 0.3740\n","Epoch 3/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3201 - accuracy: 0.3693\n","Epoch 4/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3215 - accuracy: 0.3671\n","Epoch 5/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3173 - accuracy: 0.3751\n","Epoch 6/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3180 - accuracy: 0.3739\n","Epoch 7/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3169 - accuracy: 0.3746\n","Epoch 8/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3161 - accuracy: 0.3761\n","Epoch 9/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3166 - accuracy: 0.3749\n","Epoch 10/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3149 - accuracy: 0.3760\n","dict_keys(['loss', 'accuracy'])\n","Now testing ...\n","Split-exercise. The data for this test/validation are taken from another source than the train data\n","Train-test split ratio set to: 0.259125\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2728: UserWarning: `Model.evaluate_generator` is deprecated and will be removed in a future version. Please use `Model.evaluate`, which supports generators.\n","  score, acc = net.evaluate_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps = np.int(float(nrTestSamples)/batchSize))\n"]},{"output_type":"stream","name":"stdout","text":["Test score: 1.3150471448898315\n","Test accuracy: 0.374210000038147\n","I've now reloaded the model from the previous iteration (for test-only: for this repeatNr:  /content/drive/MyDrive/DNA_proj/results/training/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr7\n","Next: compile it ..\n","Compiled model ...\n","Epoch 2/10\n"]},{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.7/dist-packages/keras/optimizer_v2/adam.py:105: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.\n","  super(Adam, self).__init__(name, **kwargs)\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2559: UserWarning: `Model.fit_generator` is deprecated and will be removed in a future version. Please use `Model.fit`, which supports generators.\n","  history = net.fit_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps_per_epoch= stepsPerEpoch, epochs=nrEpochs, verbose=1, callbacks=None, validation_data=None, validation_steps=None, class_weight=None, max_queue_size=2, workers=1, use_multiprocessing=False,  initial_epoch=1)\n"]},{"output_type":"stream","name":"stdout","text":["100/100 [==============================] - 29s 231ms/step - loss: 1.3183 - accuracy: 0.3738\n","Epoch 3/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3203 - accuracy: 0.3716\n","Epoch 4/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3198 - accuracy: 0.3739\n","Epoch 5/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3216 - accuracy: 0.3727\n","Epoch 6/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3174 - accuracy: 0.3739\n","Epoch 7/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3184 - accuracy: 0.3716\n","Epoch 8/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3186 - accuracy: 0.3745\n","Epoch 9/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3173 - accuracy: 0.3741\n","Epoch 10/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3171 - accuracy: 0.3719\n","dict_keys(['loss', 'accuracy'])\n","Now testing ...\n","Split-exercise. The data for this test/validation are taken from another source than the train data\n","Train-test split ratio set to: 0.259125\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2728: UserWarning: `Model.evaluate_generator` is deprecated and will be removed in a future version. Please use `Model.evaluate`, which supports generators.\n","  score, acc = net.evaluate_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps = np.int(float(nrTestSamples)/batchSize))\n"]},{"output_type":"stream","name":"stdout","text":["Test score: 1.3148387670516968\n","Test accuracy: 0.3754499852657318\n","I've now reloaded the model from the previous iteration (for test-only: for this repeatNr:  /content/drive/MyDrive/DNA_proj/results/training/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr8\n","Next: compile it ..\n","Compiled model ...\n","Epoch 2/10\n"]},{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.7/dist-packages/keras/optimizer_v2/adam.py:105: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.\n","  super(Adam, self).__init__(name, **kwargs)\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2559: UserWarning: `Model.fit_generator` is deprecated and will be removed in a future version. Please use `Model.fit`, which supports generators.\n","  history = net.fit_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps_per_epoch= stepsPerEpoch, epochs=nrEpochs, verbose=1, callbacks=None, validation_data=None, validation_steps=None, class_weight=None, max_queue_size=2, workers=1, use_multiprocessing=False,  initial_epoch=1)\n"]},{"output_type":"stream","name":"stdout","text":["100/100 [==============================] - 30s 230ms/step - loss: 1.3133 - accuracy: 0.3789\n","Epoch 3/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3173 - accuracy: 0.3723\n","Epoch 4/10\n","100/100 [==============================] - 23s 231ms/step - loss: 1.3136 - accuracy: 0.3750\n","Epoch 5/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3173 - accuracy: 0.3717\n","Epoch 6/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3164 - accuracy: 0.3763\n","Epoch 7/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3100 - accuracy: 0.3819\n","Epoch 8/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3126 - accuracy: 0.3786\n","Epoch 9/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3108 - accuracy: 0.3806\n","Epoch 10/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3134 - accuracy: 0.3759\n","dict_keys(['loss', 'accuracy'])\n","Now testing ...\n","Split-exercise. The data for this test/validation are taken from another source than the train data\n","Train-test split ratio set to: 0.259125\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2728: UserWarning: `Model.evaluate_generator` is deprecated and will be removed in a future version. Please use `Model.evaluate`, which supports generators.\n","  score, acc = net.evaluate_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps = np.int(float(nrTestSamples)/batchSize))\n"]},{"output_type":"stream","name":"stdout","text":["Test score: 1.3147965669631958\n","Test accuracy: 0.3738600015640259\n","I've now reloaded the model from the previous iteration (for test-only: for this repeatNr:  /content/drive/MyDrive/DNA_proj/results/training/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr9\n","Next: compile it ..\n","Compiled model ...\n","Epoch 2/10\n"]},{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.7/dist-packages/keras/optimizer_v2/adam.py:105: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.\n","  super(Adam, self).__init__(name, **kwargs)\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2559: UserWarning: `Model.fit_generator` is deprecated and will be removed in a future version. Please use `Model.fit`, which supports generators.\n","  history = net.fit_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps_per_epoch= stepsPerEpoch, epochs=nrEpochs, verbose=1, callbacks=None, validation_data=None, validation_steps=None, class_weight=None, max_queue_size=2, workers=1, use_multiprocessing=False,  initial_epoch=1)\n"]},{"output_type":"stream","name":"stdout","text":["100/100 [==============================] - 30s 231ms/step - loss: 1.3136 - accuracy: 0.3767\n","Epoch 3/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3115 - accuracy: 0.3812\n","Epoch 4/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3121 - accuracy: 0.3780\n","Epoch 5/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3115 - accuracy: 0.3784\n","Epoch 6/10\n","100/100 [==============================] - 23s 231ms/step - loss: 1.3110 - accuracy: 0.3810\n","Epoch 7/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3104 - accuracy: 0.3799\n","Epoch 8/10\n","100/100 [==============================] - 23s 231ms/step - loss: 1.3092 - accuracy: 0.3763\n","Epoch 9/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3112 - accuracy: 0.3778\n","Epoch 10/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3107 - accuracy: 0.3814\n","dict_keys(['loss', 'accuracy'])\n","Now testing ...\n","Split-exercise. The data for this test/validation are taken from another source than the train data\n","Train-test split ratio set to: 0.259125\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2728: UserWarning: `Model.evaluate_generator` is deprecated and will be removed in a future version. Please use `Model.evaluate`, which supports generators.\n","  score, acc = net.evaluate_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps = np.int(float(nrTestSamples)/batchSize))\n"]},{"output_type":"stream","name":"stdout","text":["Test score: 1.3127226829528809\n","Test accuracy: 0.3768799901008606\n","I've now reloaded the model from the previous iteration (for test-only: for this repeatNr:  /content/drive/MyDrive/DNA_proj/results/training/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr10\n","Next: compile it ..\n","Compiled model ...\n","Epoch 2/10\n"]},{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.7/dist-packages/keras/optimizer_v2/adam.py:105: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.\n","  super(Adam, self).__init__(name, **kwargs)\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2559: UserWarning: `Model.fit_generator` is deprecated and will be removed in a future version. Please use `Model.fit`, which supports generators.\n","  history = net.fit_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps_per_epoch= stepsPerEpoch, epochs=nrEpochs, verbose=1, callbacks=None, validation_data=None, validation_steps=None, class_weight=None, max_queue_size=2, workers=1, use_multiprocessing=False,  initial_epoch=1)\n"]},{"output_type":"stream","name":"stdout","text":["100/100 [==============================] - 29s 231ms/step - loss: 1.3123 - accuracy: 0.3784\n","Epoch 3/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3139 - accuracy: 0.3771\n","Epoch 4/10\n","100/100 [==============================] - 23s 231ms/step - loss: 1.3157 - accuracy: 0.3743\n","Epoch 5/10\n","100/100 [==============================] - 23s 231ms/step - loss: 1.3137 - accuracy: 0.3768\n","Epoch 6/10\n","100/100 [==============================] - 23s 231ms/step - loss: 1.3076 - accuracy: 0.3841\n","Epoch 7/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3126 - accuracy: 0.3795\n","Epoch 8/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3092 - accuracy: 0.3799\n","Epoch 9/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3081 - accuracy: 0.3817\n","Epoch 10/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3078 - accuracy: 0.3804\n","dict_keys(['loss', 'accuracy'])\n","Now testing ...\n","Split-exercise. The data for this test/validation are taken from another source than the train data\n","Train-test split ratio set to: 0.259125\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2728: UserWarning: `Model.evaluate_generator` is deprecated and will be removed in a future version. Please use `Model.evaluate`, which supports generators.\n","  score, acc = net.evaluate_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps = np.int(float(nrTestSamples)/batchSize))\n"]},{"output_type":"stream","name":"stdout","text":["Test score: 1.3095237016677856\n","Test accuracy: 0.37953001260757446\n","I've now reloaded the model from the previous iteration (for test-only: for this repeatNr:  /content/drive/MyDrive/DNA_proj/results/training/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr11\n","Next: compile it ..\n","Compiled model ...\n","Epoch 2/10\n"]},{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.7/dist-packages/keras/optimizer_v2/adam.py:105: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.\n","  super(Adam, self).__init__(name, **kwargs)\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2559: UserWarning: `Model.fit_generator` is deprecated and will be removed in a future version. Please use `Model.fit`, which supports generators.\n","  history = net.fit_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps_per_epoch= stepsPerEpoch, epochs=nrEpochs, verbose=1, callbacks=None, validation_data=None, validation_steps=None, class_weight=None, max_queue_size=2, workers=1, use_multiprocessing=False,  initial_epoch=1)\n"]},{"output_type":"stream","name":"stdout","text":["100/100 [==============================] - 30s 230ms/step - loss: 1.3105 - accuracy: 0.3790\n","Epoch 3/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3022 - accuracy: 0.3876\n","Epoch 4/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3098 - accuracy: 0.3787\n","Epoch 5/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3099 - accuracy: 0.3799\n","Epoch 6/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3053 - accuracy: 0.3818\n","Epoch 7/10\n","100/100 [==============================] - 23s 231ms/step - loss: 1.3092 - accuracy: 0.3810\n","Epoch 8/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3186 - accuracy: 0.3732\n","Epoch 9/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3266 - accuracy: 0.3686\n","Epoch 10/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3237 - accuracy: 0.3698\n","dict_keys(['loss', 'accuracy'])\n","Now testing ...\n","Split-exercise. The data for this test/validation are taken from another source than the train data\n","Train-test split ratio set to: 0.259125\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2728: UserWarning: `Model.evaluate_generator` is deprecated and will be removed in a future version. Please use `Model.evaluate`, which supports generators.\n","  score, acc = net.evaluate_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps = np.int(float(nrTestSamples)/batchSize))\n"]},{"output_type":"stream","name":"stdout","text":["Test score: 1.3220796585083008\n","Test accuracy: 0.3729900121688843\n","I've now reloaded the model from the previous iteration (for test-only: for this repeatNr:  /content/drive/MyDrive/DNA_proj/results/training/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr12\n","Next: compile it ..\n","Compiled model ...\n","Epoch 2/10\n"]},{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.7/dist-packages/keras/optimizer_v2/adam.py:105: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.\n","  super(Adam, self).__init__(name, **kwargs)\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2559: UserWarning: `Model.fit_generator` is deprecated and will be removed in a future version. Please use `Model.fit`, which supports generators.\n","  history = net.fit_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps_per_epoch= stepsPerEpoch, epochs=nrEpochs, verbose=1, callbacks=None, validation_data=None, validation_steps=None, class_weight=None, max_queue_size=2, workers=1, use_multiprocessing=False,  initial_epoch=1)\n"]},{"output_type":"stream","name":"stdout","text":["100/100 [==============================] - 30s 230ms/step - loss: 1.3182 - accuracy: 0.3759\n","Epoch 3/10\n","100/100 [==============================] - 23s 231ms/step - loss: 1.3170 - accuracy: 0.3744\n","Epoch 4/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3180 - accuracy: 0.3739\n","Epoch 5/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3146 - accuracy: 0.3770\n","Epoch 6/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3132 - accuracy: 0.3764\n","Epoch 7/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3142 - accuracy: 0.3792\n","Epoch 8/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3132 - accuracy: 0.3755\n","Epoch 9/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3085 - accuracy: 0.3823\n","Epoch 10/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3127 - accuracy: 0.3784\n","dict_keys(['loss', 'accuracy'])\n","Now testing ...\n","Split-exercise. The data for this test/validation are taken from another source than the train data\n","Train-test split ratio set to: 0.259125\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2728: UserWarning: `Model.evaluate_generator` is deprecated and will be removed in a future version. Please use `Model.evaluate`, which supports generators.\n","  score, acc = net.evaluate_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps = np.int(float(nrTestSamples)/batchSize))\n"]},{"output_type":"stream","name":"stdout","text":["Test score: 1.3110098838806152\n","Test accuracy: 0.38047000765800476\n","I've now reloaded the model from the previous iteration (for test-only: for this repeatNr:  /content/drive/MyDrive/DNA_proj/results/training/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr13\n","Next: compile it ..\n","Compiled model ...\n","Epoch 2/10\n"]},{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.7/dist-packages/keras/optimizer_v2/adam.py:105: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.\n","  super(Adam, self).__init__(name, **kwargs)\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2559: UserWarning: `Model.fit_generator` is deprecated and will be removed in a future version. Please use `Model.fit`, which supports generators.\n","  history = net.fit_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps_per_epoch= stepsPerEpoch, epochs=nrEpochs, verbose=1, callbacks=None, validation_data=None, validation_steps=None, class_weight=None, max_queue_size=2, workers=1, use_multiprocessing=False,  initial_epoch=1)\n"]},{"output_type":"stream","name":"stdout","text":["100/100 [==============================] - 30s 232ms/step - loss: 1.3123 - accuracy: 0.3766\n","Epoch 3/10\n","100/100 [==============================] - 23s 231ms/step - loss: 1.3137 - accuracy: 0.3741\n","Epoch 4/10\n","100/100 [==============================] - 23s 231ms/step - loss: 1.3107 - accuracy: 0.3811\n","Epoch 5/10\n","100/100 [==============================] - 23s 231ms/step - loss: 1.3099 - accuracy: 0.3769\n","Epoch 6/10\n","100/100 [==============================] - 23s 231ms/step - loss: 1.3080 - accuracy: 0.3823\n","Epoch 7/10\n","100/100 [==============================] - 23s 231ms/step - loss: 1.3091 - accuracy: 0.3819\n","Epoch 8/10\n","100/100 [==============================] - 23s 231ms/step - loss: 1.3062 - accuracy: 0.3839\n","Epoch 9/10\n","100/100 [==============================] - 23s 231ms/step - loss: 1.3090 - accuracy: 0.3796\n","Epoch 10/10\n","100/100 [==============================] - 23s 231ms/step - loss: 1.3107 - accuracy: 0.3802\n","dict_keys(['loss', 'accuracy'])\n","Now testing ...\n","Split-exercise. The data for this test/validation are taken from another source than the train data\n","Train-test split ratio set to: 0.259125\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2728: UserWarning: `Model.evaluate_generator` is deprecated and will be removed in a future version. Please use `Model.evaluate`, which supports generators.\n","  score, acc = net.evaluate_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps = np.int(float(nrTestSamples)/batchSize))\n"]},{"output_type":"stream","name":"stdout","text":["Test score: 1.30729341506958\n","Test accuracy: 0.3800300061702728\n","I've now reloaded the model from the previous iteration (for test-only: for this repeatNr:  /content/drive/MyDrive/DNA_proj/results/training/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr14\n","Next: compile it ..\n","Compiled model ...\n","Epoch 2/10\n"]},{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.7/dist-packages/keras/optimizer_v2/adam.py:105: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.\n","  super(Adam, self).__init__(name, **kwargs)\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2559: UserWarning: `Model.fit_generator` is deprecated and will be removed in a future version. Please use `Model.fit`, which supports generators.\n","  history = net.fit_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps_per_epoch= stepsPerEpoch, epochs=nrEpochs, verbose=1, callbacks=None, validation_data=None, validation_steps=None, class_weight=None, max_queue_size=2, workers=1, use_multiprocessing=False,  initial_epoch=1)\n"]},{"output_type":"stream","name":"stdout","text":["100/100 [==============================] - 30s 231ms/step - loss: 1.3093 - accuracy: 0.3815\n","Epoch 3/10\n","100/100 [==============================] - 23s 231ms/step - loss: 1.3073 - accuracy: 0.3817\n","Epoch 4/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3067 - accuracy: 0.3855\n","Epoch 5/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3052 - accuracy: 0.3827\n","Epoch 6/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3045 - accuracy: 0.3844\n","Epoch 7/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3051 - accuracy: 0.3829\n","Epoch 8/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3070 - accuracy: 0.3794\n","Epoch 9/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3055 - accuracy: 0.3852\n","Epoch 10/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3055 - accuracy: 0.3846\n","dict_keys(['loss', 'accuracy'])\n","Now testing ...\n","Split-exercise. The data for this test/validation are taken from another source than the train data\n","Train-test split ratio set to: 0.259125\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2728: UserWarning: `Model.evaluate_generator` is deprecated and will be removed in a future version. Please use `Model.evaluate`, which supports generators.\n","  score, acc = net.evaluate_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps = np.int(float(nrTestSamples)/batchSize))\n"]},{"output_type":"stream","name":"stdout","text":["Test score: 1.308707356452942\n","Test accuracy: 0.3793799877166748\n","I've now reloaded the model from the previous iteration (for test-only: for this repeatNr:  /content/drive/MyDrive/DNA_proj/results/training/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr15\n","Next: compile it ..\n","Compiled model ...\n","Epoch 2/10\n"]},{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.7/dist-packages/keras/optimizer_v2/adam.py:105: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.\n","  super(Adam, self).__init__(name, **kwargs)\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2559: UserWarning: `Model.fit_generator` is deprecated and will be removed in a future version. Please use `Model.fit`, which supports generators.\n","  history = net.fit_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps_per_epoch= stepsPerEpoch, epochs=nrEpochs, verbose=1, callbacks=None, validation_data=None, validation_steps=None, class_weight=None, max_queue_size=2, workers=1, use_multiprocessing=False,  initial_epoch=1)\n"]},{"output_type":"stream","name":"stdout","text":["100/100 [==============================] - 30s 230ms/step - loss: 1.3056 - accuracy: 0.3848\n","Epoch 3/10\n","100/100 [==============================] - 23s 231ms/step - loss: 1.3082 - accuracy: 0.3819\n","Epoch 4/10\n","100/100 [==============================] - 23s 231ms/step - loss: 1.3016 - accuracy: 0.3881\n","Epoch 5/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3079 - accuracy: 0.3824\n","Epoch 6/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3046 - accuracy: 0.3824\n","Epoch 7/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3020 - accuracy: 0.3890\n","Epoch 8/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3059 - accuracy: 0.3829\n","Epoch 9/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3084 - accuracy: 0.3809\n","Epoch 10/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3088 - accuracy: 0.3808\n","dict_keys(['loss', 'accuracy'])\n","Now testing ...\n","Split-exercise. The data for this test/validation are taken from another source than the train data\n","Train-test split ratio set to: 0.259125\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2728: UserWarning: `Model.evaluate_generator` is deprecated and will be removed in a future version. Please use `Model.evaluate`, which supports generators.\n","  score, acc = net.evaluate_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps = np.int(float(nrTestSamples)/batchSize))\n"]},{"output_type":"stream","name":"stdout","text":["Test score: 1.3103021383285522\n","Test accuracy: 0.37654998898506165\n","I've now reloaded the model from the previous iteration (for test-only: for this repeatNr:  /content/drive/MyDrive/DNA_proj/results/training/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr16\n","Next: compile it ..\n","Compiled model ...\n","Epoch 2/10\n"]},{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.7/dist-packages/keras/optimizer_v2/adam.py:105: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.\n","  super(Adam, self).__init__(name, **kwargs)\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2559: UserWarning: `Model.fit_generator` is deprecated and will be removed in a future version. Please use `Model.fit`, which supports generators.\n","  history = net.fit_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps_per_epoch= stepsPerEpoch, epochs=nrEpochs, verbose=1, callbacks=None, validation_data=None, validation_steps=None, class_weight=None, max_queue_size=2, workers=1, use_multiprocessing=False,  initial_epoch=1)\n"]},{"output_type":"stream","name":"stdout","text":["100/100 [==============================] - 29s 230ms/step - loss: 1.3044 - accuracy: 0.3848\n","Epoch 3/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3046 - accuracy: 0.3852\n","Epoch 4/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3044 - accuracy: 0.3856\n","Epoch 5/10\n","100/100 [==============================] - 23s 231ms/step - loss: 1.3055 - accuracy: 0.3815\n","Epoch 6/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3022 - accuracy: 0.3865\n","Epoch 7/10\n","100/100 [==============================] - 23s 231ms/step - loss: 1.3094 - accuracy: 0.3777\n","Epoch 8/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3054 - accuracy: 0.3826\n","Epoch 9/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3063 - accuracy: 0.3828\n","Epoch 10/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3086 - accuracy: 0.3805\n","dict_keys(['loss', 'accuracy'])\n","Now testing ...\n","Split-exercise. The data for this test/validation are taken from another source than the train data\n","Train-test split ratio set to: 0.259125\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2728: UserWarning: `Model.evaluate_generator` is deprecated and will be removed in a future version. Please use `Model.evaluate`, which supports generators.\n","  score, acc = net.evaluate_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps = np.int(float(nrTestSamples)/batchSize))\n"]},{"output_type":"stream","name":"stdout","text":["Test score: 1.3069136142730713\n","Test accuracy: 0.3824099898338318\n","I've now reloaded the model from the previous iteration (for test-only: for this repeatNr:  /content/drive/MyDrive/DNA_proj/results/training/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr17\n","Next: compile it ..\n","Compiled model ...\n","Epoch 2/10\n"]},{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.7/dist-packages/keras/optimizer_v2/adam.py:105: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.\n","  super(Adam, self).__init__(name, **kwargs)\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2559: UserWarning: `Model.fit_generator` is deprecated and will be removed in a future version. Please use `Model.fit`, which supports generators.\n","  history = net.fit_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps_per_epoch= stepsPerEpoch, epochs=nrEpochs, verbose=1, callbacks=None, validation_data=None, validation_steps=None, class_weight=None, max_queue_size=2, workers=1, use_multiprocessing=False,  initial_epoch=1)\n"]},{"output_type":"stream","name":"stdout","text":["100/100 [==============================] - 30s 231ms/step - loss: 1.3076 - accuracy: 0.3839\n","Epoch 3/10\n","100/100 [==============================] - 23s 231ms/step - loss: 1.3055 - accuracy: 0.3849\n","Epoch 4/10\n","100/100 [==============================] - 23s 231ms/step - loss: 1.3045 - accuracy: 0.3848\n","Epoch 5/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3033 - accuracy: 0.3854\n","Epoch 6/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3002 - accuracy: 0.3870\n","Epoch 7/10\n","100/100 [==============================] - 23s 231ms/step - loss: 1.3043 - accuracy: 0.3823\n","Epoch 8/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3053 - accuracy: 0.3842\n","Epoch 9/10\n","100/100 [==============================] - 23s 231ms/step - loss: 1.3034 - accuracy: 0.3867\n","Epoch 10/10\n","100/100 [==============================] - 23s 231ms/step - loss: 1.3033 - accuracy: 0.3845\n","dict_keys(['loss', 'accuracy'])\n","Now testing ...\n","Split-exercise. The data for this test/validation are taken from another source than the train data\n","Train-test split ratio set to: 0.259125\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2728: UserWarning: `Model.evaluate_generator` is deprecated and will be removed in a future version. Please use `Model.evaluate`, which supports generators.\n","  score, acc = net.evaluate_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps = np.int(float(nrTestSamples)/batchSize))\n"]},{"output_type":"stream","name":"stdout","text":["Test score: 1.3018290996551514\n","Test accuracy: 0.382750004529953\n","I've now reloaded the model from the previous iteration (for test-only: for this repeatNr:  /content/drive/MyDrive/DNA_proj/results/training/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr18\n","Next: compile it ..\n","Compiled model ...\n","Epoch 2/10\n"]},{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.7/dist-packages/keras/optimizer_v2/adam.py:105: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.\n","  super(Adam, self).__init__(name, **kwargs)\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2559: UserWarning: `Model.fit_generator` is deprecated and will be removed in a future version. Please use `Model.fit`, which supports generators.\n","  history = net.fit_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps_per_epoch= stepsPerEpoch, epochs=nrEpochs, verbose=1, callbacks=None, validation_data=None, validation_steps=None, class_weight=None, max_queue_size=2, workers=1, use_multiprocessing=False,  initial_epoch=1)\n"]},{"output_type":"stream","name":"stdout","text":["100/100 [==============================] - 30s 231ms/step - loss: 1.3007 - accuracy: 0.3853\n","Epoch 3/10\n","100/100 [==============================] - 23s 231ms/step - loss: 1.3015 - accuracy: 0.3856\n","Epoch 4/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3036 - accuracy: 0.3835\n","Epoch 5/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3012 - accuracy: 0.3868\n","Epoch 6/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.3029 - accuracy: 0.3851\n","Epoch 7/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.3032 - accuracy: 0.3815\n","Epoch 8/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.2992 - accuracy: 0.3867\n","Epoch 9/10\n","100/100 [==============================] - 23s 229ms/step - loss: 1.2985 - accuracy: 0.3892\n","Epoch 10/10\n","100/100 [==============================] - 23s 230ms/step - loss: 1.2988 - accuracy: 0.3872\n","dict_keys(['loss', 'accuracy'])\n","Now testing ...\n","Split-exercise. The data for this test/validation are taken from another source than the train data\n","Train-test split ratio set to: 0.259125\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_LSTM_v3_py3.py:2728: UserWarning: `Model.evaluate_generator` is deprecated and will be removed in a future version. Please use `Model.evaluate`, which supports generators.\n","  score, acc = net.evaluate_generator(myGenerator(customFlankSize,batchSize, oneSided_b, inclFrqModel_b, insertFrqModel_b, labelsCodetype, forTrain_b), steps = np.int(float(nrTestSamples)/batchSize))\n"]},{"output_type":"stream","name":"stdout","text":["Test score: 1.3017898797988892\n","Test accuracy: 0.38317999243736267\n"]}]},{"cell_type":"markdown","source":["To examine the training process you can plot training-validation results with this call: "],"metadata":{"id":"w9lrXQJOqG-j"}},{"cell_type":"code","source":["import dnaNet_plots as plots\n","\n","bigLoopIterNr = 0\n","modelFileNameList = [rootOutput + 'modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00']\n","\n","batchSizeRatioList = [1]\n","lastRepeatNrList = [nrOfRepeats -1]\n","epochsPerRepeatList = [10. -1] #you have to look up nrEpochs per repeat (the log file for the run, runData ); subtract 1 since epoch 1 in each repeat is apperently not saved\n","\n","modelNameList = ['LSTM4']\n","fileNameAcc = 'LSTM4_total_trainTest_acc_vs_allEpochs'\n","fileNameLoss = 'LSTM4_total_trainTest_loss_vs_allEpochs'\n","plots.collectivePerfPlot(modelFileNameList = modelFileNameList, batchSizeRatioList = batchSizeRatioList, lastRepeatNrList = lastRepeatNrList, epochsPerRepeatList = epochsPerRepeatList, rootOutput = rootOutput, modelNameList = modelNameList, fileNameAcc= fileNameAcc, fileNameLoss = fileNameLoss )\n"],"metadata":{"id":"CYsZniDxqF6p","executionInfo":{"status":"ok","timestamp":1650135150367,"user_tz":-120,"elapsed":515,"user":{"displayName":"Christian Grønbæk","userId":"14909394381660051993"}}},"execution_count":18,"outputs":[]},{"cell_type":"markdown","source":["When the training is done (and if we are satisfied with it), we can apply the trained model to DNA sequence. Here we want to let the model \n","'predict' on the complete yeast genome. To carry this out and many other tasks downstream of it, we use a python module dedicated to these purposes:   "],"metadata":{"id":"3TFUlRA76lM8"}},{"cell_type":"code","execution_count":9,"metadata":{"id":"H-nO1aGqqk1B","executionInfo":{"status":"ok","timestamp":1650141597028,"user_tz":-120,"elapsed":970,"user":{"displayName":"Christian Grønbæk","userId":"14909394381660051993"}}},"outputs":[],"source":["import dnaNet_stats_py3 as stats"]},{"cell_type":"markdown","source":["Settings for the prediction (we segment the genome in 100kb pieces):"],"metadata":{"id":"oZrhNkk0FqfF"}},{"cell_type":"code","source":["rootGenome = rootGenome #was set above\n","rootOutput = r\"/content/drive/MyDrive/DNA_proj/results/predictions/\"\n","rootModel = r\"/content/drive/MyDrive/DNA_proj/results/training/\"\n","lastRepeatNr = nrOfRepeats -1 \n","modelFileNameNN = \"modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr\" + str(lastRepeatNr)\n","\n","augmentWithRevComplementary_b = 0\n","leftRight_b = 1\n","customFlankSize = 50\n","computePredAcc_b = 1\n","Fourier_b = 0\n","\n","segmentLength = 1e5\n","\n","batchSize = 528\n","windowLength = 1\n","stepSize = 1\n","Fourier_b = 0\n","on_binf_b = 1\n","\n","#start positions\n","chromosomeOrderList = ['R64_chr1', 'R64_chr2', 'R64_chr3', 'R64_chr4', 'R64_chr5', 'R64_chr6', 'R64_chr7', 'R64_chr8','R64_chr9', 'R64_chr10', 'R64_chr11', 'R64_chr12','R64_chr13', 'R64_chr14', 'R64_chr15', 'R64_chr16']\n","chromosomeDict = {'R64_chr1':[0,1e8], 'R64_chr2':[0,1e8], 'R64_chr3':[0,1e8], 'R64_chr4':[0,1e8], 'R64_chr5':[0,1e8], 'R64_chr6':[0,1e8], 'R64_chr7':[0,1e8], 'R64_chr8':[0,1e8],'R64_chr9':[0,1e8], 'R64_chr10':[0,1e8], 'R64_chr11':[0,1e8], 'R64_chr12':[0,1e8],'R64_chr13':[0,1e8], 'R64_chr14':[0,1e8], 'R64_chr15':[0,1e8], 'R64_chr16':[0,1e8]}\n","startAtSegmentDict = {}\n","\n"],"metadata":{"id":"O48g8TuxK2DH","executionInfo":{"status":"ok","timestamp":1650141678636,"user_tz":-120,"elapsed":240,"user":{"displayName":"Christian Grønbæk","userId":"14909394381660051993"}}},"execution_count":16,"outputs":[]},{"cell_type":"markdown","source":["Run the predictions -- this may take a while too:"],"metadata":{"id":"bu6M726bFuS8"}},{"cell_type":"code","source":["stats.predictOnChromosomes(rootGenome = rootGenome, \n","                         chromosomeDict = chromosomeDict,\n","                         chromosomeOrderList = chromosomeOrderList, \n","                         rootOutput = rootOutput,\n","                         rootModel = rootModel,\n","                         modelFileName = modelFileNameNN,\n","                        segmentLength = segmentLength,\n","                        augmentWithRevComplementary_b = augmentWithRevComplementary_b, #!!!!!\n","                        startAtSegmentDict = startAtSegmentDict,\n","                        customFlankSize = customFlankSize,\n","                        computePredAcc_b = computePredAcc_b, \n","                        overlap = 0,\n","                        leftRight_b = leftRight_b, #use 1 for bi-directional models\n","                        batchSize = batchSize,\n","                        windowLength = windowLength,\n","                        stepSize = stepSize,\n","                        Fourier_b = Fourier_b,\n","                        on_binf_b = on_binf_b)\n","                        "],"metadata":{"id":"eB54LM2FK2Lc","executionInfo":{"status":"ok","timestamp":1650138352191,"user_tz":-120,"elapsed":2972563,"user":{"displayName":"Christian Grønbæk","userId":"14909394381660051993"}},"colab":{"base_uri":"https://localhost:8080/"},"outputId":"632bf222-17ec-436f-a5da-2385a23f6251"},"execution_count":31,"outputs":[{"output_type":"stream","name":"stdout","text":["Reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 100\n","Fast reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 100\n","OBS: no file containing exonic info was provided, so exonic status is set to 0 from 0 - 100000001\n","Genome data file 1st line:\n","  >ref|NC_001133| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=I]\n","\n","Found data for this chromosome: ref|NC_001133| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=I]\n","60 60 60\n","[['ref|NC_001133| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=I]', 230218]]\n","[]\n","Length of genome sequence read in:230218\n","Length of exonic-info sequence read in:230218\n","ACGTacgt checked 0 tokens\n","Length genome sequence, ACGT's and W's:230218\n","Of these 0 are W's\n","Length genome sequence, only ACGT's:230218\n","nrSegments:  2\n","Now at segment  0\n","Length of encoded genome sequence read in:  230218\n","Nr of samples to be generated:  99950\n","Nr of sampled positions: 99950\n","Yt shape  (99950, 4)\n","Xt_lr shape  (99950, 2, 50, 4)\n","Nr of samples: 99950 ; of which are predicted: 99950\n","lSamples  99950 , nrSteps  99950\n","cntCorr: 38664 , cntTot: 99950 ; average prediction acc : 0.386833\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.386833\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 0: 0.386833\n","Now at segment  1\n","Length of encoded genome sequence read in:  230218\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38355 , cntTot: 100000 ; average prediction acc : 0.383550\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.383550\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 1: 0.385191\n","Reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 100\n","Fast reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 100\n","OBS: no file containing exonic info was provided, so exonic status is set to 0 from 0 - 100000001\n","Genome data file 1st line:\n","  >ref|NC_001134| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=II]\n","\n","Found data for this chromosome: ref|NC_001134| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=II]\n","60 60 60\n","[['ref|NC_001134| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=II]', 813184]]\n","[]\n","Length of genome sequence read in:813184\n","Length of exonic-info sequence read in:813184\n","ACGTacgt checked 0 tokens\n","Length genome sequence, ACGT's and W's:813184\n","Of these 0 are W's\n","Length genome sequence, only ACGT's:813184\n","Directory /content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/ created. Output will be placed there.\n","nrSegments:  8\n","Now at segment  0\n","Length of encoded genome sequence read in:  813184\n","Nr of samples to be generated:  99950\n","Nr of sampled positions: 99950\n","Yt shape  (99950, 4)\n","Xt_lr shape  (99950, 2, 50, 4)\n","Nr of samples: 99950 ; of which are predicted: 99950\n","lSamples  99950 , nrSteps  99950\n","cntCorr: 38153 , cntTot: 99950 ; average prediction acc : 0.381721\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.381721\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 0: 0.381721\n","Now at segment  1\n","Length of encoded genome sequence read in:  813184\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38914 , cntTot: 100000 ; average prediction acc : 0.389140\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.389140\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 1: 0.385431\n","Now at segment  2\n","Length of encoded genome sequence read in:  813184\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38091 , cntTot: 100000 ; average prediction acc : 0.380910\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.380910\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 2: 0.383924\n","Now at segment  3\n","Length of encoded genome sequence read in:  813184\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 39014 , cntTot: 100000 ; average prediction acc : 0.390140\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.390140\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 3: 0.385478\n","Now at segment  4\n","Length of encoded genome sequence read in:  813184\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 39363 , cntTot: 100000 ; average prediction acc : 0.393630\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.393630\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 4: 0.387109\n","Now at segment  5\n","Length of encoded genome sequence read in:  813184\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38837 , cntTot: 100000 ; average prediction acc : 0.388370\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.388370\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 5: 0.387319\n","Now at segment  6\n","Length of encoded genome sequence read in:  813184\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38322 , cntTot: 100000 ; average prediction acc : 0.383220\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.383220\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 6: 0.386733\n","Now at segment  7\n","Length of encoded genome sequence read in:  813184\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38069 , cntTot: 100000 ; average prediction acc : 0.380690\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.380690\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 7: 0.385978\n","Reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 100\n","Fast reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 100\n","OBS: no file containing exonic info was provided, so exonic status is set to 0 from 0 - 100000001\n","Genome data file 1st line:\n","  >ref|NC_001135| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=III]\n","\n","Found data for this chromosome: ref|NC_001135| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=III]\n","60 60 60\n","[['ref|NC_001135| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=III]', 316620]]\n","[]\n","Length of genome sequence read in:316620\n","Length of exonic-info sequence read in:316620\n","ACGTacgt checked 0 tokens\n","Length genome sequence, ACGT's and W's:316620\n","Of these 0 are W's\n","Length genome sequence, only ACGT's:316620\n","Directory /content/drive/MyDrive/DNA_proj/results/predictions/R64_chr3/ created. Output will be placed there.\n","nrSegments:  3\n","Now at segment  0\n","Length of encoded genome sequence read in:  316620\n","Nr of samples to be generated:  99950\n","Nr of sampled positions: 99950\n","Yt shape  (99950, 4)\n","Xt_lr shape  (99950, 2, 50, 4)\n","Nr of samples: 99950 ; of which are predicted: 99950\n","lSamples  99950 , nrSteps  99950\n","cntCorr: 38543 , cntTot: 99950 ; average prediction acc : 0.385623\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.385623\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 0: 0.385623\n","Now at segment  1\n","Length of encoded genome sequence read in:  316620\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38274 , cntTot: 100000 ; average prediction acc : 0.382740\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.382740\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 1: 0.384181\n","Now at segment  2\n","Length of encoded genome sequence read in:  316620\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38196 , cntTot: 100000 ; average prediction acc : 0.381960\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.381960\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 2: 0.383441\n","Reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 100\n","Fast reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 100\n","OBS: no file containing exonic info was provided, so exonic status is set to 0 from 0 - 100000001\n","Genome data file 1st line:\n","  >ref|NC_001136| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=IV]\n","\n","Found data for this chromosome: ref|NC_001136| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=IV]\n","60 60 60\n","[['ref|NC_001136| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=IV]', 1531933]]\n","[]\n","Length of genome sequence read in:1531933\n","Length of exonic-info sequence read in:1531933\n","ACGTacgt checked 0 tokens\n","Length genome sequence, ACGT's and W's:1531933\n","Of these 0 are W's\n","Length genome sequence, only ACGT's:1531933\n","Directory /content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/ created. Output will be placed there.\n","nrSegments:  15\n","Now at segment  0\n","Length of encoded genome sequence read in:  1531933\n","Nr of samples to be generated:  99950\n","Nr of sampled positions: 99950\n","Yt shape  (99950, 4)\n","Xt_lr shape  (99950, 2, 50, 4)\n","Nr of samples: 99950 ; of which are predicted: 99950\n","lSamples  99950 , nrSteps  99950\n","cntCorr: 38074 , cntTot: 99950 ; average prediction acc : 0.380930\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.380930\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 0: 0.380930\n","Now at segment  1\n","Length of encoded genome sequence read in:  1531933\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 39277 , cntTot: 100000 ; average prediction acc : 0.392770\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.392770\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 1: 0.386852\n","Now at segment  2\n","Length of encoded genome sequence read in:  1531933\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38552 , cntTot: 100000 ; average prediction acc : 0.385520\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.385520\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 2: 0.386408\n","Now at segment  3\n","Length of encoded genome sequence read in:  1531933\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 39382 , cntTot: 100000 ; average prediction acc : 0.393820\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.393820\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 3: 0.388261\n","Now at segment  4\n","Length of encoded genome sequence read in:  1531933\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38831 , cntTot: 100000 ; average prediction acc : 0.388310\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.388310\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 4: 0.388271\n","Now at segment  5\n","Length of encoded genome sequence read in:  1531933\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 39290 , cntTot: 100000 ; average prediction acc : 0.392900\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.392900\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 5: 0.389042\n","Now at segment  6\n","Length of encoded genome sequence read in:  1531933\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38802 , cntTot: 100000 ; average prediction acc : 0.388020\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.388020\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 6: 0.388896\n","Now at segment  7\n","Length of encoded genome sequence read in:  1531933\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38857 , cntTot: 100000 ; average prediction acc : 0.388570\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.388570\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 7: 0.388856\n","Now at segment  8\n","Length of encoded genome sequence read in:  1531933\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 39128 , cntTot: 100000 ; average prediction acc : 0.391280\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.391280\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 8: 0.389125\n","Now at segment  9\n","Length of encoded genome sequence read in:  1531933\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38661 , cntTot: 100000 ; average prediction acc : 0.386610\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.386610\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 9: 0.388873\n","Now at segment  10\n","Length of encoded genome sequence read in:  1531933\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38568 , cntTot: 100000 ; average prediction acc : 0.385680\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.385680\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 10: 0.388583\n","Now at segment  11\n","Length of encoded genome sequence read in:  1531933\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38903 , cntTot: 100000 ; average prediction acc : 0.389030\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.389030\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 11: 0.388620\n","Now at segment  12\n","Length of encoded genome sequence read in:  1531933\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38765 , cntTot: 100000 ; average prediction acc : 0.387650\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.387650\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 12: 0.388546\n","Now at segment  13\n","Length of encoded genome sequence read in:  1531933\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38662 , cntTot: 100000 ; average prediction acc : 0.386620\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.386620\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 13: 0.388408\n","Now at segment  14\n","Length of encoded genome sequence read in:  1531933\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38890 , cntTot: 100000 ; average prediction acc : 0.388900\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.388900\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 14: 0.388441\n","Reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 100\n","Fast reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 100\n","OBS: no file containing exonic info was provided, so exonic status is set to 0 from 0 - 100000001\n","Genome data file 1st line:\n","  >ref|NC_001137| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=V]\n","\n","Found data for this chromosome: ref|NC_001137| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=V]\n","60 60 60\n","[['ref|NC_001137| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=V]', 576874]]\n","[]\n","Length of genome sequence read in:576874\n","Length of exonic-info sequence read in:576874\n","ACGTacgt checked 0 tokens\n","Length genome sequence, ACGT's and W's:576874\n","Of these 0 are W's\n","Length genome sequence, only ACGT's:576874\n","Directory /content/drive/MyDrive/DNA_proj/results/predictions/R64_chr5/ created. Output will be placed there.\n","nrSegments:  5\n","Now at segment  0\n","Length of encoded genome sequence read in:  576874\n","Nr of samples to be generated:  99950\n","Nr of sampled positions: 99950\n","Yt shape  (99950, 4)\n","Xt_lr shape  (99950, 2, 50, 4)\n","Nr of samples: 99950 ; of which are predicted: 99950\n","lSamples  99950 , nrSteps  99950\n","cntCorr: 37746 , cntTot: 99950 ; average prediction acc : 0.377649\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.377649\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 0: 0.377649\n","Now at segment  1\n","Length of encoded genome sequence read in:  576874\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38442 , cntTot: 100000 ; average prediction acc : 0.384420\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.384420\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 1: 0.381035\n","Now at segment  2\n","Length of encoded genome sequence read in:  576874\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38379 , cntTot: 100000 ; average prediction acc : 0.383790\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.383790\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 2: 0.381954\n","Now at segment  3\n","Length of encoded genome sequence read in:  576874\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38791 , cntTot: 100000 ; average prediction acc : 0.387910\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.387910\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 3: 0.383443\n","Now at segment  4\n","Length of encoded genome sequence read in:  576874\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38139 , cntTot: 100000 ; average prediction acc : 0.381390\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.381390\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 4: 0.383032\n","Reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 100\n","Fast reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 100\n","OBS: no file containing exonic info was provided, so exonic status is set to 0 from 0 - 100000001\n","Genome data file 1st line:\n","  >ref|NC_001138| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VI]\n","\n","Found data for this chromosome: ref|NC_001138| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VI]\n","60 60 60\n","[['ref|NC_001138| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VI]', 270161]]\n","[]\n","Length of genome sequence read in:270161\n","Length of exonic-info sequence read in:270161\n","ACGTacgt checked 0 tokens\n","Length genome sequence, ACGT's and W's:270161\n","Of these 0 are W's\n","Length genome sequence, only ACGT's:270161\n","Directory /content/drive/MyDrive/DNA_proj/results/predictions/R64_chr6/ created. Output will be placed there.\n","nrSegments:  2\n","Now at segment  0\n","Length of encoded genome sequence read in:  270161\n","Nr of samples to be generated:  99950\n","Nr of sampled positions: 99950\n","Yt shape  (99950, 4)\n","Xt_lr shape  (99950, 2, 50, 4)\n","Nr of samples: 99950 ; of which are predicted: 99950\n","lSamples  99950 , nrSteps  99950\n","cntCorr: 37893 , cntTot: 99950 ; average prediction acc : 0.379120\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.379120\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 0: 0.379120\n","Now at segment  1\n","Length of encoded genome sequence read in:  270161\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38924 , cntTot: 100000 ; average prediction acc : 0.389240\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.389240\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 1: 0.384181\n","Reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 100\n","Fast reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 100\n","OBS: no file containing exonic info was provided, so exonic status is set to 0 from 0 - 100000001\n","Genome data file 1st line:\n","  >ref|NC_001139| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VII]\n","\n","Found data for this chromosome: ref|NC_001139| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VII]\n","60 60 60\n","[['ref|NC_001139| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VII]', 1090940]]\n","[]\n","Length of genome sequence read in:1090940\n","Length of exonic-info sequence read in:1090940\n","ACGTacgt checked 0 tokens\n","Length genome sequence, ACGT's and W's:1090940\n","Of these 0 are W's\n","Length genome sequence, only ACGT's:1090940\n","Directory /content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/ created. Output will be placed there.\n","nrSegments:  10\n","Now at segment  0\n","Length of encoded genome sequence read in:  1090940\n","Nr of samples to be generated:  99950\n","Nr of sampled positions: 99950\n","Yt shape  (99950, 4)\n","Xt_lr shape  (99950, 2, 50, 4)\n","Nr of samples: 99950 ; of which are predicted: 99950\n","lSamples  99950 , nrSteps  99950\n","cntCorr: 38767 , cntTot: 99950 ; average prediction acc : 0.387864\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.387864\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 0: 0.387864\n","Now at segment  1\n","Length of encoded genome sequence read in:  1090940\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38784 , cntTot: 100000 ; average prediction acc : 0.387840\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.387840\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 1: 0.387852\n","Now at segment  2\n","Length of encoded genome sequence read in:  1090940\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38343 , cntTot: 100000 ; average prediction acc : 0.383430\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.383430\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 2: 0.386378\n","Now at segment  3\n","Length of encoded genome sequence read in:  1090940\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38822 , cntTot: 100000 ; average prediction acc : 0.388220\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.388220\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 3: 0.386838\n","Now at segment  4\n","Length of encoded genome sequence read in:  1090940\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 39015 , cntTot: 100000 ; average prediction acc : 0.390150\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.390150\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 4: 0.387501\n","Now at segment  5\n","Length of encoded genome sequence read in:  1090940\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 37850 , cntTot: 100000 ; average prediction acc : 0.378500\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.378500\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 5: 0.386001\n","Now at segment  6\n","Length of encoded genome sequence read in:  1090940\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 39154 , cntTot: 100000 ; average prediction acc : 0.391540\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.391540\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 6: 0.386792\n","Now at segment  7\n","Length of encoded genome sequence read in:  1090940\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38584 , cntTot: 100000 ; average prediction acc : 0.385840\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.385840\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 7: 0.386673\n","Now at segment  8\n","Length of encoded genome sequence read in:  1090940\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 39181 , cntTot: 100000 ; average prediction acc : 0.391810\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.391810\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 8: 0.387244\n","Now at segment  9\n","Length of encoded genome sequence read in:  1090940\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38486 , cntTot: 100000 ; average prediction acc : 0.384860\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.384860\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 9: 0.387005\n","Reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 100\n","Fast reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 100\n","OBS: no file containing exonic info was provided, so exonic status is set to 0 from 0 - 100000001\n","Genome data file 1st line:\n","  >ref|NC_001140| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VIII]\n","\n","Found data for this chromosome: ref|NC_001140| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VIII]\n","60 60 60\n","[['ref|NC_001140| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VIII]', 562643]]\n","[]\n","Length of genome sequence read in:562643\n","Length of exonic-info sequence read in:562643\n","ACGTacgt checked 0 tokens\n","Length genome sequence, ACGT's and W's:562643\n","Of these 0 are W's\n","Length genome sequence, only ACGT's:562643\n","Directory /content/drive/MyDrive/DNA_proj/results/predictions/R64_chr8/ created. Output will be placed there.\n","nrSegments:  5\n","Now at segment  0\n","Length of encoded genome sequence read in:  562643\n","Nr of samples to be generated:  99950\n","Nr of sampled positions: 99950\n","Yt shape  (99950, 4)\n","Xt_lr shape  (99950, 2, 50, 4)\n","Nr of samples: 99950 ; of which are predicted: 99950\n","lSamples  99950 , nrSteps  99950\n","cntCorr: 37862 , cntTot: 99950 ; average prediction acc : 0.378809\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.378809\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 0: 0.378809\n","Now at segment  1\n","Length of encoded genome sequence read in:  562643\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38443 , cntTot: 100000 ; average prediction acc : 0.384430\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.384430\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 1: 0.381620\n","Now at segment  2\n","Length of encoded genome sequence read in:  562643\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38745 , cntTot: 100000 ; average prediction acc : 0.387450\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.387450\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 2: 0.383564\n","Now at segment  3\n","Length of encoded genome sequence read in:  562643\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38117 , cntTot: 100000 ; average prediction acc : 0.381170\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.381170\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 3: 0.382965\n","Now at segment  4\n","Length of encoded genome sequence read in:  562643\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38193 , cntTot: 100000 ; average prediction acc : 0.381930\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.381930\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 4: 0.382758\n","Reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 100\n","Fast reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 100\n","OBS: no file containing exonic info was provided, so exonic status is set to 0 from 0 - 100000001\n","Genome data file 1st line:\n","  >ref|NC_001141| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=IX]\n","\n","Found data for this chromosome: ref|NC_001141| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=IX]\n","60 60 60\n","[['ref|NC_001141| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=IX]', 439888]]\n","[]\n","Length of genome sequence read in:439888\n","Length of exonic-info sequence read in:439888\n","ACGTacgt checked 0 tokens\n","Length genome sequence, ACGT's and W's:439888\n","Of these 0 are W's\n","Length genome sequence, only ACGT's:439888\n","Directory /content/drive/MyDrive/DNA_proj/results/predictions/R64_chr9/ created. Output will be placed there.\n","nrSegments:  4\n","Now at segment  0\n","Length of encoded genome sequence read in:  439888\n","Nr of samples to be generated:  99950\n","Nr of sampled positions: 99950\n","Yt shape  (99950, 4)\n","Xt_lr shape  (99950, 2, 50, 4)\n","Nr of samples: 99950 ; of which are predicted: 99950\n","lSamples  99950 , nrSteps  99950\n","cntCorr: 38361 , cntTot: 99950 ; average prediction acc : 0.383802\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.383802\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 0: 0.383802\n","Now at segment  1\n","Length of encoded genome sequence read in:  439888\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 39048 , cntTot: 100000 ; average prediction acc : 0.390480\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.390480\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 1: 0.387142\n","Now at segment  2\n","Length of encoded genome sequence read in:  439888\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38339 , cntTot: 100000 ; average prediction acc : 0.383390\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.383390\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 2: 0.385891\n","Now at segment  3\n","Length of encoded genome sequence read in:  439888\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 37891 , cntTot: 100000 ; average prediction acc : 0.378910\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.378910\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 3: 0.384146\n","Reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 100\n","Fast reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 100\n","OBS: no file containing exonic info was provided, so exonic status is set to 0 from 0 - 100000001\n","Genome data file 1st line:\n","  >ref|NC_001142| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=X]\n","\n","Found data for this chromosome: ref|NC_001142| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=X]\n","60 60 60\n","[['ref|NC_001142| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=X]', 745751]]\n","[]\n","Length of genome sequence read in:745751\n","Length of exonic-info sequence read in:745751\n","ACGTacgt checked 0 tokens\n","Length genome sequence, ACGT's and W's:745751\n","Of these 0 are W's\n","Length genome sequence, only ACGT's:745751\n","Directory /content/drive/MyDrive/DNA_proj/results/predictions/R64_chr10/ created. Output will be placed there.\n","nrSegments:  7\n","Now at segment  0\n","Length of encoded genome sequence read in:  745751\n","Nr of samples to be generated:  99950\n","Nr of sampled positions: 99950\n","Yt shape  (99950, 4)\n","Xt_lr shape  (99950, 2, 50, 4)\n","Nr of samples: 99950 ; of which are predicted: 99950\n","lSamples  99950 , nrSteps  99950\n","cntCorr: 37867 , cntTot: 99950 ; average prediction acc : 0.378859\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.378859\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 0: 0.378859\n","Now at segment  1\n","Length of encoded genome sequence read in:  745751\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38620 , cntTot: 100000 ; average prediction acc : 0.386200\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.386200\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 1: 0.382531\n","Now at segment  2\n","Length of encoded genome sequence read in:  745751\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38509 , cntTot: 100000 ; average prediction acc : 0.385090\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.385090\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 2: 0.383384\n","Now at segment  3\n","Length of encoded genome sequence read in:  745751\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38765 , cntTot: 100000 ; average prediction acc : 0.387650\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.387650\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 3: 0.384451\n","Now at segment  4\n","Length of encoded genome sequence read in:  745751\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38193 , cntTot: 100000 ; average prediction acc : 0.381930\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.381930\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 4: 0.383946\n","Now at segment  5\n","Length of encoded genome sequence read in:  745751\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38473 , cntTot: 100000 ; average prediction acc : 0.384730\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.384730\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 5: 0.384077\n","Now at segment  6\n","Length of encoded genome sequence read in:  745751\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38585 , cntTot: 100000 ; average prediction acc : 0.385850\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.385850\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 6: 0.384330\n","Reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 100\n","Fast reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 100\n","OBS: no file containing exonic info was provided, so exonic status is set to 0 from 0 - 100000001\n","Genome data file 1st line:\n","  >ref|NC_001143| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XI]\n","\n","Found data for this chromosome: ref|NC_001143| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XI]\n","60 60 60\n","[['ref|NC_001143| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XI]', 666816]]\n","[]\n","Length of genome sequence read in:666816\n","Length of exonic-info sequence read in:666816\n","ACGTacgt checked 0 tokens\n","Length genome sequence, ACGT's and W's:666816\n","Of these 0 are W's\n","Length genome sequence, only ACGT's:666816\n","Directory /content/drive/MyDrive/DNA_proj/results/predictions/R64_chr11/ created. Output will be placed there.\n","nrSegments:  6\n","Now at segment  0\n","Length of encoded genome sequence read in:  666816\n","Nr of samples to be generated:  99950\n","Nr of sampled positions: 99950\n","Yt shape  (99950, 4)\n","Xt_lr shape  (99950, 2, 50, 4)\n","Nr of samples: 99950 ; of which are predicted: 99950\n","lSamples  99950 , nrSteps  99950\n","cntCorr: 37612 , cntTot: 99950 ; average prediction acc : 0.376308\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.376308\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 0: 0.376308\n","Now at segment  1\n","Length of encoded genome sequence read in:  666816\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38989 , cntTot: 100000 ; average prediction acc : 0.389890\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.389890\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 1: 0.383101\n","Now at segment  2\n","Length of encoded genome sequence read in:  666816\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38962 , cntTot: 100000 ; average prediction acc : 0.389620\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.389620\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 2: 0.385274\n","Now at segment  3\n","Length of encoded genome sequence read in:  666816\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38842 , cntTot: 100000 ; average prediction acc : 0.388420\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.388420\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 3: 0.386061\n","Now at segment  4\n","Length of encoded genome sequence read in:  666816\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38953 , cntTot: 100000 ; average prediction acc : 0.389530\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.389530\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 4: 0.386755\n","Now at segment  5\n","Length of encoded genome sequence read in:  666816\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38692 , cntTot: 100000 ; average prediction acc : 0.386920\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.386920\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 5: 0.386782\n","Reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 100\n","Fast reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 100\n","OBS: no file containing exonic info was provided, so exonic status is set to 0 from 0 - 100000001\n","Genome data file 1st line:\n","  >ref|NC_001144| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XII]\n","\n","Found data for this chromosome: ref|NC_001144| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XII]\n","60 60 60\n","[['ref|NC_001144| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XII]', 1078177]]\n","[]\n","Length of genome sequence read in:1078177\n","Length of exonic-info sequence read in:1078177\n","ACGTacgt checked 0 tokens\n","Length genome sequence, ACGT's and W's:1078177\n","Of these 0 are W's\n","Length genome sequence, only ACGT's:1078177\n","Directory /content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/ created. Output will be placed there.\n","nrSegments:  10\n","Now at segment  0\n","Length of encoded genome sequence read in:  1078177\n","Nr of samples to be generated:  99950\n","Nr of sampled positions: 99950\n","Yt shape  (99950, 4)\n","Xt_lr shape  (99950, 2, 50, 4)\n","Nr of samples: 99950 ; of which are predicted: 99950\n","lSamples  99950 , nrSteps  99950\n","cntCorr: 37849 , cntTot: 99950 ; average prediction acc : 0.378679\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.378679\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 0: 0.378679\n","Now at segment  1\n","Length of encoded genome sequence read in:  1078177\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38661 , cntTot: 100000 ; average prediction acc : 0.386610\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.386610\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 1: 0.382646\n","Now at segment  2\n","Length of encoded genome sequence read in:  1078177\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38723 , cntTot: 100000 ; average prediction acc : 0.387230\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.387230\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 2: 0.384174\n","Now at segment  3\n","Length of encoded genome sequence read in:  1078177\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38241 , cntTot: 100000 ; average prediction acc : 0.382410\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.382410\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 3: 0.383733\n","Now at segment  4\n","Length of encoded genome sequence read in:  1078177\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 37030 , cntTot: 100000 ; average prediction acc : 0.370300\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.370300\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 4: 0.381046\n","Now at segment  5\n","Length of encoded genome sequence read in:  1078177\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 39064 , cntTot: 100000 ; average prediction acc : 0.390640\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.390640\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 5: 0.382645\n","Now at segment  6\n","Length of encoded genome sequence read in:  1078177\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38854 , cntTot: 100000 ; average prediction acc : 0.388540\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.388540\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 6: 0.383487\n","Now at segment  7\n","Length of encoded genome sequence read in:  1078177\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 39051 , cntTot: 100000 ; average prediction acc : 0.390510\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.390510\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 7: 0.384365\n","Now at segment  8\n","Length of encoded genome sequence read in:  1078177\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38435 , cntTot: 100000 ; average prediction acc : 0.384350\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.384350\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 8: 0.384364\n","Now at segment  9\n","Length of encoded genome sequence read in:  1078177\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 37962 , cntTot: 100000 ; average prediction acc : 0.379620\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.379620\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 9: 0.383889\n","Reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 100\n","Fast reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 100\n","OBS: no file containing exonic info was provided, so exonic status is set to 0 from 0 - 100000001\n","Genome data file 1st line:\n","  >ref|NC_001145| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XIII]\n","\n","Found data for this chromosome: ref|NC_001145| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XIII]\n","60 60 60\n","[['ref|NC_001145| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XIII]', 924431]]\n","[]\n","Length of genome sequence read in:924431\n","Length of exonic-info sequence read in:924431\n","ACGTacgt checked 0 tokens\n","Length genome sequence, ACGT's and W's:924431\n","Of these 0 are W's\n","Length genome sequence, only ACGT's:924431\n","Directory /content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/ created. Output will be placed there.\n","nrSegments:  9\n","Now at segment  0\n","Length of encoded genome sequence read in:  924431\n","Nr of samples to be generated:  99950\n","Nr of sampled positions: 99950\n","Yt shape  (99950, 4)\n","Xt_lr shape  (99950, 2, 50, 4)\n","Nr of samples: 99950 ; of which are predicted: 99950\n","lSamples  99950 , nrSteps  99950\n","cntCorr: 38777 , cntTot: 99950 ; average prediction acc : 0.387964\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.387964\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 0: 0.387964\n","Now at segment  1\n","Length of encoded genome sequence read in:  924431\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38597 , cntTot: 100000 ; average prediction acc : 0.385970\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.385970\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 1: 0.386967\n","Now at segment  2\n","Length of encoded genome sequence read in:  924431\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38889 , cntTot: 100000 ; average prediction acc : 0.388890\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.388890\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 2: 0.387608\n","Now at segment  3\n","Length of encoded genome sequence read in:  924431\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38437 , cntTot: 100000 ; average prediction acc : 0.384370\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.384370\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 3: 0.386798\n","Now at segment  4\n","Length of encoded genome sequence read in:  924431\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38665 , cntTot: 100000 ; average prediction acc : 0.386650\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.386650\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 4: 0.386769\n","Now at segment  5\n","Length of encoded genome sequence read in:  924431\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38868 , cntTot: 100000 ; average prediction acc : 0.388680\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.388680\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 5: 0.387087\n","Now at segment  6\n","Length of encoded genome sequence read in:  924431\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38810 , cntTot: 100000 ; average prediction acc : 0.388100\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.388100\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 6: 0.387232\n","Now at segment  7\n","Length of encoded genome sequence read in:  924431\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38892 , cntTot: 100000 ; average prediction acc : 0.388920\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.388920\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 7: 0.387443\n","Now at segment  8\n","Length of encoded genome sequence read in:  924431\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38447 , cntTot: 100000 ; average prediction acc : 0.384470\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.384470\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 8: 0.387113\n","Reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 100\n","Fast reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 100\n","OBS: no file containing exonic info was provided, so exonic status is set to 0 from 0 - 100000001\n","Genome data file 1st line:\n","  >ref|NC_001146| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XIV]\n","\n","Found data for this chromosome: ref|NC_001146| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XIV]\n","60 60 60\n","[['ref|NC_001146| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XIV]', 784333]]\n","[]\n","Length of genome sequence read in:784333\n","Length of exonic-info sequence read in:784333\n","ACGTacgt checked 0 tokens\n","Length genome sequence, ACGT's and W's:784333\n","Of these 0 are W's\n","Length genome sequence, only ACGT's:784333\n","Directory /content/drive/MyDrive/DNA_proj/results/predictions/R64_chr14/ created. Output will be placed there.\n","nrSegments:  7\n","Now at segment  0\n","Length of encoded genome sequence read in:  784333\n","Nr of samples to be generated:  99950\n","Nr of sampled positions: 99950\n","Yt shape  (99950, 4)\n","Xt_lr shape  (99950, 2, 50, 4)\n","Nr of samples: 99950 ; of which are predicted: 99950\n","lSamples  99950 , nrSteps  99950\n","cntCorr: 37360 , cntTot: 99950 ; average prediction acc : 0.373787\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.373787\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 0: 0.373787\n","Now at segment  1\n","Length of encoded genome sequence read in:  784333\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38404 , cntTot: 100000 ; average prediction acc : 0.384040\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.384040\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 1: 0.378915\n","Now at segment  2\n","Length of encoded genome sequence read in:  784333\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 39373 , cntTot: 100000 ; average prediction acc : 0.393730\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.393730\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 2: 0.383854\n","Now at segment  3\n","Length of encoded genome sequence read in:  784333\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38506 , cntTot: 100000 ; average prediction acc : 0.385060\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.385060\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 3: 0.384156\n","Now at segment  4\n","Length of encoded genome sequence read in:  784333\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38656 , cntTot: 100000 ; average prediction acc : 0.386560\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.386560\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 4: 0.384636\n","Now at segment  5\n","Length of encoded genome sequence read in:  784333\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38180 , cntTot: 100000 ; average prediction acc : 0.381800\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.381800\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 5: 0.384164\n","Now at segment  6\n","Length of encoded genome sequence read in:  784333\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38129 , cntTot: 100000 ; average prediction acc : 0.381290\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.381290\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 6: 0.383753\n","Reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 100\n","Fast reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 100\n","OBS: no file containing exonic info was provided, so exonic status is set to 0 from 0 - 100000001\n","Genome data file 1st line:\n","  >ref|NC_001147| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XV]\n","\n","Found data for this chromosome: ref|NC_001147| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XV]\n","60 60 60\n","[['ref|NC_001147| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XV]', 1091291]]\n","[]\n","Length of genome sequence read in:1091291\n","Length of exonic-info sequence read in:1091291\n","ACGTacgt checked 0 tokens\n","Length genome sequence, ACGT's and W's:1091291\n","Of these 0 are W's\n","Length genome sequence, only ACGT's:1091291\n","Directory /content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/ created. Output will be placed there.\n","nrSegments:  10\n","Now at segment  0\n","Length of encoded genome sequence read in:  1091291\n","Nr of samples to be generated:  99950\n","Nr of sampled positions: 99950\n","Yt shape  (99950, 4)\n","Xt_lr shape  (99950, 2, 50, 4)\n","Nr of samples: 99950 ; of which are predicted: 99950\n","lSamples  99950 , nrSteps  99950\n","cntCorr: 37578 , cntTot: 99950 ; average prediction acc : 0.375968\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.375968\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 0: 0.375968\n","Now at segment  1\n","Length of encoded genome sequence read in:  1091291\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38555 , cntTot: 100000 ; average prediction acc : 0.385550\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.385550\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 1: 0.380760\n","Now at segment  2\n","Length of encoded genome sequence read in:  1091291\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38412 , cntTot: 100000 ; average prediction acc : 0.384120\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.384120\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 2: 0.381880\n","Now at segment  3\n","Length of encoded genome sequence read in:  1091291\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38320 , cntTot: 100000 ; average prediction acc : 0.383200\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.383200\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 3: 0.382210\n","Now at segment  4\n","Length of encoded genome sequence read in:  1091291\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38554 , cntTot: 100000 ; average prediction acc : 0.385540\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.385540\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 4: 0.382876\n","Now at segment  5\n","Length of encoded genome sequence read in:  1091291\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38659 , cntTot: 100000 ; average prediction acc : 0.386590\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.386590\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 5: 0.383495\n","Now at segment  6\n","Length of encoded genome sequence read in:  1091291\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38696 , cntTot: 100000 ; average prediction acc : 0.386960\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.386960\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 6: 0.383990\n","Now at segment  7\n","Length of encoded genome sequence read in:  1091291\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 39005 , cntTot: 100000 ; average prediction acc : 0.390050\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.390050\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 7: 0.384748\n","Now at segment  8\n","Length of encoded genome sequence read in:  1091291\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38564 , cntTot: 100000 ; average prediction acc : 0.385640\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.385640\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 8: 0.384847\n","Now at segment  9\n","Length of encoded genome sequence read in:  1091291\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38236 , cntTot: 100000 ; average prediction acc : 0.382360\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.382360\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 9: 0.384598\n","Reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 100\n","Fast reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 100\n","OBS: no file containing exonic info was provided, so exonic status is set to 0 from 0 - 100000001\n","Genome data file 1st line:\n","  >ref|NC_001148| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XVI]\n","\n","Found data for this chromosome: ref|NC_001148| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XVI]\n","60 60 60\n","[['ref|NC_001148| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XVI]', 948066]]\n","[]\n","Length of genome sequence read in:948066\n","Length of exonic-info sequence read in:948066\n","ACGTacgt checked 0 tokens\n","Length genome sequence, ACGT's and W's:948066\n","Of these 0 are W's\n","Length genome sequence, only ACGT's:948066\n","Directory /content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/ created. Output will be placed there.\n","nrSegments:  9\n","Now at segment  0\n","Length of encoded genome sequence read in:  948066\n","Nr of samples to be generated:  99950\n","Nr of sampled positions: 99950\n","Yt shape  (99950, 4)\n","Xt_lr shape  (99950, 2, 50, 4)\n","Nr of samples: 99950 ; of which are predicted: 99950\n","lSamples  99950 , nrSteps  99950\n","cntCorr: 37833 , cntTot: 99950 ; average prediction acc : 0.378519\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.378519\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 0: 0.378519\n","Now at segment  1\n","Length of encoded genome sequence read in:  948066\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 39882 , cntTot: 100000 ; average prediction acc : 0.398820\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.398820\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 1: 0.388672\n","Now at segment  2\n","Length of encoded genome sequence read in:  948066\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38087 , cntTot: 100000 ; average prediction acc : 0.380870\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.380870\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 2: 0.386071\n","Now at segment  3\n","Length of encoded genome sequence read in:  948066\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38961 , cntTot: 100000 ; average prediction acc : 0.389610\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.389610\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 3: 0.386956\n","Now at segment  4\n","Length of encoded genome sequence read in:  948066\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38854 , cntTot: 100000 ; average prediction acc : 0.388540\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.388540\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 4: 0.387273\n","Now at segment  5\n","Length of encoded genome sequence read in:  948066\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38274 , cntTot: 100000 ; average prediction acc : 0.382740\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.382740\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 5: 0.386517\n","Now at segment  6\n","Length of encoded genome sequence read in:  948066\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38385 , cntTot: 100000 ; average prediction acc : 0.383850\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.383850\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 6: 0.386136\n","Now at segment  7\n","Length of encoded genome sequence read in:  948066\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38334 , cntTot: 100000 ; average prediction acc : 0.383340\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.383340\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 7: 0.385787\n","Now at segment  8\n","Length of encoded genome sequence read in:  948066\n","Nr of samples to be generated:  100000\n","Nr of sampled positions: 100000\n","Yt shape  (100000, 4)\n","Xt_lr shape  (100000, 2, 50, 4)\n","Nr of samples: 100000 ; of which are predicted: 100000\n","lSamples  100000 , nrSteps  100000\n","cntCorr: 38498 , cntTot: 100000 ; average prediction acc : 0.384980\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.384980\n","No repeat sections were recorded in the genome data.\n","Average prediction acc after segment 8: 0.385697\n"]}]},{"cell_type":"code","execution_count":17,"metadata":{"id":"yubV3C9Gqk1B","executionInfo":{"status":"ok","timestamp":1650141687747,"user_tz":-120,"elapsed":267,"user":{"displayName":"Christian Grønbæk","userId":"14909394381660051993"}}},"outputs":[],"source":["#Yeast:\n","segmentLength = 100000\n","\n","averageRevComplementary_b = 0\n","\n","chromosomeOrderList = ['R64_chr1', 'R64_chr2', 'R64_chr3', 'R64_chr4', 'R64_chr5', 'R64_chr6', 'R64_chr7', 'R64_chr8','R64_chr9', 'R64_chr10', 'R64_chr11', 'R64_chr12','R64_chr13', 'R64_chr14', 'R64_chr15', 'R64_chr16' ]\n","chromosomeDict = {'R64_chr1':[0,1e8], 'R64_chr2':[0,1e8], 'R64_chr3':[0,1e8], 'R64_chr4':[0,1e8], 'R64_chr5':[0,1e8], 'R64_chr6':[0,1e8], 'R64_chr7':[0,1e8], 'R64_chr8':[0,1e8],'R64_chr9':[0,1e8], 'R64_chr10':[0,1e8], 'R64_chr11':[0,1e8], 'R64_chr12':[0,1e8],'R64_chr13':[0,1e8], 'R64_chr14':[0,1e8], 'R64_chr15':[0,1e8], 'R64_chr16':[0,1e8]}\n","rootAnnotationFiles = r\"/content/drive/MyDrive/DNA_proj/data/S288C_reference_genome_R64-1-1_20110203/\"\n","annotationTypes = ['simpleRepeats']\n","\n"]},{"cell_type":"markdown","source":["This call computes accuracy-figures per segment and saves the results in dictionaries:  "],"metadata":{"id":"H0GaKif7F3Pe"}},{"cell_type":"code","source":["resultsDictByAnnoSeg, resultsDictByAnno  = stats.getAccuracyChromosomes(chromosomeOrderList = chromosomeOrderList, \n","                         rootOutput = rootOutput,\n","                         modelFileName = modelFileNameNN, \n","                         segmentLength = segmentLength,\n","                         averageRevComplementary_b = averageRevComplementary_b,\n","                         windowLength = windowLength,\n","                         stepSize = stepSize, \n","                         annotationTypes = annotationTypes,\n","                         rootAnnotationFiles = rootAnnotationFiles,\n","                         chromosomeDict = chromosomeDict)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"eok-LTR8KEQk","executionInfo":{"status":"ok","timestamp":1650138353612,"user_tz":-120,"elapsed":1427,"user":{"displayName":"Christian Grønbæk","userId":"14909394381660051993"}},"outputId":"47b8efc2-4e5a-444e-a8e3-a44f892c7cf1"},"execution_count":33,"outputs":[{"output_type":"stream","name":"stdout","text":["/content/drive/MyDrive/DNA_proj/data/S288C_reference_genome_R64-1-1_20110203/R64_chr1_annotationArray_simpleRepeats\n","Annotation file /content/drive/MyDrive/DNA_proj/data/S288C_reference_genome_R64-1-1_20110203/R64_chr1_annotationArray_simpleRepeats not found\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr1/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr1_seg100000_segment0_avgRevCompl0_win1_step1\n","0 (99950,)\n","99950\n","99950\n","Acc at R64_chr1_seg100000_segment0: 0.386830 based on corr 38664 and tot 99950\n","Acc at R64_chr1_seg100000_segment0 as recorded/done in computeAcc: 0.386830 , based on nr of corr 38664, tot cnt 99950\n","Acc at R64_chr1_seg100000_segment0 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr1/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr1_seg100000_segment1_avgRevCompl0_win1_step1\n","1 (100000,)\n","100000\n","100000\n","Acc at R64_chr1_seg100000_segment1: 0.383546 based on corr 38355 and tot 100000\n","Acc at R64_chr1_seg100000_segment1 as recorded/done in computeAcc: 0.383546 , based on nr of corr 38355, tot cnt 100000\n","Acc at R64_chr1_seg100000_segment1 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr1/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr1_seg100000_segment2_avgRevCompl0_win1_step1\n","/content/drive/MyDrive/DNA_proj/data/S288C_reference_genome_R64-1-1_20110203/R64_chr2_annotationArray_simpleRepeats\n","Annotation file /content/drive/MyDrive/DNA_proj/data/S288C_reference_genome_R64-1-1_20110203/R64_chr2_annotationArray_simpleRepeats not found\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr2_seg100000_segment0_avgRevCompl0_win1_step1\n","0 (99950,)\n","99950\n","99950\n","Acc at R64_chr2_seg100000_segment0: 0.381717 based on corr 38153 and tot 99950\n","Acc at R64_chr2_seg100000_segment0 as recorded/done in computeAcc: 0.381717 , based on nr of corr 38153, tot cnt 99950\n","Acc at R64_chr2_seg100000_segment0 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr2_seg100000_segment1_avgRevCompl0_win1_step1\n","1 (100000,)\n","100000\n","100000\n","Acc at R64_chr2_seg100000_segment1: 0.389136 based on corr 38914 and tot 100000\n","Acc at R64_chr2_seg100000_segment1 as recorded/done in computeAcc: 0.389136 , based on nr of corr 38914, tot cnt 100000\n","Acc at R64_chr2_seg100000_segment1 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr2_seg100000_segment2_avgRevCompl0_win1_step1\n","2 (100000,)\n","100000\n","100000\n","Acc at R64_chr2_seg100000_segment2: 0.380906 based on corr 38091 and tot 100000\n","Acc at R64_chr2_seg100000_segment2 as recorded/done in computeAcc: 0.380906 , based on nr of corr 38091, tot cnt 100000\n","Acc at R64_chr2_seg100000_segment2 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr2_seg100000_segment3_avgRevCompl0_win1_step1\n","3 (100000,)\n","100000\n","100000\n","Acc at R64_chr2_seg100000_segment3: 0.390136 based on corr 39014 and tot 100000\n","Acc at R64_chr2_seg100000_segment3 as recorded/done in computeAcc: 0.390136 , based on nr of corr 39014, tot cnt 100000\n","Acc at R64_chr2_seg100000_segment3 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr2_seg100000_segment4_avgRevCompl0_win1_step1\n","4 (100000,)\n","100000\n","100000\n","Acc at R64_chr2_seg100000_segment4: 0.393626 based on corr 39363 and tot 100000\n","Acc at R64_chr2_seg100000_segment4 as recorded/done in computeAcc: 0.393626 , based on nr of corr 39363, tot cnt 100000\n","Acc at R64_chr2_seg100000_segment4 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr2_seg100000_segment5_avgRevCompl0_win1_step1\n","5 (100000,)\n","100000\n","100000\n","Acc at R64_chr2_seg100000_segment5: 0.388366 based on corr 38837 and tot 100000\n","Acc at R64_chr2_seg100000_segment5 as recorded/done in computeAcc: 0.388366 , based on nr of corr 38837, tot cnt 100000\n","Acc at R64_chr2_seg100000_segment5 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr2_seg100000_segment6_avgRevCompl0_win1_step1\n","6 (100000,)\n","100000\n","100000\n","Acc at R64_chr2_seg100000_segment6: 0.383216 based on corr 38322 and tot 100000\n","Acc at R64_chr2_seg100000_segment6 as recorded/done in computeAcc: 0.383216 , based on nr of corr 38322, tot cnt 100000\n","Acc at R64_chr2_seg100000_segment6 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr2_seg100000_segment7_avgRevCompl0_win1_step1\n","7 (100000,)\n","100000\n","100000\n","Acc at R64_chr2_seg100000_segment7: 0.380686 based on corr 38069 and tot 100000\n","Acc at R64_chr2_seg100000_segment7 as recorded/done in computeAcc: 0.380686 , based on nr of corr 38069, tot cnt 100000\n","Acc at R64_chr2_seg100000_segment7 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr2_seg100000_segment8_avgRevCompl0_win1_step1\n","/content/drive/MyDrive/DNA_proj/data/S288C_reference_genome_R64-1-1_20110203/R64_chr3_annotationArray_simpleRepeats\n","Annotation file /content/drive/MyDrive/DNA_proj/data/S288C_reference_genome_R64-1-1_20110203/R64_chr3_annotationArray_simpleRepeats not found\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr3/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr3_seg100000_segment0_avgRevCompl0_win1_step1\n","0 (99950,)\n","99950\n","99950\n","Acc at R64_chr3_seg100000_segment0: 0.385619 based on corr 38543 and tot 99950\n","Acc at R64_chr3_seg100000_segment0 as recorded/done in computeAcc: 0.385619 , based on nr of corr 38543, tot cnt 99950\n","Acc at R64_chr3_seg100000_segment0 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr3/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr3_seg100000_segment1_avgRevCompl0_win1_step1\n","1 (100000,)\n","100000\n","100000\n","Acc at R64_chr3_seg100000_segment1: 0.382736 based on corr 38274 and tot 100000\n","Acc at R64_chr3_seg100000_segment1 as recorded/done in computeAcc: 0.382736 , based on nr of corr 38274, tot cnt 100000\n","Acc at R64_chr3_seg100000_segment1 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr3/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr3_seg100000_segment2_avgRevCompl0_win1_step1\n","2 (100000,)\n","100000\n","100000\n","Acc at R64_chr3_seg100000_segment2: 0.381956 based on corr 38196 and tot 100000\n","Acc at R64_chr3_seg100000_segment2 as recorded/done in computeAcc: 0.381956 , based on nr of corr 38196, tot cnt 100000\n","Acc at R64_chr3_seg100000_segment2 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr3/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr3_seg100000_segment3_avgRevCompl0_win1_step1\n","/content/drive/MyDrive/DNA_proj/data/S288C_reference_genome_R64-1-1_20110203/R64_chr4_annotationArray_simpleRepeats\n","Annotation file /content/drive/MyDrive/DNA_proj/data/S288C_reference_genome_R64-1-1_20110203/R64_chr4_annotationArray_simpleRepeats not found\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr4_seg100000_segment0_avgRevCompl0_win1_step1\n","0 (99950,)\n","99950\n","99950\n","Acc at R64_chr4_seg100000_segment0: 0.380927 based on corr 38074 and tot 99950\n","Acc at R64_chr4_seg100000_segment0 as recorded/done in computeAcc: 0.380927 , based on nr of corr 38074, tot cnt 99950\n","Acc at R64_chr4_seg100000_segment0 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr4_seg100000_segment1_avgRevCompl0_win1_step1\n","1 (100000,)\n","100000\n","100000\n","Acc at R64_chr4_seg100000_segment1: 0.392766 based on corr 39277 and tot 100000\n","Acc at R64_chr4_seg100000_segment1 as recorded/done in computeAcc: 0.392766 , based on nr of corr 39277, tot cnt 100000\n","Acc at R64_chr4_seg100000_segment1 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr4_seg100000_segment2_avgRevCompl0_win1_step1\n","2 (100000,)\n","100000\n","100000\n","Acc at R64_chr4_seg100000_segment2: 0.385516 based on corr 38552 and tot 100000\n","Acc at R64_chr4_seg100000_segment2 as recorded/done in computeAcc: 0.385516 , based on nr of corr 38552, tot cnt 100000\n","Acc at R64_chr4_seg100000_segment2 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr4_seg100000_segment3_avgRevCompl0_win1_step1\n","3 (100000,)\n","100000\n","100000\n","Acc at R64_chr4_seg100000_segment3: 0.393816 based on corr 39382 and tot 100000\n","Acc at R64_chr4_seg100000_segment3 as recorded/done in computeAcc: 0.393816 , based on nr of corr 39382, tot cnt 100000\n","Acc at R64_chr4_seg100000_segment3 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr4_seg100000_segment4_avgRevCompl0_win1_step1\n","4 (100000,)\n","100000\n","100000\n","Acc at R64_chr4_seg100000_segment4: 0.388306 based on corr 38831 and tot 100000\n","Acc at R64_chr4_seg100000_segment4 as recorded/done in computeAcc: 0.388306 , based on nr of corr 38831, tot cnt 100000\n","Acc at R64_chr4_seg100000_segment4 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr4_seg100000_segment5_avgRevCompl0_win1_step1\n","5 (100000,)\n","100000\n","100000\n","Acc at R64_chr4_seg100000_segment5: 0.392896 based on corr 39290 and tot 100000\n","Acc at R64_chr4_seg100000_segment5 as recorded/done in computeAcc: 0.392896 , based on nr of corr 39290, tot cnt 100000\n","Acc at R64_chr4_seg100000_segment5 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr4_seg100000_segment6_avgRevCompl0_win1_step1\n","6 (100000,)\n","100000\n","100000\n","Acc at R64_chr4_seg100000_segment6: 0.388016 based on corr 38802 and tot 100000\n","Acc at R64_chr4_seg100000_segment6 as recorded/done in computeAcc: 0.388016 , based on nr of corr 38802, tot cnt 100000\n","Acc at R64_chr4_seg100000_segment6 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr4_seg100000_segment7_avgRevCompl0_win1_step1\n","7 (100000,)\n","100000\n","100000\n","Acc at R64_chr4_seg100000_segment7: 0.388566 based on corr 38857 and tot 100000\n","Acc at R64_chr4_seg100000_segment7 as recorded/done in computeAcc: 0.388566 , based on nr of corr 38857, tot cnt 100000\n","Acc at R64_chr4_seg100000_segment7 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr4_seg100000_segment8_avgRevCompl0_win1_step1\n","8 (100000,)\n","100000\n","100000\n","Acc at R64_chr4_seg100000_segment8: 0.391276 based on corr 39128 and tot 100000\n","Acc at R64_chr4_seg100000_segment8 as recorded/done in computeAcc: 0.391276 , based on nr of corr 39128, tot cnt 100000\n","Acc at R64_chr4_seg100000_segment8 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr4_seg100000_segment9_avgRevCompl0_win1_step1\n","9 (100000,)\n","100000\n","100000\n","Acc at R64_chr4_seg100000_segment9: 0.386606 based on corr 38661 and tot 100000\n","Acc at R64_chr4_seg100000_segment9 as recorded/done in computeAcc: 0.386606 , based on nr of corr 38661, tot cnt 100000\n","Acc at R64_chr4_seg100000_segment9 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr4_seg100000_segment10_avgRevCompl0_win1_step1\n","10 (100000,)\n","100000\n","100000\n","Acc at R64_chr4_seg100000_segment10: 0.385676 based on corr 38568 and tot 100000\n","Acc at R64_chr4_seg100000_segment10 as recorded/done in computeAcc: 0.385676 , based on nr of corr 38568, tot cnt 100000\n","Acc at R64_chr4_seg100000_segment10 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr4_seg100000_segment11_avgRevCompl0_win1_step1\n","11 (100000,)\n","100000\n","100000\n","Acc at R64_chr4_seg100000_segment11: 0.389026 based on corr 38903 and tot 100000\n","Acc at R64_chr4_seg100000_segment11 as recorded/done in computeAcc: 0.389026 , based on nr of corr 38903, tot cnt 100000\n","Acc at R64_chr4_seg100000_segment11 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr4_seg100000_segment12_avgRevCompl0_win1_step1\n","12 (100000,)\n","100000\n","100000\n","Acc at R64_chr4_seg100000_segment12: 0.387646 based on corr 38765 and tot 100000\n","Acc at R64_chr4_seg100000_segment12 as recorded/done in computeAcc: 0.387646 , based on nr of corr 38765, tot cnt 100000\n","Acc at R64_chr4_seg100000_segment12 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr4_seg100000_segment13_avgRevCompl0_win1_step1\n","13 (100000,)\n","100000\n","100000\n","Acc at R64_chr4_seg100000_segment13: 0.386616 based on corr 38662 and tot 100000\n","Acc at R64_chr4_seg100000_segment13 as recorded/done in computeAcc: 0.386616 , based on nr of corr 38662, tot cnt 100000\n","Acc at R64_chr4_seg100000_segment13 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr4_seg100000_segment14_avgRevCompl0_win1_step1\n","14 (100000,)\n","100000\n","100000\n","Acc at R64_chr4_seg100000_segment14: 0.388896 based on corr 38890 and tot 100000\n","Acc at R64_chr4_seg100000_segment14 as recorded/done in computeAcc: 0.388896 , based on nr of corr 38890, tot cnt 100000\n","Acc at R64_chr4_seg100000_segment14 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr4_seg100000_segment15_avgRevCompl0_win1_step1\n","/content/drive/MyDrive/DNA_proj/data/S288C_reference_genome_R64-1-1_20110203/R64_chr5_annotationArray_simpleRepeats\n","Annotation file /content/drive/MyDrive/DNA_proj/data/S288C_reference_genome_R64-1-1_20110203/R64_chr5_annotationArray_simpleRepeats not found\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr5/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr5_seg100000_segment0_avgRevCompl0_win1_step1\n","0 (99950,)\n","99950\n","99950\n","Acc at R64_chr5_seg100000_segment0: 0.377645 based on corr 37746 and tot 99950\n","Acc at R64_chr5_seg100000_segment0 as recorded/done in computeAcc: 0.377645 , based on nr of corr 37746, tot cnt 99950\n","Acc at R64_chr5_seg100000_segment0 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr5/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr5_seg100000_segment1_avgRevCompl0_win1_step1\n","1 (100000,)\n","100000\n","100000\n","Acc at R64_chr5_seg100000_segment1: 0.384416 based on corr 38442 and tot 100000\n","Acc at R64_chr5_seg100000_segment1 as recorded/done in computeAcc: 0.384416 , based on nr of corr 38442, tot cnt 100000\n","Acc at R64_chr5_seg100000_segment1 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr5/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr5_seg100000_segment2_avgRevCompl0_win1_step1\n","2 (100000,)\n","100000\n","100000\n","Acc at R64_chr5_seg100000_segment2: 0.383786 based on corr 38379 and tot 100000\n","Acc at R64_chr5_seg100000_segment2 as recorded/done in computeAcc: 0.383786 , based on nr of corr 38379, tot cnt 100000\n","Acc at R64_chr5_seg100000_segment2 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr5/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr5_seg100000_segment3_avgRevCompl0_win1_step1\n","3 (100000,)\n","100000\n","100000\n","Acc at R64_chr5_seg100000_segment3: 0.387906 based on corr 38791 and tot 100000\n","Acc at R64_chr5_seg100000_segment3 as recorded/done in computeAcc: 0.387906 , based on nr of corr 38791, tot cnt 100000\n","Acc at R64_chr5_seg100000_segment3 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr5/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr5_seg100000_segment4_avgRevCompl0_win1_step1\n","4 (100000,)\n","100000\n","100000\n","Acc at R64_chr5_seg100000_segment4: 0.381386 based on corr 38139 and tot 100000\n","Acc at R64_chr5_seg100000_segment4 as recorded/done in computeAcc: 0.381386 , based on nr of corr 38139, tot cnt 100000\n","Acc at R64_chr5_seg100000_segment4 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr5/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr5_seg100000_segment5_avgRevCompl0_win1_step1\n","/content/drive/MyDrive/DNA_proj/data/S288C_reference_genome_R64-1-1_20110203/R64_chr6_annotationArray_simpleRepeats\n","Annotation file /content/drive/MyDrive/DNA_proj/data/S288C_reference_genome_R64-1-1_20110203/R64_chr6_annotationArray_simpleRepeats not found\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr6/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr6_seg100000_segment0_avgRevCompl0_win1_step1\n","0 (99950,)\n","99950\n","99950\n","Acc at R64_chr6_seg100000_segment0: 0.379116 based on corr 37893 and tot 99950\n","Acc at R64_chr6_seg100000_segment0 as recorded/done in computeAcc: 0.379116 , based on nr of corr 37893, tot cnt 99950\n","Acc at R64_chr6_seg100000_segment0 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr6/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr6_seg100000_segment1_avgRevCompl0_win1_step1\n","1 (100000,)\n","100000\n","100000\n","Acc at R64_chr6_seg100000_segment1: 0.389236 based on corr 38924 and tot 100000\n","Acc at R64_chr6_seg100000_segment1 as recorded/done in computeAcc: 0.389236 , based on nr of corr 38924, tot cnt 100000\n","Acc at R64_chr6_seg100000_segment1 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr6/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr6_seg100000_segment2_avgRevCompl0_win1_step1\n","/content/drive/MyDrive/DNA_proj/data/S288C_reference_genome_R64-1-1_20110203/R64_chr7_annotationArray_simpleRepeats\n","Annotation file /content/drive/MyDrive/DNA_proj/data/S288C_reference_genome_R64-1-1_20110203/R64_chr7_annotationArray_simpleRepeats not found\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr7_seg100000_segment0_avgRevCompl0_win1_step1\n","0 (99950,)\n","99950\n","99950\n","Acc at R64_chr7_seg100000_segment0: 0.387860 based on corr 38767 and tot 99950\n","Acc at R64_chr7_seg100000_segment0 as recorded/done in computeAcc: 0.387860 , based on nr of corr 38767, tot cnt 99950\n","Acc at R64_chr7_seg100000_segment0 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr7_seg100000_segment1_avgRevCompl0_win1_step1\n","1 (100000,)\n","100000\n","100000\n","Acc at R64_chr7_seg100000_segment1: 0.387836 based on corr 38784 and tot 100000\n","Acc at R64_chr7_seg100000_segment1 as recorded/done in computeAcc: 0.387836 , based on nr of corr 38784, tot cnt 100000\n","Acc at R64_chr7_seg100000_segment1 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr7_seg100000_segment2_avgRevCompl0_win1_step1\n","2 (100000,)\n","100000\n","100000\n","Acc at R64_chr7_seg100000_segment2: 0.383426 based on corr 38343 and tot 100000\n","Acc at R64_chr7_seg100000_segment2 as recorded/done in computeAcc: 0.383426 , based on nr of corr 38343, tot cnt 100000\n","Acc at R64_chr7_seg100000_segment2 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr7_seg100000_segment3_avgRevCompl0_win1_step1\n","3 (100000,)\n","100000\n","100000\n","Acc at R64_chr7_seg100000_segment3: 0.388216 based on corr 38822 and tot 100000\n","Acc at R64_chr7_seg100000_segment3 as recorded/done in computeAcc: 0.388216 , based on nr of corr 38822, tot cnt 100000\n","Acc at R64_chr7_seg100000_segment3 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr7_seg100000_segment4_avgRevCompl0_win1_step1\n","4 (100000,)\n","100000\n","100000\n","Acc at R64_chr7_seg100000_segment4: 0.390146 based on corr 39015 and tot 100000\n","Acc at R64_chr7_seg100000_segment4 as recorded/done in computeAcc: 0.390146 , based on nr of corr 39015, tot cnt 100000\n","Acc at R64_chr7_seg100000_segment4 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr7_seg100000_segment5_avgRevCompl0_win1_step1\n","5 (100000,)\n","100000\n","100000\n","Acc at R64_chr7_seg100000_segment5: 0.378496 based on corr 37850 and tot 100000\n","Acc at R64_chr7_seg100000_segment5 as recorded/done in computeAcc: 0.378496 , based on nr of corr 37850, tot cnt 100000\n","Acc at R64_chr7_seg100000_segment5 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr7_seg100000_segment6_avgRevCompl0_win1_step1\n","6 (100000,)\n","100000\n","100000\n","Acc at R64_chr7_seg100000_segment6: 0.391536 based on corr 39154 and tot 100000\n","Acc at R64_chr7_seg100000_segment6 as recorded/done in computeAcc: 0.391536 , based on nr of corr 39154, tot cnt 100000\n","Acc at R64_chr7_seg100000_segment6 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr7_seg100000_segment7_avgRevCompl0_win1_step1\n","7 (100000,)\n","100000\n","100000\n","Acc at R64_chr7_seg100000_segment7: 0.385836 based on corr 38584 and tot 100000\n","Acc at R64_chr7_seg100000_segment7 as recorded/done in computeAcc: 0.385836 , based on nr of corr 38584, tot cnt 100000\n","Acc at R64_chr7_seg100000_segment7 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr7_seg100000_segment8_avgRevCompl0_win1_step1\n","8 (100000,)\n","100000\n","100000\n","Acc at R64_chr7_seg100000_segment8: 0.391806 based on corr 39181 and tot 100000\n","Acc at R64_chr7_seg100000_segment8 as recorded/done in computeAcc: 0.391806 , based on nr of corr 39181, tot cnt 100000\n","Acc at R64_chr7_seg100000_segment8 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr7_seg100000_segment9_avgRevCompl0_win1_step1\n","9 (100000,)\n","100000\n","100000\n","Acc at R64_chr7_seg100000_segment9: 0.384856 based on corr 38486 and tot 100000\n","Acc at R64_chr7_seg100000_segment9 as recorded/done in computeAcc: 0.384856 , based on nr of corr 38486, tot cnt 100000\n","Acc at R64_chr7_seg100000_segment9 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr7_seg100000_segment10_avgRevCompl0_win1_step1\n","/content/drive/MyDrive/DNA_proj/data/S288C_reference_genome_R64-1-1_20110203/R64_chr8_annotationArray_simpleRepeats\n","Annotation file /content/drive/MyDrive/DNA_proj/data/S288C_reference_genome_R64-1-1_20110203/R64_chr8_annotationArray_simpleRepeats not found\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr8/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr8_seg100000_segment0_avgRevCompl0_win1_step1\n","0 (99950,)\n","99950\n","99950\n","Acc at R64_chr8_seg100000_segment0: 0.378806 based on corr 37862 and tot 99950\n","Acc at R64_chr8_seg100000_segment0 as recorded/done in computeAcc: 0.378806 , based on nr of corr 37862, tot cnt 99950\n","Acc at R64_chr8_seg100000_segment0 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr8/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr8_seg100000_segment1_avgRevCompl0_win1_step1\n","1 (100000,)\n","100000\n","100000\n","Acc at R64_chr8_seg100000_segment1: 0.384426 based on corr 38443 and tot 100000\n","Acc at R64_chr8_seg100000_segment1 as recorded/done in computeAcc: 0.384426 , based on nr of corr 38443, tot cnt 100000\n","Acc at R64_chr8_seg100000_segment1 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr8/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr8_seg100000_segment2_avgRevCompl0_win1_step1\n","2 (100000,)\n","100000\n","100000\n","Acc at R64_chr8_seg100000_segment2: 0.387446 based on corr 38745 and tot 100000\n","Acc at R64_chr8_seg100000_segment2 as recorded/done in computeAcc: 0.387446 , based on nr of corr 38745, tot cnt 100000\n","Acc at R64_chr8_seg100000_segment2 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr8/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr8_seg100000_segment3_avgRevCompl0_win1_step1\n","3 (100000,)\n","100000\n","100000\n","Acc at R64_chr8_seg100000_segment3: 0.381166 based on corr 38117 and tot 100000\n","Acc at R64_chr8_seg100000_segment3 as recorded/done in computeAcc: 0.381166 , based on nr of corr 38117, tot cnt 100000\n","Acc at R64_chr8_seg100000_segment3 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr8/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr8_seg100000_segment4_avgRevCompl0_win1_step1\n","4 (100000,)\n","100000\n","100000\n","Acc at R64_chr8_seg100000_segment4: 0.381926 based on corr 38193 and tot 100000\n","Acc at R64_chr8_seg100000_segment4 as recorded/done in computeAcc: 0.381926 , based on nr of corr 38193, tot cnt 100000\n","Acc at R64_chr8_seg100000_segment4 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr8/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr8_seg100000_segment5_avgRevCompl0_win1_step1\n","/content/drive/MyDrive/DNA_proj/data/S288C_reference_genome_R64-1-1_20110203/R64_chr9_annotationArray_simpleRepeats\n","Annotation file /content/drive/MyDrive/DNA_proj/data/S288C_reference_genome_R64-1-1_20110203/R64_chr9_annotationArray_simpleRepeats not found\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr9/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr9_seg100000_segment0_avgRevCompl0_win1_step1\n","0 (99950,)\n","99950\n","99950\n","Acc at R64_chr9_seg100000_segment0: 0.383798 based on corr 38361 and tot 99950\n","Acc at R64_chr9_seg100000_segment0 as recorded/done in computeAcc: 0.383798 , based on nr of corr 38361, tot cnt 99950\n","Acc at R64_chr9_seg100000_segment0 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr9/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr9_seg100000_segment1_avgRevCompl0_win1_step1\n","1 (100000,)\n","100000\n","100000\n","Acc at R64_chr9_seg100000_segment1: 0.390476 based on corr 39048 and tot 100000\n","Acc at R64_chr9_seg100000_segment1 as recorded/done in computeAcc: 0.390476 , based on nr of corr 39048, tot cnt 100000\n","Acc at R64_chr9_seg100000_segment1 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr9/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr9_seg100000_segment2_avgRevCompl0_win1_step1\n","2 (100000,)\n","100000\n","100000\n","Acc at R64_chr9_seg100000_segment2: 0.383386 based on corr 38339 and tot 100000\n","Acc at R64_chr9_seg100000_segment2 as recorded/done in computeAcc: 0.383386 , based on nr of corr 38339, tot cnt 100000\n","Acc at R64_chr9_seg100000_segment2 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr9/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr9_seg100000_segment3_avgRevCompl0_win1_step1\n","3 (100000,)\n","100000\n","100000\n","Acc at R64_chr9_seg100000_segment3: 0.378906 based on corr 37891 and tot 100000\n","Acc at R64_chr9_seg100000_segment3 as recorded/done in computeAcc: 0.378906 , based on nr of corr 37891, tot cnt 100000\n","Acc at R64_chr9_seg100000_segment3 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr9/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr9_seg100000_segment4_avgRevCompl0_win1_step1\n","/content/drive/MyDrive/DNA_proj/data/S288C_reference_genome_R64-1-1_20110203/R64_chr10_annotationArray_simpleRepeats\n","Annotation file /content/drive/MyDrive/DNA_proj/data/S288C_reference_genome_R64-1-1_20110203/R64_chr10_annotationArray_simpleRepeats not found\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr10/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr10_seg100000_segment0_avgRevCompl0_win1_step1\n","0 (99950,)\n","99950\n","99950\n","Acc at R64_chr10_seg100000_segment0: 0.378856 based on corr 37867 and tot 99950\n","Acc at R64_chr10_seg100000_segment0 as recorded/done in computeAcc: 0.378856 , based on nr of corr 37867, tot cnt 99950\n","Acc at R64_chr10_seg100000_segment0 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr10/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr10_seg100000_segment1_avgRevCompl0_win1_step1\n","1 (100000,)\n","100000\n","100000\n","Acc at R64_chr10_seg100000_segment1: 0.386196 based on corr 38620 and tot 100000\n","Acc at R64_chr10_seg100000_segment1 as recorded/done in computeAcc: 0.386196 , based on nr of corr 38620, tot cnt 100000\n","Acc at R64_chr10_seg100000_segment1 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr10/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr10_seg100000_segment2_avgRevCompl0_win1_step1\n","2 (100000,)\n","100000\n","100000\n","Acc at R64_chr10_seg100000_segment2: 0.385086 based on corr 38509 and tot 100000\n","Acc at R64_chr10_seg100000_segment2 as recorded/done in computeAcc: 0.385086 , based on nr of corr 38509, tot cnt 100000\n","Acc at R64_chr10_seg100000_segment2 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr10/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr10_seg100000_segment3_avgRevCompl0_win1_step1\n","3 (100000,)\n","100000\n","100000\n","Acc at R64_chr10_seg100000_segment3: 0.387646 based on corr 38765 and tot 100000\n","Acc at R64_chr10_seg100000_segment3 as recorded/done in computeAcc: 0.387646 , based on nr of corr 38765, tot cnt 100000\n","Acc at R64_chr10_seg100000_segment3 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr10/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr10_seg100000_segment4_avgRevCompl0_win1_step1\n","4 (100000,)\n","100000\n","100000\n","Acc at R64_chr10_seg100000_segment4: 0.381926 based on corr 38193 and tot 100000\n","Acc at R64_chr10_seg100000_segment4 as recorded/done in computeAcc: 0.381926 , based on nr of corr 38193, tot cnt 100000\n","Acc at R64_chr10_seg100000_segment4 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr10/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr10_seg100000_segment5_avgRevCompl0_win1_step1\n","5 (100000,)\n","100000\n","100000\n","Acc at R64_chr10_seg100000_segment5: 0.384726 based on corr 38473 and tot 100000\n","Acc at R64_chr10_seg100000_segment5 as recorded/done in computeAcc: 0.384726 , based on nr of corr 38473, tot cnt 100000\n","Acc at R64_chr10_seg100000_segment5 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr10/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr10_seg100000_segment6_avgRevCompl0_win1_step1\n","6 (100000,)\n","100000\n","100000\n","Acc at R64_chr10_seg100000_segment6: 0.385846 based on corr 38585 and tot 100000\n","Acc at R64_chr10_seg100000_segment6 as recorded/done in computeAcc: 0.385846 , based on nr of corr 38585, tot cnt 100000\n","Acc at R64_chr10_seg100000_segment6 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr10/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr10_seg100000_segment7_avgRevCompl0_win1_step1\n","/content/drive/MyDrive/DNA_proj/data/S288C_reference_genome_R64-1-1_20110203/R64_chr11_annotationArray_simpleRepeats\n","Annotation file /content/drive/MyDrive/DNA_proj/data/S288C_reference_genome_R64-1-1_20110203/R64_chr11_annotationArray_simpleRepeats not found\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr11/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr11_seg100000_segment0_avgRevCompl0_win1_step1\n","0 (99950,)\n","99950\n","99950\n","Acc at R64_chr11_seg100000_segment0: 0.376304 based on corr 37612 and tot 99950\n","Acc at R64_chr11_seg100000_segment0 as recorded/done in computeAcc: 0.376304 , based on nr of corr 37612, tot cnt 99950\n","Acc at R64_chr11_seg100000_segment0 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr11/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr11_seg100000_segment1_avgRevCompl0_win1_step1\n","1 (100000,)\n","100000\n","100000\n","Acc at R64_chr11_seg100000_segment1: 0.389886 based on corr 38989 and tot 100000\n","Acc at R64_chr11_seg100000_segment1 as recorded/done in computeAcc: 0.389886 , based on nr of corr 38989, tot cnt 100000\n","Acc at R64_chr11_seg100000_segment1 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr11/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr11_seg100000_segment2_avgRevCompl0_win1_step1\n","2 (100000,)\n","100000\n","100000\n","Acc at R64_chr11_seg100000_segment2: 0.389616 based on corr 38962 and tot 100000\n","Acc at R64_chr11_seg100000_segment2 as recorded/done in computeAcc: 0.389616 , based on nr of corr 38962, tot cnt 100000\n","Acc at R64_chr11_seg100000_segment2 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr11/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr11_seg100000_segment3_avgRevCompl0_win1_step1\n","3 (100000,)\n","100000\n","100000\n","Acc at R64_chr11_seg100000_segment3: 0.388416 based on corr 38842 and tot 100000\n","Acc at R64_chr11_seg100000_segment3 as recorded/done in computeAcc: 0.388416 , based on nr of corr 38842, tot cnt 100000\n","Acc at R64_chr11_seg100000_segment3 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr11/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr11_seg100000_segment4_avgRevCompl0_win1_step1\n","4 (100000,)\n","100000\n","100000\n","Acc at R64_chr11_seg100000_segment4: 0.389526 based on corr 38953 and tot 100000\n","Acc at R64_chr11_seg100000_segment4 as recorded/done in computeAcc: 0.389526 , based on nr of corr 38953, tot cnt 100000\n","Acc at R64_chr11_seg100000_segment4 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr11/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr11_seg100000_segment5_avgRevCompl0_win1_step1\n","5 (100000,)\n","100000\n","100000\n","Acc at R64_chr11_seg100000_segment5: 0.386916 based on corr 38692 and tot 100000\n","Acc at R64_chr11_seg100000_segment5 as recorded/done in computeAcc: 0.386916 , based on nr of corr 38692, tot cnt 100000\n","Acc at R64_chr11_seg100000_segment5 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr11/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr11_seg100000_segment6_avgRevCompl0_win1_step1\n","/content/drive/MyDrive/DNA_proj/data/S288C_reference_genome_R64-1-1_20110203/R64_chr12_annotationArray_simpleRepeats\n","Annotation file /content/drive/MyDrive/DNA_proj/data/S288C_reference_genome_R64-1-1_20110203/R64_chr12_annotationArray_simpleRepeats not found\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr12_seg100000_segment0_avgRevCompl0_win1_step1\n","0 (99950,)\n","99950\n","99950\n","Acc at R64_chr12_seg100000_segment0: 0.378676 based on corr 37849 and tot 99950\n","Acc at R64_chr12_seg100000_segment0 as recorded/done in computeAcc: 0.378676 , based on nr of corr 37849, tot cnt 99950\n","Acc at R64_chr12_seg100000_segment0 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr12_seg100000_segment1_avgRevCompl0_win1_step1\n","1 (100000,)\n","100000\n","100000\n","Acc at R64_chr12_seg100000_segment1: 0.386606 based on corr 38661 and tot 100000\n","Acc at R64_chr12_seg100000_segment1 as recorded/done in computeAcc: 0.386606 , based on nr of corr 38661, tot cnt 100000\n","Acc at R64_chr12_seg100000_segment1 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr12_seg100000_segment2_avgRevCompl0_win1_step1\n","2 (100000,)\n","100000\n","100000\n","Acc at R64_chr12_seg100000_segment2: 0.387226 based on corr 38723 and tot 100000\n","Acc at R64_chr12_seg100000_segment2 as recorded/done in computeAcc: 0.387226 , based on nr of corr 38723, tot cnt 100000\n","Acc at R64_chr12_seg100000_segment2 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr12_seg100000_segment3_avgRevCompl0_win1_step1\n","3 (100000,)\n","100000\n","100000\n","Acc at R64_chr12_seg100000_segment3: 0.382406 based on corr 38241 and tot 100000\n","Acc at R64_chr12_seg100000_segment3 as recorded/done in computeAcc: 0.382406 , based on nr of corr 38241, tot cnt 100000\n","Acc at R64_chr12_seg100000_segment3 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr12_seg100000_segment4_avgRevCompl0_win1_step1\n","4 (100000,)\n","100000\n","100000\n","Acc at R64_chr12_seg100000_segment4: 0.370296 based on corr 37030 and tot 100000\n","Acc at R64_chr12_seg100000_segment4 as recorded/done in computeAcc: 0.370296 , based on nr of corr 37030, tot cnt 100000\n","Acc at R64_chr12_seg100000_segment4 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr12_seg100000_segment5_avgRevCompl0_win1_step1\n","5 (100000,)\n","100000\n","100000\n","Acc at R64_chr12_seg100000_segment5: 0.390636 based on corr 39064 and tot 100000\n","Acc at R64_chr12_seg100000_segment5 as recorded/done in computeAcc: 0.390636 , based on nr of corr 39064, tot cnt 100000\n","Acc at R64_chr12_seg100000_segment5 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr12_seg100000_segment6_avgRevCompl0_win1_step1\n","6 (100000,)\n","100000\n","100000\n","Acc at R64_chr12_seg100000_segment6: 0.388536 based on corr 38854 and tot 100000\n","Acc at R64_chr12_seg100000_segment6 as recorded/done in computeAcc: 0.388536 , based on nr of corr 38854, tot cnt 100000\n","Acc at R64_chr12_seg100000_segment6 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr12_seg100000_segment7_avgRevCompl0_win1_step1\n","7 (100000,)\n","100000\n","100000\n","Acc at R64_chr12_seg100000_segment7: 0.390506 based on corr 39051 and tot 100000\n","Acc at R64_chr12_seg100000_segment7 as recorded/done in computeAcc: 0.390506 , based on nr of corr 39051, tot cnt 100000\n","Acc at R64_chr12_seg100000_segment7 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr12_seg100000_segment8_avgRevCompl0_win1_step1\n","8 (100000,)\n","100000\n","100000\n","Acc at R64_chr12_seg100000_segment8: 0.384346 based on corr 38435 and tot 100000\n","Acc at R64_chr12_seg100000_segment8 as recorded/done in computeAcc: 0.384346 , based on nr of corr 38435, tot cnt 100000\n","Acc at R64_chr12_seg100000_segment8 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr12_seg100000_segment9_avgRevCompl0_win1_step1\n","9 (100000,)\n","100000\n","100000\n","Acc at R64_chr12_seg100000_segment9: 0.379616 based on corr 37962 and tot 100000\n","Acc at R64_chr12_seg100000_segment9 as recorded/done in computeAcc: 0.379616 , based on nr of corr 37962, tot cnt 100000\n","Acc at R64_chr12_seg100000_segment9 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr12_seg100000_segment10_avgRevCompl0_win1_step1\n","/content/drive/MyDrive/DNA_proj/data/S288C_reference_genome_R64-1-1_20110203/R64_chr13_annotationArray_simpleRepeats\n","Annotation file /content/drive/MyDrive/DNA_proj/data/S288C_reference_genome_R64-1-1_20110203/R64_chr13_annotationArray_simpleRepeats not found\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr13_seg100000_segment0_avgRevCompl0_win1_step1\n","0 (99950,)\n","99950\n","99950\n","Acc at R64_chr13_seg100000_segment0: 0.387960 based on corr 38777 and tot 99950\n","Acc at R64_chr13_seg100000_segment0 as recorded/done in computeAcc: 0.387960 , based on nr of corr 38777, tot cnt 99950\n","Acc at R64_chr13_seg100000_segment0 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr13_seg100000_segment1_avgRevCompl0_win1_step1\n","1 (100000,)\n","100000\n","100000\n","Acc at R64_chr13_seg100000_segment1: 0.385966 based on corr 38597 and tot 100000\n","Acc at R64_chr13_seg100000_segment1 as recorded/done in computeAcc: 0.385966 , based on nr of corr 38597, tot cnt 100000\n","Acc at R64_chr13_seg100000_segment1 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr13_seg100000_segment2_avgRevCompl0_win1_step1\n","2 (100000,)\n","100000\n","100000\n","Acc at R64_chr13_seg100000_segment2: 0.388886 based on corr 38889 and tot 100000\n","Acc at R64_chr13_seg100000_segment2 as recorded/done in computeAcc: 0.388886 , based on nr of corr 38889, tot cnt 100000\n","Acc at R64_chr13_seg100000_segment2 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr13_seg100000_segment3_avgRevCompl0_win1_step1\n","3 (100000,)\n","100000\n","100000\n","Acc at R64_chr13_seg100000_segment3: 0.384366 based on corr 38437 and tot 100000\n","Acc at R64_chr13_seg100000_segment3 as recorded/done in computeAcc: 0.384366 , based on nr of corr 38437, tot cnt 100000\n","Acc at R64_chr13_seg100000_segment3 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr13_seg100000_segment4_avgRevCompl0_win1_step1\n","4 (100000,)\n","100000\n","100000\n","Acc at R64_chr13_seg100000_segment4: 0.386646 based on corr 38665 and tot 100000\n","Acc at R64_chr13_seg100000_segment4 as recorded/done in computeAcc: 0.386646 , based on nr of corr 38665, tot cnt 100000\n","Acc at R64_chr13_seg100000_segment4 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr13_seg100000_segment5_avgRevCompl0_win1_step1\n","5 (100000,)\n","100000\n","100000\n","Acc at R64_chr13_seg100000_segment5: 0.388676 based on corr 38868 and tot 100000\n","Acc at R64_chr13_seg100000_segment5 as recorded/done in computeAcc: 0.388676 , based on nr of corr 38868, tot cnt 100000\n","Acc at R64_chr13_seg100000_segment5 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr13_seg100000_segment6_avgRevCompl0_win1_step1\n","6 (100000,)\n","100000\n","100000\n","Acc at R64_chr13_seg100000_segment6: 0.388096 based on corr 38810 and tot 100000\n","Acc at R64_chr13_seg100000_segment6 as recorded/done in computeAcc: 0.388096 , based on nr of corr 38810, tot cnt 100000\n","Acc at R64_chr13_seg100000_segment6 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr13_seg100000_segment7_avgRevCompl0_win1_step1\n","7 (100000,)\n","100000\n","100000\n","Acc at R64_chr13_seg100000_segment7: 0.388916 based on corr 38892 and tot 100000\n","Acc at R64_chr13_seg100000_segment7 as recorded/done in computeAcc: 0.388916 , based on nr of corr 38892, tot cnt 100000\n","Acc at R64_chr13_seg100000_segment7 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr13_seg100000_segment8_avgRevCompl0_win1_step1\n","8 (100000,)\n","100000\n","100000\n","Acc at R64_chr13_seg100000_segment8: 0.384466 based on corr 38447 and tot 100000\n","Acc at R64_chr13_seg100000_segment8 as recorded/done in computeAcc: 0.384466 , based on nr of corr 38447, tot cnt 100000\n","Acc at R64_chr13_seg100000_segment8 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr13_seg100000_segment9_avgRevCompl0_win1_step1\n","/content/drive/MyDrive/DNA_proj/data/S288C_reference_genome_R64-1-1_20110203/R64_chr14_annotationArray_simpleRepeats\n","Annotation file /content/drive/MyDrive/DNA_proj/data/S288C_reference_genome_R64-1-1_20110203/R64_chr14_annotationArray_simpleRepeats not found\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr14/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr14_seg100000_segment0_avgRevCompl0_win1_step1\n","0 (99950,)\n","99950\n","99950\n","Acc at R64_chr14_seg100000_segment0: 0.373783 based on corr 37360 and tot 99950\n","Acc at R64_chr14_seg100000_segment0 as recorded/done in computeAcc: 0.373783 , based on nr of corr 37360, tot cnt 99950\n","Acc at R64_chr14_seg100000_segment0 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr14/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr14_seg100000_segment1_avgRevCompl0_win1_step1\n","1 (100000,)\n","100000\n","100000\n","Acc at R64_chr14_seg100000_segment1: 0.384036 based on corr 38404 and tot 100000\n","Acc at R64_chr14_seg100000_segment1 as recorded/done in computeAcc: 0.384036 , based on nr of corr 38404, tot cnt 100000\n","Acc at R64_chr14_seg100000_segment1 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr14/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr14_seg100000_segment2_avgRevCompl0_win1_step1\n","2 (100000,)\n","100000\n","100000\n","Acc at R64_chr14_seg100000_segment2: 0.393726 based on corr 39373 and tot 100000\n","Acc at R64_chr14_seg100000_segment2 as recorded/done in computeAcc: 0.393726 , based on nr of corr 39373, tot cnt 100000\n","Acc at R64_chr14_seg100000_segment2 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr14/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr14_seg100000_segment3_avgRevCompl0_win1_step1\n","3 (100000,)\n","100000\n","100000\n","Acc at R64_chr14_seg100000_segment3: 0.385056 based on corr 38506 and tot 100000\n","Acc at R64_chr14_seg100000_segment3 as recorded/done in computeAcc: 0.385056 , based on nr of corr 38506, tot cnt 100000\n","Acc at R64_chr14_seg100000_segment3 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr14/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr14_seg100000_segment4_avgRevCompl0_win1_step1\n","4 (100000,)\n","100000\n","100000\n","Acc at R64_chr14_seg100000_segment4: 0.386556 based on corr 38656 and tot 100000\n","Acc at R64_chr14_seg100000_segment4 as recorded/done in computeAcc: 0.386556 , based on nr of corr 38656, tot cnt 100000\n","Acc at R64_chr14_seg100000_segment4 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr14/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr14_seg100000_segment5_avgRevCompl0_win1_step1\n","5 (100000,)\n","100000\n","100000\n","Acc at R64_chr14_seg100000_segment5: 0.381796 based on corr 38180 and tot 100000\n","Acc at R64_chr14_seg100000_segment5 as recorded/done in computeAcc: 0.381796 , based on nr of corr 38180, tot cnt 100000\n","Acc at R64_chr14_seg100000_segment5 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr14/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr14_seg100000_segment6_avgRevCompl0_win1_step1\n","6 (100000,)\n","100000\n","100000\n","Acc at R64_chr14_seg100000_segment6: 0.381286 based on corr 38129 and tot 100000\n","Acc at R64_chr14_seg100000_segment6 as recorded/done in computeAcc: 0.381286 , based on nr of corr 38129, tot cnt 100000\n","Acc at R64_chr14_seg100000_segment6 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr14/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr14_seg100000_segment7_avgRevCompl0_win1_step1\n","/content/drive/MyDrive/DNA_proj/data/S288C_reference_genome_R64-1-1_20110203/R64_chr15_annotationArray_simpleRepeats\n","Annotation file /content/drive/MyDrive/DNA_proj/data/S288C_reference_genome_R64-1-1_20110203/R64_chr15_annotationArray_simpleRepeats not found\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr15_seg100000_segment0_avgRevCompl0_win1_step1\n","0 (99950,)\n","99950\n","99950\n","Acc at R64_chr15_seg100000_segment0: 0.375964 based on corr 37578 and tot 99950\n","Acc at R64_chr15_seg100000_segment0 as recorded/done in computeAcc: 0.375964 , based on nr of corr 37578, tot cnt 99950\n","Acc at R64_chr15_seg100000_segment0 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr15_seg100000_segment1_avgRevCompl0_win1_step1\n","1 (100000,)\n","100000\n","100000\n","Acc at R64_chr15_seg100000_segment1: 0.385546 based on corr 38555 and tot 100000\n","Acc at R64_chr15_seg100000_segment1 as recorded/done in computeAcc: 0.385546 , based on nr of corr 38555, tot cnt 100000\n","Acc at R64_chr15_seg100000_segment1 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr15_seg100000_segment2_avgRevCompl0_win1_step1\n","2 (100000,)\n","100000\n","100000\n","Acc at R64_chr15_seg100000_segment2: 0.384116 based on corr 38412 and tot 100000\n","Acc at R64_chr15_seg100000_segment2 as recorded/done in computeAcc: 0.384116 , based on nr of corr 38412, tot cnt 100000\n","Acc at R64_chr15_seg100000_segment2 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr15_seg100000_segment3_avgRevCompl0_win1_step1\n","3 (100000,)\n","100000\n","100000\n","Acc at R64_chr15_seg100000_segment3: 0.383196 based on corr 38320 and tot 100000\n","Acc at R64_chr15_seg100000_segment3 as recorded/done in computeAcc: 0.383196 , based on nr of corr 38320, tot cnt 100000\n","Acc at R64_chr15_seg100000_segment3 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr15_seg100000_segment4_avgRevCompl0_win1_step1\n","4 (100000,)\n","100000\n","100000\n","Acc at R64_chr15_seg100000_segment4: 0.385536 based on corr 38554 and tot 100000\n","Acc at R64_chr15_seg100000_segment4 as recorded/done in computeAcc: 0.385536 , based on nr of corr 38554, tot cnt 100000\n","Acc at R64_chr15_seg100000_segment4 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr15_seg100000_segment5_avgRevCompl0_win1_step1\n","5 (100000,)\n","100000\n","100000\n","Acc at R64_chr15_seg100000_segment5: 0.386586 based on corr 38659 and tot 100000\n","Acc at R64_chr15_seg100000_segment5 as recorded/done in computeAcc: 0.386586 , based on nr of corr 38659, tot cnt 100000\n","Acc at R64_chr15_seg100000_segment5 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr15_seg100000_segment6_avgRevCompl0_win1_step1\n","6 (100000,)\n","100000\n","100000\n","Acc at R64_chr15_seg100000_segment6: 0.386956 based on corr 38696 and tot 100000\n","Acc at R64_chr15_seg100000_segment6 as recorded/done in computeAcc: 0.386956 , based on nr of corr 38696, tot cnt 100000\n","Acc at R64_chr15_seg100000_segment6 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr15_seg100000_segment7_avgRevCompl0_win1_step1\n","7 (100000,)\n","100000\n","100000\n","Acc at R64_chr15_seg100000_segment7: 0.390046 based on corr 39005 and tot 100000\n","Acc at R64_chr15_seg100000_segment7 as recorded/done in computeAcc: 0.390046 , based on nr of corr 39005, tot cnt 100000\n","Acc at R64_chr15_seg100000_segment7 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr15_seg100000_segment8_avgRevCompl0_win1_step1\n","8 (100000,)\n","100000\n","100000\n","Acc at R64_chr15_seg100000_segment8: 0.385636 based on corr 38564 and tot 100000\n","Acc at R64_chr15_seg100000_segment8 as recorded/done in computeAcc: 0.385636 , based on nr of corr 38564, tot cnt 100000\n","Acc at R64_chr15_seg100000_segment8 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr15_seg100000_segment9_avgRevCompl0_win1_step1\n","9 (100000,)\n","100000\n","100000\n","Acc at R64_chr15_seg100000_segment9: 0.382356 based on corr 38236 and tot 100000\n","Acc at R64_chr15_seg100000_segment9 as recorded/done in computeAcc: 0.382356 , based on nr of corr 38236, tot cnt 100000\n","Acc at R64_chr15_seg100000_segment9 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr15_seg100000_segment10_avgRevCompl0_win1_step1\n","/content/drive/MyDrive/DNA_proj/data/S288C_reference_genome_R64-1-1_20110203/R64_chr16_annotationArray_simpleRepeats\n","Annotation file /content/drive/MyDrive/DNA_proj/data/S288C_reference_genome_R64-1-1_20110203/R64_chr16_annotationArray_simpleRepeats not found\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr16_seg100000_segment0_avgRevCompl0_win1_step1\n","0 (99950,)\n","99950\n","99950\n","Acc at R64_chr16_seg100000_segment0: 0.378515 based on corr 37833 and tot 99950\n","Acc at R64_chr16_seg100000_segment0 as recorded/done in computeAcc: 0.378515 , based on nr of corr 37833, tot cnt 99950\n","Acc at R64_chr16_seg100000_segment0 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr16_seg100000_segment1_avgRevCompl0_win1_step1\n","1 (100000,)\n","100000\n","100000\n","Acc at R64_chr16_seg100000_segment1: 0.398816 based on corr 39882 and tot 100000\n","Acc at R64_chr16_seg100000_segment1 as recorded/done in computeAcc: 0.398816 , based on nr of corr 39882, tot cnt 100000\n","Acc at R64_chr16_seg100000_segment1 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr16_seg100000_segment2_avgRevCompl0_win1_step1\n","2 (100000,)\n","100000\n","100000\n","Acc at R64_chr16_seg100000_segment2: 0.380866 based on corr 38087 and tot 100000\n","Acc at R64_chr16_seg100000_segment2 as recorded/done in computeAcc: 0.380866 , based on nr of corr 38087, tot cnt 100000\n","Acc at R64_chr16_seg100000_segment2 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr16_seg100000_segment3_avgRevCompl0_win1_step1\n","3 (100000,)\n","100000\n","100000\n","Acc at R64_chr16_seg100000_segment3: 0.389606 based on corr 38961 and tot 100000\n","Acc at R64_chr16_seg100000_segment3 as recorded/done in computeAcc: 0.389606 , based on nr of corr 38961, tot cnt 100000\n","Acc at R64_chr16_seg100000_segment3 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr16_seg100000_segment4_avgRevCompl0_win1_step1\n","4 (100000,)\n","100000\n","100000\n","Acc at R64_chr16_seg100000_segment4: 0.388536 based on corr 38854 and tot 100000\n","Acc at R64_chr16_seg100000_segment4 as recorded/done in computeAcc: 0.388536 , based on nr of corr 38854, tot cnt 100000\n","Acc at R64_chr16_seg100000_segment4 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr16_seg100000_segment5_avgRevCompl0_win1_step1\n","5 (100000,)\n","100000\n","100000\n","Acc at R64_chr16_seg100000_segment5: 0.382736 based on corr 38274 and tot 100000\n","Acc at R64_chr16_seg100000_segment5 as recorded/done in computeAcc: 0.382736 , based on nr of corr 38274, tot cnt 100000\n","Acc at R64_chr16_seg100000_segment5 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr16_seg100000_segment6_avgRevCompl0_win1_step1\n","6 (100000,)\n","100000\n","100000\n","Acc at R64_chr16_seg100000_segment6: 0.383846 based on corr 38385 and tot 100000\n","Acc at R64_chr16_seg100000_segment6 as recorded/done in computeAcc: 0.383846 , based on nr of corr 38385, tot cnt 100000\n","Acc at R64_chr16_seg100000_segment6 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr16_seg100000_segment7_avgRevCompl0_win1_step1\n","7 (100000,)\n","100000\n","100000\n","Acc at R64_chr16_seg100000_segment7: 0.383336 based on corr 38334 and tot 100000\n","Acc at R64_chr16_seg100000_segment7 as recorded/done in computeAcc: 0.383336 , based on nr of corr 38334, tot cnt 100000\n","Acc at R64_chr16_seg100000_segment7 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr16_seg100000_segment8_avgRevCompl0_win1_step1\n","8 (100000,)\n","100000\n","100000\n","Acc at R64_chr16_seg100000_segment8: 0.384976 based on corr 38498 and tot 100000\n","Acc at R64_chr16_seg100000_segment8 as recorded/done in computeAcc: 0.384976 , based on nr of corr 38498, tot cnt 100000\n","Acc at R64_chr16_seg100000_segment8 for repeats as recorded/done in computeAcc: 0.000000 , based on nr of corr 0, tot cnt 0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predReturn_R64_chr16_seg100000_segment9_avgRevCompl0_win1_step1\n"]}]},{"cell_type":"markdown","source":["We can then aggregate the results per chromosome and get the results in a table (in .txt and Latex format too) and in a plot:"],"metadata":{"id":"MU-eLyFfGVTK"}},{"cell_type":"code","execution_count":34,"metadata":{"id":"4Os7r5S-qk1B","executionInfo":{"status":"ok","timestamp":1650138353613,"user_tz":-120,"elapsed":35,"user":{"displayName":"Christian Grønbæk","userId":"14909394381660051993"}}},"outputs":[],"source":["#Do the aggregation so that it covers all chromosomes/anotations for which results are had:\n","dictionaryName = 'accuracyChromoByAnnoDictionary'\n","stats.calculateAggrAccOnChromos(rootOutput =rootOutput , chromosomeDict= chromosomeDict,  dictionaryName = dictionaryName)\n"]},{"cell_type":"code","execution_count":35,"metadata":{"id":"lyn-vGZOqk1C","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1650138353613,"user_tz":-120,"elapsed":33,"user":{"displayName":"Christian Grønbæk","userId":"14909394381660051993"}},"outputId":"837fddaf-7ed1-4468-85e3-2e1d6892d13f"},"outputs":[{"output_type":"execute_result","data":{"text/plain":["('\\\\begin{table}[h!]\\n  \\\\begin{center}\\n    \\\\label{tab:table1c}\\n    \\\\begin{tabular}{l | c | c | r} % <-- Alignments: 1st column left, 2nd middle and 3rd right, with vertical lines in between\\n      \\\\textbf{chr/annotation}& \\\\textbf{all}\\\\\\\\\\n      \\\\hline\\nR64_chr1 & 0.3852\\\\\\\\\\nR64_chr2 & 0.386\\\\\\\\\\nR64_chr3 & 0.3834\\\\\\\\\\nR64_chr4 & 0.3884\\\\\\\\\\nR64_chr5 & 0.383\\\\\\\\\\nR64_chr6 & 0.3842\\\\\\\\\\nR64_chr7 & 0.387\\\\\\\\\\nR64_chr8 & 0.3828\\\\\\\\\\nR64_chr9 & 0.3841\\\\\\\\\\nR64_chr10 & 0.3843\\\\\\\\\\nR64_chr11 & 0.3868\\\\\\\\\\nR64_chr12 & 0.3839\\\\\\\\\\nR64_chr13 & 0.3871\\\\\\\\\\nR64_chr14 & 0.3838\\\\\\\\\\nR64_chr15 & 0.3846\\\\\\\\\\nR64_chr16 & 0.3857\\\\\\\\\\nAll & 0.3855\\\\\\\\\\n    \\\\end{tabular}\\n        \\\\caption{LSTM50 on R64}\\n  \\\\end{center}\\n \\\\end{table}',\n"," ['all'])"]},"metadata":{},"execution_count":35}],"source":["#Yeast, w. train test split\n","rootResults = rootOutput\n","captionText = 'LSTM50 on R64'\n","\n","import pickle\n","\n","#Results per chromo/anno\n","loadFile = rootResults + 'accuracyByChromoAnnoDictionary' \n","resultsDictChromo = pickle.load(open( loadFile, \"rb\"))\n","\n","#Aggr over all chromos\n","loadFile = rootResults + 'accuracyByAnnoDictionary' \n","resultsDict = pickle.load(open( loadFile, \"rb\"))\n","\n","fileName = 'table_chromo_anno.txt'\n","rowNames = chromosomeOrderList\n","colNames = ['all']   #the models will appear in columns in this order\n","#with final aggregation row:\n","stats.makeTexTable(inputDict = resultsDictChromo , rowColHeading = 'chr/annotation', rowNames = rowNames,\n","                                  columnNames = colNames, inputDict2 = resultsDict, captionText = captionText, rootOutput = rootOutput, fileName = fileName )\n"]},{"cell_type":"code","source":["chromosomeOrderList = chromosomeOrderList[::-1]\n","saveAtDpi = 300\n","addAvg_b = 1\n","avgLevel = 0.3751\n","resultsDictByChromoModel = stats.collectAccuracyChromosomesSeveralModels(rootOutput = rootOutput, \n","                                  rootPredictModelList = [rootOutput], \n","                                  modelFileList = [modelFileNameNN],\n","                                  modelNameList = ['LSTM4'],\n","                                  chromosomeOrderList = chromosomeOrderList,\n","                                  plot_b = 1,\n","                                  addAvg_b = addAvg_b,\n","                                  avgLevel = avgLevel,\n","                                  saveAtDpi = saveAtDpi)\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"YWu4B7F7un6n","executionInfo":{"status":"ok","timestamp":1650138354205,"user_tz":-120,"elapsed":600,"user":{"displayName":"Christian Grønbæk","userId":"14909394381660051993"}},"outputId":"64417524-0eed-4da7-e5af-1f660efb079d"},"execution_count":36,"outputs":[{"output_type":"stream","name":"stdout","text":["{'R64_chr1': {'all': [0.3851912978244561, 77019.0, 199950]}, 'R64_chr2': {'all': [0.38597787361710106, 308763.0, 799950]}, 'R64_chr3': {'all': [0.3834405734289048, 115013.0, 299950]}, 'R64_chr4': {'all': [0.38844094803160106, 582642.0, 1499950]}, 'R64_chr5': {'all': [0.38303230323032306, 191497.0, 499950]}, 'R64_chr6': {'all': [0.3841810452613153, 76817.0, 199950]}, 'R64_chr7': {'all': [0.38700535026751337, 386986.0, 999950]}, 'R64_chr8': {'all': [0.38275827582758276, 191360.0, 499950]}, 'R64_chr9': {'all': [0.3841455181897737, 153639.0, 399950]}, 'R64_chr10': {'all': [0.3843303093078077, 269012.0, 699950]}, 'R64_chr11': {'all': [0.38678223185265437, 232050.0, 599950]}, 'R64_chr12': {'all': [0.38388919445972297, 383870.0, 999950]}, 'R64_chr13': {'all': [0.38711261736763153, 348382.0, 899950]}, 'R64_chr14': {'all': [0.38375312522323024, 268608.0, 699950]}, 'R64_chr15': {'all': [0.38459822991149556, 384579.0, 999950]}, 'R64_chr16': {'all': [0.3856969831657314, 347108.0, 899950]}}\n","nChromos  16\n"]}]},{"cell_type":"markdown","source":["Running the Fourier transforms on the model's probability arrays (prob of reference base at each position) is done in a single call to a wrapper. First some settings, and then the call (it runs surprisingly fast):  "],"metadata":{"id":"egXVi1FfzXP0"}},{"cell_type":"code","execution_count":18,"metadata":{"id":"XWeGjT21qk1C","executionInfo":{"status":"ok","timestamp":1650141701950,"user_tz":-120,"elapsed":282,"user":{"displayName":"Christian Grønbæk","userId":"14909394381660051993"}}},"outputs":[],"source":["rootOutput =  r\"/content/drive/MyDrive/DNA_proj/results/predictions/\"\n","\n","#We reuse some of the settings from above\n","modelFileNameNN = modelFileNameNN\n","chromosomeOrderList = chromosomeOrderList\n","chromosomeDict = chromosomeDict \n","\n","#placeholders\n","modelFileName_forATorGCbias = ''\n","rootOutput_forATorGCbias = ''\n","rootInput_forATorGCbias = ''\n","forATorGCbias_b = 0 #!\n","\n","#General settings:\n","segmentLength = 100000\n","\n","augmentWithRevComplementary_b = 0  #!\n","\n","#window lgth and stepsize used in generating the avg prediction\n","windowLength = 1\n","stepSize = 1\n","                      \n","#Param's for Fourier plots:\n","shuffle_b = 0 #!\n","randomizeDisqualified_b = 0 #!\n","randomizingByShuffle_b = 0 #!\n","fullGC_b = 0 #!\n","\n","#Which input to use:\n","inputArrayType = 1 # 1: ref base prob's; 0: pred returns\n","plotOnlyNorm_b = 1 #default:1 \n","\n","#Here we only plot the 20-2000 range; extend the list with more starts/stops to cover several ranges\n","fourierStep = 10\n","fourierRawPlotFrq = 5\n","fourierStartList = [20]\n","fourierStopList = [2000]\n","fourierWindowLengthList = [100]\n","\n","ratioQcutoff = 0.9 #0.7 w hg18\n","dumpFourier_b = 0\n","dumpFileNamePrefix = ''"]},{"cell_type":"code","source":["#Call to Fourier-wrapper:\n","for i in range(len(fourierStartList)):\n","    fourierStart = fourierStartList[i]\n","    fourierStop = fourierStopList[i]\n","    fourierWindowLength = fourierWindowLengthList[i]\n","    stats.computeFourierChromosomes(chromosomeOrderList = chromosomeOrderList,\n","                                    rootOutput = rootOutput,\n","                                    modelFileName = modelFileNameNN,  \n","                                    segmentLength = segmentLength,\n","                                    inputArrayType = inputArrayType,\n","                                    averageRevComplementary_b = augmentWithRevComplementary_b,\n","                                    ratioQcutoff = ratioQcutoff,\n","                                    windowLength = windowLength,\n","                                    stepSize = stepSize,\n","                                    plotOnlyNorm_b = plotOnlyNorm_b,\n","                                    fourierWindowLength = fourierWindowLength,\n","                                    fourierStart = fourierStart,\n","                                    fourierStop = fourierStop,\n","                                    fourierStep = fourierStep, \n","                                    fourierRawPlotFrq = fourierRawPlotFrq,\n","                                    shuffle_b = shuffle_b,\n","                                    randomizeDisqualified_b =randomizeDisqualified_b,\n","                                    randomizingByShuffle_b = randomizingByShuffle_b,\n","                                    forATorGCbias_b = forATorGCbias_b, \n","                                    rootOutput_forATorGCbias= rootOutput_forATorGCbias,\n","                                    rootInput_forATorGCbias = rootInput_forATorGCbias,\n","                                    fullGC_b = fullGC_b,\n","                                    dumpFourier_b = dumpFourier_b,\n","                                    dumpFileNamePrefix = dumpFileNamePrefix, \n","                                    modelFileName_forATorGCbias = modelFileName_forATorGCbias)\n"],"metadata":{"id":"vjs0uCNn450l","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1650142016993,"user_tz":-120,"elapsed":149690,"user":{"displayName":"Christian Grønbæk","userId":"14909394381660051993"}},"outputId":"814d6fda-08e5-4f66-dd85-bbce83fb5ee8"},"execution_count":21,"outputs":[{"output_type":"stream","name":"stdout","text":["/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr1/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr1_seg100000_segment0_avgRevCompl0\n","99950\n","avgPredArray size:  99950\n","ratioQ:  1.0\n","Fourier transf lenght:  99950\n","frqs  [ 0.00000000e+00  1.00050025e-05  2.00100050e-05 ... -3.00150075e-05\n"," -2.00100050e-05 -1.00050025e-05]\n","First 10 (positive) frqs:  [0.00000000e+00 1.00050025e-05 2.00100050e-05 3.00150075e-05\n"," 4.00200100e-05 5.00250125e-05 6.00300150e-05 7.00350175e-05\n"," 8.00400200e-05 9.00450225e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.00050025e-05 -2.00100050e-05 -3.00150075e-05 -4.00200100e-05\n"," -5.00250125e-05 -6.00300150e-05 -7.00350175e-05 -8.00400200e-05\n"," -9.00450225e-05 -1.00050025e-04]\n","Up to stop (2000) with frq 0.020010, the max (min) modulus of Fourier coeff is: 29426.839650 (1.049864)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 210.230853 (0.187132)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr1/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr1_seg100000_segment1_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29318.820430 (0.360726)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 236.996151 (0.039374)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr1/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr1_seg100000_segment2_avgRevCompl0\n","Directory /content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/FourierOnRefBaseProb/20to2000/ created. Output will be placed there.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr2_seg100000_segment0_avgRevCompl0\n","99950\n","avgPredArray size:  99950\n","ratioQ:  1.0\n","Fourier transf lenght:  99950\n","frqs  [ 0.00000000e+00  1.00050025e-05  2.00100050e-05 ... -3.00150075e-05\n"," -2.00100050e-05 -1.00050025e-05]\n","First 10 (positive) frqs:  [0.00000000e+00 1.00050025e-05 2.00100050e-05 3.00150075e-05\n"," 4.00200100e-05 5.00250125e-05 6.00300150e-05 7.00350175e-05\n"," 8.00400200e-05 9.00450225e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.00050025e-05 -2.00100050e-05 -3.00150075e-05 -4.00200100e-05\n"," -5.00250125e-05 -6.00300150e-05 -7.00350175e-05 -8.00400200e-05\n"," -9.00450225e-05 -1.00050025e-04]\n","Up to stop (2000) with frq 0.020010, the max (min) modulus of Fourier coeff is: 29238.289271 (1.090319)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 191.145671 (0.227573)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr2_seg100000_segment1_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29530.853347 (1.948767)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 197.694144 (0.069005)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr2_seg100000_segment2_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29073.469150 (2.761018)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 156.258643 (0.164225)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr2_seg100000_segment3_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29382.304674 (0.957062)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 139.777304 (0.169242)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr2_seg100000_segment4_avgRevCompl0\n","100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7006: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7007: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29697.831064 (1.643515)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 209.900434 (0.119637)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr2_seg100000_segment5_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29271.045999 (0.305688)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 161.698544 (0.021876)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr2_seg100000_segment6_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29230.293078 (0.705434)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 141.075621 (0.061374)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr2_seg100000_segment7_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29160.384917 (0.375086)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 181.606134 (0.139173)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr2_seg100000_segment8_avgRevCompl0\n","Directory /content/drive/MyDrive/DNA_proj/results/predictions/R64_chr3/FourierOnRefBaseProb/20to2000/ created. Output will be placed there.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr3/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr3_seg100000_segment0_avgRevCompl0\n","99950\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6875: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6876: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6877: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1, ax1 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6878: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6879: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2, ax2 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6880: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["avgPredArray size:  99950\n","ratioQ:  1.0\n","Fourier transf lenght:  99950\n","frqs  [ 0.00000000e+00  1.00050025e-05  2.00100050e-05 ... -3.00150075e-05\n"," -2.00100050e-05 -1.00050025e-05]\n","First 10 (positive) frqs:  [0.00000000e+00 1.00050025e-05 2.00100050e-05 3.00150075e-05\n"," 4.00200100e-05 5.00250125e-05 6.00300150e-05 7.00350175e-05\n"," 8.00400200e-05 9.00450225e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.00050025e-05 -2.00100050e-05 -3.00150075e-05 -4.00200100e-05\n"," -5.00250125e-05 -6.00300150e-05 -7.00350175e-05 -8.00400200e-05\n"," -9.00450225e-05 -1.00050025e-04]\n","Up to stop (2000) with frq 0.020010, the max (min) modulus of Fourier coeff is: 29416.182640 (0.740562)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 166.824541 (0.195305)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr3/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr3_seg100000_segment1_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29090.580511 (0.892642)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 152.500331 (0.212665)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr3/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr3_seg100000_segment2_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29329.501045 (1.741137)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 216.171316 (0.182436)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr3/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr3_seg100000_segment3_avgRevCompl0\n","Directory /content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/FourierOnRefBaseProb/20to2000/ created. Output will be placed there.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr4_seg100000_segment0_avgRevCompl0\n","99950\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6875: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6876: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6877: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1, ax1 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6878: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6879: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2, ax2 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6880: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["avgPredArray size:  99950\n","ratioQ:  1.0\n","Fourier transf lenght:  99950\n","frqs  [ 0.00000000e+00  1.00050025e-05  2.00100050e-05 ... -3.00150075e-05\n"," -2.00100050e-05 -1.00050025e-05]\n","First 10 (positive) frqs:  [0.00000000e+00 1.00050025e-05 2.00100050e-05 3.00150075e-05\n"," 4.00200100e-05 5.00250125e-05 6.00300150e-05 7.00350175e-05\n"," 8.00400200e-05 9.00450225e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.00050025e-05 -2.00100050e-05 -3.00150075e-05 -4.00200100e-05\n"," -5.00250125e-05 -6.00300150e-05 -7.00350175e-05 -8.00400200e-05\n"," -9.00450225e-05 -1.00050025e-04]\n","Up to stop (2000) with frq 0.020010, the max (min) modulus of Fourier coeff is: 29052.058680 (0.986998)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 130.222999 (0.095254)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr4_seg100000_segment1_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29590.387126 (0.430683)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 186.933494 (0.131328)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr4_seg100000_segment2_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29358.292179 (1.585267)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 135.545171 (0.179864)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr4_seg100000_segment3_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29759.178336 (0.936540)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 254.464148 (0.234641)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr4_seg100000_segment4_avgRevCompl0\n","100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7004: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7005: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7006: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7007: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29404.292610 (1.600772)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 139.787868 (0.044269)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr4_seg100000_segment5_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29526.141414 (0.511913)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 175.181781 (0.140106)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr4_seg100000_segment6_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29313.866862 (2.285492)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 182.350272 (0.065309)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr4_seg100000_segment7_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29473.642520 (1.710377)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 239.700329 (0.079242)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr4_seg100000_segment8_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29545.120455 (1.293766)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 207.421840 (0.230303)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr4_seg100000_segment9_avgRevCompl0\n","100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7004: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7005: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7006: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7007: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29265.734605 (1.386578)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 154.602397 (0.131552)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr4_seg100000_segment10_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29221.385872 (1.999509)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 180.283397 (0.077504)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr4_seg100000_segment11_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29379.597887 (1.615393)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 233.639015 (0.235818)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr4_seg100000_segment12_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29426.759939 (1.330783)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 153.528469 (0.141222)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr4_seg100000_segment13_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29334.953486 (1.306488)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 183.021903 (0.066533)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr4_seg100000_segment14_avgRevCompl0\n","100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7004: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7005: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7006: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7007: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29396.891083 (0.993020)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 234.877937 (0.004686)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr4_seg100000_segment15_avgRevCompl0\n","Directory /content/drive/MyDrive/DNA_proj/results/predictions/R64_chr5/FourierOnRefBaseProb/20to2000/ created. Output will be placed there.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr5/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr5_seg100000_segment0_avgRevCompl0\n","99950\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6875: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6876: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6877: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1, ax1 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6878: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6879: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2, ax2 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6880: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["avgPredArray size:  99950\n","ratioQ:  1.0\n","Fourier transf lenght:  99950\n","frqs  [ 0.00000000e+00  1.00050025e-05  2.00100050e-05 ... -3.00150075e-05\n"," -2.00100050e-05 -1.00050025e-05]\n","First 10 (positive) frqs:  [0.00000000e+00 1.00050025e-05 2.00100050e-05 3.00150075e-05\n"," 4.00200100e-05 5.00250125e-05 6.00300150e-05 7.00350175e-05\n"," 8.00400200e-05 9.00450225e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.00050025e-05 -2.00100050e-05 -3.00150075e-05 -4.00200100e-05\n"," -5.00250125e-05 -6.00300150e-05 -7.00350175e-05 -8.00400200e-05\n"," -9.00450225e-05 -1.00050025e-04]\n","Up to stop (2000) with frq 0.020010, the max (min) modulus of Fourier coeff is: 29102.621265 (1.641079)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 193.948657 (0.152533)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr5/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr5_seg100000_segment1_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29237.631731 (1.361717)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 152.503479 (0.058687)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr5/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr5_seg100000_segment2_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29279.525364 (0.598714)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 194.375809 (0.196587)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr5/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr5_seg100000_segment3_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29378.721858 (0.900946)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 200.155758 (0.093475)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr5/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr5_seg100000_segment4_avgRevCompl0\n","100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7004: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7005: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7006: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7007: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29105.743320 (0.795037)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 179.534320 (0.093396)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr5/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr5_seg100000_segment5_avgRevCompl0\n","Directory /content/drive/MyDrive/DNA_proj/results/predictions/R64_chr6/FourierOnRefBaseProb/20to2000/ created. Output will be placed there.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr6/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr6_seg100000_segment0_avgRevCompl0\n","99950\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6875: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6876: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6877: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1, ax1 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6878: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6879: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2, ax2 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6880: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["avgPredArray size:  99950\n","ratioQ:  1.0\n","Fourier transf lenght:  99950\n","frqs  [ 0.00000000e+00  1.00050025e-05  2.00100050e-05 ... -3.00150075e-05\n"," -2.00100050e-05 -1.00050025e-05]\n","First 10 (positive) frqs:  [0.00000000e+00 1.00050025e-05 2.00100050e-05 3.00150075e-05\n"," 4.00200100e-05 5.00250125e-05 6.00300150e-05 7.00350175e-05\n"," 8.00400200e-05 9.00450225e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.00050025e-05 -2.00100050e-05 -3.00150075e-05 -4.00200100e-05\n"," -5.00250125e-05 -6.00300150e-05 -7.00350175e-05 -8.00400200e-05\n"," -9.00450225e-05 -1.00050025e-04]\n","Up to stop (2000) with frq 0.020010, the max (min) modulus of Fourier coeff is: 29117.474822 (1.102197)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 204.773702 (0.089226)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr6/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr6_seg100000_segment1_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29471.030247 (0.183576)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 237.127451 (0.127469)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr6/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr6_seg100000_segment2_avgRevCompl0\n","Directory /content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/FourierOnRefBaseProb/20to2000/ created. Output will be placed there.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr7_seg100000_segment0_avgRevCompl0\n","99950\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6875: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6876: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6877: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1, ax1 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6878: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6879: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2, ax2 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6880: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["avgPredArray size:  99950\n","ratioQ:  1.0\n","Fourier transf lenght:  99950\n","frqs  [ 0.00000000e+00  1.00050025e-05  2.00100050e-05 ... -3.00150075e-05\n"," -2.00100050e-05 -1.00050025e-05]\n","First 10 (positive) frqs:  [0.00000000e+00 1.00050025e-05 2.00100050e-05 3.00150075e-05\n"," 4.00200100e-05 5.00250125e-05 6.00300150e-05 7.00350175e-05\n"," 8.00400200e-05 9.00450225e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.00050025e-05 -2.00100050e-05 -3.00150075e-05 -4.00200100e-05\n"," -5.00250125e-05 -6.00300150e-05 -7.00350175e-05 -8.00400200e-05\n"," -9.00450225e-05 -1.00050025e-04]\n","Up to stop (2000) with frq 0.020010, the max (min) modulus of Fourier coeff is: 29331.874089 (2.260079)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 181.821248 (0.175409)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr7_seg100000_segment1_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29394.879787 (1.003466)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 189.677139 (0.112374)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr7_seg100000_segment2_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29264.904462 (0.877677)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 177.514211 (0.066210)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr7_seg100000_segment3_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29359.616000 (2.474305)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 209.715024 (0.109686)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr7_seg100000_segment4_avgRevCompl0\n","100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7004: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7005: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7006: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7007: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29524.286480 (2.331624)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 200.636579 (0.142829)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr7_seg100000_segment5_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29077.596322 (0.410617)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 232.956032 (0.204514)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr7_seg100000_segment6_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29537.653317 (1.027334)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 166.157558 (0.154723)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr7_seg100000_segment7_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29219.116226 (0.307662)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 177.776387 (0.139021)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr7_seg100000_segment8_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29573.872471 (0.788738)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 167.596347 (0.103897)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr7_seg100000_segment9_avgRevCompl0\n","100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7004: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7005: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7006: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7007: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29259.128834 (1.452040)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 159.207896 (0.135655)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr7_seg100000_segment10_avgRevCompl0\n","Directory /content/drive/MyDrive/DNA_proj/results/predictions/R64_chr8/FourierOnRefBaseProb/20to2000/ created. Output will be placed there.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr8/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr8_seg100000_segment0_avgRevCompl0\n","99950\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6875: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6876: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6877: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1, ax1 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6878: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6879: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2, ax2 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6880: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["avgPredArray size:  99950\n","ratioQ:  1.0\n","Fourier transf lenght:  99950\n","frqs  [ 0.00000000e+00  1.00050025e-05  2.00100050e-05 ... -3.00150075e-05\n"," -2.00100050e-05 -1.00050025e-05]\n","First 10 (positive) frqs:  [0.00000000e+00 1.00050025e-05 2.00100050e-05 3.00150075e-05\n"," 4.00200100e-05 5.00250125e-05 6.00300150e-05 7.00350175e-05\n"," 8.00400200e-05 9.00450225e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.00050025e-05 -2.00100050e-05 -3.00150075e-05 -4.00200100e-05\n"," -5.00250125e-05 -6.00300150e-05 -7.00350175e-05 -8.00400200e-05\n"," -9.00450225e-05 -1.00050025e-04]\n","Up to stop (2000) with frq 0.020010, the max (min) modulus of Fourier coeff is: 29114.367970 (1.281803)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 182.982966 (0.141092)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr8/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr8_seg100000_segment1_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29307.549824 (1.093302)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 195.051574 (0.113933)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr8/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr8_seg100000_segment2_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29326.616403 (1.041897)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 166.377391 (0.109206)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr8/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr8_seg100000_segment3_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29127.915512 (0.763730)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 163.448475 (0.244965)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr8/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr8_seg100000_segment4_avgRevCompl0\n","100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7004: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7005: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7006: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7007: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29214.955211 (1.177778)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 163.762453 (0.084975)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr8/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr8_seg100000_segment5_avgRevCompl0\n","Directory /content/drive/MyDrive/DNA_proj/results/predictions/R64_chr9/FourierOnRefBaseProb/20to2000/ created. Output will be placed there.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr9/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr9_seg100000_segment0_avgRevCompl0\n","99950\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6875: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6876: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6877: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1, ax1 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6878: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6879: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2, ax2 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6880: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["avgPredArray size:  99950\n","ratioQ:  1.0\n","Fourier transf lenght:  99950\n","frqs  [ 0.00000000e+00  1.00050025e-05  2.00100050e-05 ... -3.00150075e-05\n"," -2.00100050e-05 -1.00050025e-05]\n","First 10 (positive) frqs:  [0.00000000e+00 1.00050025e-05 2.00100050e-05 3.00150075e-05\n"," 4.00200100e-05 5.00250125e-05 6.00300150e-05 7.00350175e-05\n"," 8.00400200e-05 9.00450225e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.00050025e-05 -2.00100050e-05 -3.00150075e-05 -4.00200100e-05\n"," -5.00250125e-05 -6.00300150e-05 -7.00350175e-05 -8.00400200e-05\n"," -9.00450225e-05 -1.00050025e-04]\n","Up to stop (2000) with frq 0.020010, the max (min) modulus of Fourier coeff is: 29236.555776 (0.627924)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 189.140804 (0.063725)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr9/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr9_seg100000_segment1_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29533.858804 (1.070828)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 232.262425 (0.080812)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr9/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr9_seg100000_segment2_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29177.240888 (0.847017)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 164.841169 (0.114387)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr9/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr9_seg100000_segment3_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29191.753572 (2.186414)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 263.336576 (0.056360)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr9/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr9_seg100000_segment4_avgRevCompl0\n","Directory /content/drive/MyDrive/DNA_proj/results/predictions/R64_chr10/FourierOnRefBaseProb/20to2000/ created. Output will be placed there.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr10/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr10_seg100000_segment0_avgRevCompl0\n","99950\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6875: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6876: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6877: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1, ax1 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6878: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6879: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2, ax2 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6880: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["avgPredArray size:  99950\n","ratioQ:  1.0\n","Fourier transf lenght:  99950\n","frqs  [ 0.00000000e+00  1.00050025e-05  2.00100050e-05 ... -3.00150075e-05\n"," -2.00100050e-05 -1.00050025e-05]\n","First 10 (positive) frqs:  [0.00000000e+00 1.00050025e-05 2.00100050e-05 3.00150075e-05\n"," 4.00200100e-05 5.00250125e-05 6.00300150e-05 7.00350175e-05\n"," 8.00400200e-05 9.00450225e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.00050025e-05 -2.00100050e-05 -3.00150075e-05 -4.00200100e-05\n"," -5.00250125e-05 -6.00300150e-05 -7.00350175e-05 -8.00400200e-05\n"," -9.00450225e-05 -1.00050025e-04]\n","Up to stop (2000) with frq 0.020010, the max (min) modulus of Fourier coeff is: 28948.790239 (0.570886)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 157.065084 (0.205506)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr10/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr10_seg100000_segment1_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29394.526334 (1.392831)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 187.825431 (0.395886)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr10/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr10_seg100000_segment2_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29260.767876 (0.681055)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 230.998766 (0.080599)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr10/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr10_seg100000_segment3_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29398.195112 (0.327711)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 215.459063 (0.049385)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr10/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr10_seg100000_segment4_avgRevCompl0\n","100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7004: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7005: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7006: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7007: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29234.369282 (0.912876)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 173.485094 (0.029691)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr10/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr10_seg100000_segment5_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29200.283387 (0.536532)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 158.966616 (0.165391)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr10/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr10_seg100000_segment6_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29300.097024 (1.021956)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 195.396036 (0.201651)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr10/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr10_seg100000_segment7_avgRevCompl0\n","Directory /content/drive/MyDrive/DNA_proj/results/predictions/R64_chr11/FourierOnRefBaseProb/20to2000/ created. Output will be placed there.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr11/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr11_seg100000_segment0_avgRevCompl0\n","99950\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6875: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6876: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6877: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1, ax1 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6878: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6879: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2, ax2 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6880: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["avgPredArray size:  99950\n","ratioQ:  1.0\n","Fourier transf lenght:  99950\n","frqs  [ 0.00000000e+00  1.00050025e-05  2.00100050e-05 ... -3.00150075e-05\n"," -2.00100050e-05 -1.00050025e-05]\n","First 10 (positive) frqs:  [0.00000000e+00 1.00050025e-05 2.00100050e-05 3.00150075e-05\n"," 4.00200100e-05 5.00250125e-05 6.00300150e-05 7.00350175e-05\n"," 8.00400200e-05 9.00450225e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.00050025e-05 -2.00100050e-05 -3.00150075e-05 -4.00200100e-05\n"," -5.00250125e-05 -6.00300150e-05 -7.00350175e-05 -8.00400200e-05\n"," -9.00450225e-05 -1.00050025e-04]\n","Up to stop (2000) with frq 0.020010, the max (min) modulus of Fourier coeff is: 29020.654668 (1.446970)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 217.127108 (0.135478)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr11/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr11_seg100000_segment1_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29446.650132 (1.311551)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 151.372890 (0.170009)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr11/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr11_seg100000_segment2_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29574.773217 (1.199518)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 233.333651 (0.095354)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr11/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr11_seg100000_segment3_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29585.827029 (1.563441)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 215.813154 (0.072270)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr11/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr11_seg100000_segment4_avgRevCompl0\n","100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7004: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7005: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7006: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7007: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29379.177113 (0.814187)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 239.826590 (0.022053)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr11/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr11_seg100000_segment5_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29335.703879 (1.886642)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 173.761128 (0.140023)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr11/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr11_seg100000_segment6_avgRevCompl0\n","Directory /content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/FourierOnRefBaseProb/20to2000/ created. Output will be placed there.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr12_seg100000_segment0_avgRevCompl0\n","99950\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6875: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6876: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6877: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1, ax1 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6878: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6879: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2, ax2 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6880: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["avgPredArray size:  99950\n","ratioQ:  1.0\n","Fourier transf lenght:  99950\n","frqs  [ 0.00000000e+00  1.00050025e-05  2.00100050e-05 ... -3.00150075e-05\n"," -2.00100050e-05 -1.00050025e-05]\n","First 10 (positive) frqs:  [0.00000000e+00 1.00050025e-05 2.00100050e-05 3.00150075e-05\n"," 4.00200100e-05 5.00250125e-05 6.00300150e-05 7.00350175e-05\n"," 8.00400200e-05 9.00450225e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.00050025e-05 -2.00100050e-05 -3.00150075e-05 -4.00200100e-05\n"," -5.00250125e-05 -6.00300150e-05 -7.00350175e-05 -8.00400200e-05\n"," -9.00450225e-05 -1.00050025e-04]\n","Up to stop (2000) with frq 0.020010, the max (min) modulus of Fourier coeff is: 29137.326646 (0.935978)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 215.289807 (0.144831)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr12_seg100000_segment1_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29398.852262 (1.065505)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 178.715377 (0.080945)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr12_seg100000_segment2_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29406.941410 (0.947662)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 172.406727 (0.155657)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr12_seg100000_segment3_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29165.909986 (0.702573)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 276.407490 (0.053485)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr12_seg100000_segment4_avgRevCompl0\n","100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7004: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7005: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7006: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7007: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 28668.860124 (1.011173)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 137.502665 (0.115112)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr12_seg100000_segment5_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29589.990628 (1.330137)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 203.896325 (0.176079)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr12_seg100000_segment6_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29409.193503 (1.023437)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 225.738857 (0.059729)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr12_seg100000_segment7_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29480.249907 (1.315127)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 193.809677 (0.146391)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr12_seg100000_segment8_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29239.617323 (0.725372)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 161.993440 (0.043419)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr12_seg100000_segment9_avgRevCompl0\n","100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7004: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7005: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7006: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7007: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 28983.946808 (0.479065)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 154.375063 (0.176360)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr12_seg100000_segment10_avgRevCompl0\n","Directory /content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/FourierOnRefBaseProb/20to2000/ created. Output will be placed there.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr13_seg100000_segment0_avgRevCompl0\n","99950\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6875: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6876: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6877: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1, ax1 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6878: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6879: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2, ax2 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6880: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["avgPredArray size:  99950\n","ratioQ:  1.0\n","Fourier transf lenght:  99950\n","frqs  [ 0.00000000e+00  1.00050025e-05  2.00100050e-05 ... -3.00150075e-05\n"," -2.00100050e-05 -1.00050025e-05]\n","First 10 (positive) frqs:  [0.00000000e+00 1.00050025e-05 2.00100050e-05 3.00150075e-05\n"," 4.00200100e-05 5.00250125e-05 6.00300150e-05 7.00350175e-05\n"," 8.00400200e-05 9.00450225e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.00050025e-05 -2.00100050e-05 -3.00150075e-05 -4.00200100e-05\n"," -5.00250125e-05 -6.00300150e-05 -7.00350175e-05 -8.00400200e-05\n"," -9.00450225e-05 -1.00050025e-04]\n","Up to stop (2000) with frq 0.020010, the max (min) modulus of Fourier coeff is: 29461.290629 (0.123043)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 227.881130 (0.192511)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr13_seg100000_segment1_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29474.981671 (0.472478)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 225.729071 (0.149143)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr13_seg100000_segment2_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29392.639506 (2.151009)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 173.206110 (0.188750)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr13_seg100000_segment3_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29249.835147 (0.672128)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 181.467142 (0.161474)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr13_seg100000_segment4_avgRevCompl0\n","100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7004: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7005: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7006: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7007: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29242.932006 (1.086159)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 203.873191 (0.071302)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr13_seg100000_segment5_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29395.990130 (1.422726)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 163.068905 (0.126582)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr13_seg100000_segment6_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29433.386078 (1.088562)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 210.614778 (0.230848)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr13_seg100000_segment7_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29424.819590 (2.434429)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 204.482743 (0.164442)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr13_seg100000_segment8_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29255.152713 (1.664943)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 177.681646 (0.210244)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr13_seg100000_segment9_avgRevCompl0\n","Directory /content/drive/MyDrive/DNA_proj/results/predictions/R64_chr14/FourierOnRefBaseProb/20to2000/ created. Output will be placed there.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr14/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr14_seg100000_segment0_avgRevCompl0\n","99950\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6875: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6876: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6877: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1, ax1 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6878: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6879: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2, ax2 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6880: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["avgPredArray size:  99950\n","ratioQ:  1.0\n","Fourier transf lenght:  99950\n","frqs  [ 0.00000000e+00  1.00050025e-05  2.00100050e-05 ... -3.00150075e-05\n"," -2.00100050e-05 -1.00050025e-05]\n","First 10 (positive) frqs:  [0.00000000e+00 1.00050025e-05 2.00100050e-05 3.00150075e-05\n"," 4.00200100e-05 5.00250125e-05 6.00300150e-05 7.00350175e-05\n"," 8.00400200e-05 9.00450225e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.00050025e-05 -2.00100050e-05 -3.00150075e-05 -4.00200100e-05\n"," -5.00250125e-05 -6.00300150e-05 -7.00350175e-05 -8.00400200e-05\n"," -9.00450225e-05 -1.00050025e-04]\n","Up to stop (2000) with frq 0.020010, the max (min) modulus of Fourier coeff is: 28902.672975 (0.947985)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 155.659557 (0.107302)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr14/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr14_seg100000_segment1_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29282.667573 (0.240013)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 199.444990 (0.139288)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr14/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr14_seg100000_segment2_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29536.597133 (1.545502)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 141.866296 (0.126026)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr14/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr14_seg100000_segment3_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29344.481701 (1.408048)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 196.347002 (0.060558)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr14/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr14_seg100000_segment4_avgRevCompl0\n","100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7004: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7005: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7006: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7007: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29404.368736 (0.725354)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 166.239475 (0.155845)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr14/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr14_seg100000_segment5_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29165.157996 (1.478016)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 172.725369 (0.167364)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr14/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr14_seg100000_segment6_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29310.532138 (1.308674)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 163.109082 (0.090038)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr14/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr14_seg100000_segment7_avgRevCompl0\n","Directory /content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/FourierOnRefBaseProb/20to2000/ created. Output will be placed there.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr15_seg100000_segment0_avgRevCompl0\n","99950\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6875: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6876: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6877: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1, ax1 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6878: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6879: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2, ax2 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6880: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["avgPredArray size:  99950\n","ratioQ:  1.0\n","Fourier transf lenght:  99950\n","frqs  [ 0.00000000e+00  1.00050025e-05  2.00100050e-05 ... -3.00150075e-05\n"," -2.00100050e-05 -1.00050025e-05]\n","First 10 (positive) frqs:  [0.00000000e+00 1.00050025e-05 2.00100050e-05 3.00150075e-05\n"," 4.00200100e-05 5.00250125e-05 6.00300150e-05 7.00350175e-05\n"," 8.00400200e-05 9.00450225e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.00050025e-05 -2.00100050e-05 -3.00150075e-05 -4.00200100e-05\n"," -5.00250125e-05 -6.00300150e-05 -7.00350175e-05 -8.00400200e-05\n"," -9.00450225e-05 -1.00050025e-04]\n","Up to stop (2000) with frq 0.020010, the max (min) modulus of Fourier coeff is: 28942.617415 (1.128687)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 165.060105 (0.043624)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr15_seg100000_segment1_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29216.447311 (0.554088)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 187.180583 (0.057903)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr15_seg100000_segment2_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29365.291443 (1.562458)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 176.781444 (0.115612)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr15_seg100000_segment3_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29084.267403 (1.829638)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 168.365584 (0.166227)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr15_seg100000_segment4_avgRevCompl0\n","100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7004: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7005: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7006: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7007: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29280.054831 (2.720182)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 215.674137 (0.116464)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr15_seg100000_segment5_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29381.353305 (0.336463)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 215.400935 (0.114501)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr15_seg100000_segment6_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29191.676909 (0.564287)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 158.144389 (0.047695)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr15_seg100000_segment7_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29477.193800 (0.297428)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 149.809221 (0.096138)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr15_seg100000_segment8_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29313.414808 (0.709066)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 173.470165 (0.078490)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr15_seg100000_segment9_avgRevCompl0\n","100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7004: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7005: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7006: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7007: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29139.058846 (1.109937)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 173.394296 (0.069434)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr15_seg100000_segment10_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr16_seg100000_segment0_avgRevCompl0\n","99950\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6875: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6876: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6877: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1, ax1 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6878: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6879: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2, ax2 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6880: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["avgPredArray size:  99950\n","ratioQ:  1.0\n","Fourier transf lenght:  99950\n","frqs  [ 0.00000000e+00  1.00050025e-05  2.00100050e-05 ... -3.00150075e-05\n"," -2.00100050e-05 -1.00050025e-05]\n","First 10 (positive) frqs:  [0.00000000e+00 1.00050025e-05 2.00100050e-05 3.00150075e-05\n"," 4.00200100e-05 5.00250125e-05 6.00300150e-05 7.00350175e-05\n"," 8.00400200e-05 9.00450225e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.00050025e-05 -2.00100050e-05 -3.00150075e-05 -4.00200100e-05\n"," -5.00250125e-05 -6.00300150e-05 -7.00350175e-05 -8.00400200e-05\n"," -9.00450225e-05 -1.00050025e-04]\n","Up to stop (2000) with frq 0.020010, the max (min) modulus of Fourier coeff is: 28952.290704 (0.650513)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 191.460033 (0.212649)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr16_seg100000_segment1_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29858.577803 (0.727732)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 180.458974 (0.124903)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr16_seg100000_segment2_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29122.071718 (1.361520)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 163.315573 (0.105481)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr16_seg100000_segment3_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29514.527424 (0.728264)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 180.321495 (0.125774)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr16_seg100000_segment4_avgRevCompl0\n","100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7004: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7005: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7006: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7007: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29342.472081 (1.194753)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 135.979735 (0.027423)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr16_seg100000_segment5_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29186.730451 (1.236069)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 164.866802 (0.074335)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr16_seg100000_segment6_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29286.250912 (2.186335)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 151.434222 (0.122156)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr16_seg100000_segment7_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29188.159836 (0.223435)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 228.138809 (0.113080)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr16_seg100000_segment8_avgRevCompl0\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 29308.605545 (1.630856)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 150.616714 (0.132629)\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr16_seg100000_segment9_avgRevCompl0\n"]}]},{"cell_type":"markdown","source":["You should now be able to find the output plots in  FouriersOnRefBaseProb/20To2000 subfolders of the predictions/chrNN folders.\n","\n","To run the Fourier on GC content you must first create the GC-indicator arrays for the purpose. This can be done by abusing the code aimed for computing accuracies (by a little 'hack' we arrange it so that the model is always right if the given position holds a G or C, and else it is wrong):"],"metadata":{"id":"njKOcsDw2b0D"}},{"cell_type":"code","source":["#This cell is just for the case that we have changed the code and need to reload:\n","import imp\n","imp.reload(stats)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"PzhBHtzrlYjt","executionInfo":{"status":"ok","timestamp":1650143069107,"user_tz":-120,"elapsed":659,"user":{"displayName":"Christian Grønbæk","userId":"14909394381660051993"}},"outputId":"39a2882e-44d5-4f9e-917c-55bb388b4208"},"execution_count":27,"outputs":[{"output_type":"execute_result","data":{"text/plain":["<module 'dnaNet_stats_py3' from '/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py'>"]},"metadata":{},"execution_count":27}]},{"cell_type":"code","source":["rootGenome = rootGenome\n","\n","rootOutput = r\"/content/drive/MyDrive/DNA_proj/results/predictions/\"\n","rootModel = r\"/content/drive/MyDrive/DNA_proj/results/training/\"\n","lastRepeatNr = nrOfRepeats -1 \n","modelFileNameNN = \"modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr\" + str(lastRepeatNr)\n","\n","rootOutputBias = r\"/content/drive/MyDrive/DNA_proj/results/\"  \n","\n","chromosomeOrderList = ['R64_chr1', 'R64_chr2', 'R64_chr3', 'R64_chr4', 'R64_chr5', 'R64_chr6', 'R64_chr7', 'R64_chr8','R64_chr9', 'R64_chr10', 'R64_chr11', 'R64_chr12','R64_chr13', 'R64_chr14', 'R64_chr15', 'R64_chr16']\n","chromosomeDict = {'R64_chr1':[0,1e8], 'R64_chr2':[0,1e8], 'R64_chr3':[0,1e8], 'R64_chr4':[0,1e8], 'R64_chr5':[0,1e8], 'R64_chr6':[0,1e8], 'R64_chr7':[0,1e8], 'R64_chr8':[0,1e8],'R64_chr9':[0,1e8], 'R64_chr10':[0,1e8], 'R64_chr11':[0,1e8], 'R64_chr12':[0,1e8],'R64_chr13':[0,1e8], 'R64_chr14':[0,1e8], 'R64_chr15':[0,1e8], 'R64_chr16':[0,1e8]}\n","startAtSegmentDict = {}\n"],"metadata":{"id":"I7N4S6yq2bDF","executionInfo":{"status":"ok","timestamp":1650142100950,"user_tz":-120,"elapsed":236,"user":{"displayName":"Christian Grønbæk","userId":"14909394381660051993"}}},"execution_count":22,"outputs":[]},{"cell_type":"markdown","source":["The accuracies (as 'average prediction' or so) that get printed to the screen are the GC%'s .. it takes a little while to run this (a few minutes):"],"metadata":{"id":"buec0jizMf4r"}},{"cell_type":"code","source":["#set this so that it matches the setting for the computed predArray(s)\n","segmentLength = 100000\n","averageRevComplementary_b = 0\n","\n","#set these as desired\n","windowLength = 1\n","stepSize = 1\n","Fourier_b = 0\n","on_binf_b = 1\n","defaultAccuracy = 0.25\n","\n","#CG bias, little 'hack':\n","forATorGCbias_b = 1 #!\n","recodeA = [0,0,0,0]\n","recodeC = [1,1,1,1]\n","recodeG = [1,1,1,1]\n","recodeT = [0,0,0,0] \n","modelFileName_forATorGCbias =\"GCbias\"\n","rootOutput_forATorGCbias  = rootOutputBias + r\"GCbias/\"\n","\n","stats.computeAccuracyOnChromosomes(rootGenome = rootGenome, \n","                         chromosomeDict = chromosomeDict,\n","                         chromosomeOrderList = chromosomeOrderList, \n","                         rootOutput = rootOutput,\n","                         rootModel = rootModel,\n","                         modelFileName = modelFileNameNN,\n","                        segmentLength = segmentLength,\n","                        startAtSegmentDict = startAtSegmentDict,\n","                        averageRevComplementary_b = averageRevComplementary_b, #!!!!!\n","                        windowLength = windowLength,\n","                        stepSize = stepSize,\n","                        Fourier_b = Fourier_b,\n","                        defaultAccuracy = defaultAccuracy,\n","                        on_binf_b = on_binf_b,\n","                        forATorGCbias_b = forATorGCbias_b, \n","                        rootOutput_forATorGCbias= rootOutput_forATorGCbias,\n","                        modelFileName_forATorGCbias = modelFileName_forATorGCbias,\n","                        recodeA = recodeA,\n","                             recodeC = recodeC,\n","                             recodeG = recodeG,\n","                             recodeT = recodeT\n","                             )"],"metadata":{"id":"wuajDMn37h2x","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1650143439818,"user_tz":-120,"elapsed":366961,"user":{"displayName":"Christian Grønbæk","userId":"14909394381660051993"}},"outputId":"68dabfe5-39a4-48aa-b566-64741b023952"},"execution_count":28,"outputs":[{"output_type":"stream","name":"stdout","text":["Reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 200\n","Fast reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 200\n","OBS: no file containing exonic info was provided, so exonic status is set to 0 from 0 - 100000001\n","Genome data file 1st line:\n","  >ref|NC_001133| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=I]\n","\n","Found data for this chromosome: ref|NC_001133| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=I]\n","60 60 60\n","[['ref|NC_001133| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=I]', 230218]]\n","[]\n","Length of genome sequence read in:230218\n","Length of exonic-info sequence read in:230218\n","ACGTacgt checked 0 tokens\n","Length genome sequence, ACGT's and W's:230218\n","Of these 0 are W's\n","Length genome sequence, only ACGT's:230218\n","I'm doing a forATorGCbias run!\n","nrSegments:  2\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr1/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr1_seg100000_segment0_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr1/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr1_seg100000_segment0_avgRevCompl0\n","lSamples  99950 , nrSteps  99950\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 39135 , cntTot: 99950 ; average prediction acc : 0.391546\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.391546\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr1/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr1_seg100000_segment1_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr1/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr1_seg100000_segment1_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 39455 , cntTot: 100000 ; average prediction acc : 0.394550\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.394550\n","No repeat sections were recorded in the genome data.\n","Reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 200\n","Fast reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 200\n","OBS: no file containing exonic info was provided, so exonic status is set to 0 from 0 - 100000001\n","Genome data file 1st line:\n","  >ref|NC_001134| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=II]\n","\n","Found data for this chromosome: ref|NC_001134| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=II]\n","60 60 60\n","[['ref|NC_001134| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=II]', 813184]]\n","[]\n","Length of genome sequence read in:813184\n","Length of exonic-info sequence read in:813184\n","ACGTacgt checked 0 tokens\n","Length genome sequence, ACGT's and W's:813184\n","Of these 0 are W's\n","Length genome sequence, only ACGT's:813184\n","I'm doing a forATorGCbias run!\n","nrSegments:  8\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr2_seg100000_segment0_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr2_seg100000_segment0_avgRevCompl0\n","lSamples  99950 , nrSteps  99950\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38526 , cntTot: 99950 ; average prediction acc : 0.385453\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.385453\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr2_seg100000_segment1_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr2_seg100000_segment1_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38442 , cntTot: 100000 ; average prediction acc : 0.384420\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.384420\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr2_seg100000_segment2_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr2_seg100000_segment2_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38137 , cntTot: 100000 ; average prediction acc : 0.381370\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.381370\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr2_seg100000_segment3_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr2_seg100000_segment3_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37965 , cntTot: 100000 ; average prediction acc : 0.379650\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.379650\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr2_seg100000_segment4_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr2_seg100000_segment4_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38670 , cntTot: 100000 ; average prediction acc : 0.386700\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.386700\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr2_seg100000_segment5_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr2_seg100000_segment5_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37394 , cntTot: 100000 ; average prediction acc : 0.373940\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.373940\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr2_seg100000_segment6_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr2_seg100000_segment6_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 39083 , cntTot: 100000 ; average prediction acc : 0.390830\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.390830\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr2_seg100000_segment7_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr2/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr2_seg100000_segment7_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38455 , cntTot: 100000 ; average prediction acc : 0.384550\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.384550\n","No repeat sections were recorded in the genome data.\n","Reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 200\n","Fast reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 200\n","OBS: no file containing exonic info was provided, so exonic status is set to 0 from 0 - 100000001\n","Genome data file 1st line:\n","  >ref|NC_001135| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=III]\n","\n","Found data for this chromosome: ref|NC_001135| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=III]\n","60 60 60\n","[['ref|NC_001135| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=III]', 316620]]\n","[]\n","Length of genome sequence read in:316620\n","Length of exonic-info sequence read in:316620\n","ACGTacgt checked 0 tokens\n","Length genome sequence, ACGT's and W's:316620\n","Of these 0 are W's\n","Length genome sequence, only ACGT's:316620\n","I'm doing a forATorGCbias run!\n","nrSegments:  3\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr3/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr3_seg100000_segment0_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr3/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr3_seg100000_segment0_avgRevCompl0\n","lSamples  99950 , nrSteps  99950\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 39102 , cntTot: 99950 ; average prediction acc : 0.391216\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.391216\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr3/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr3_seg100000_segment1_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr3/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr3_seg100000_segment1_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37197 , cntTot: 100000 ; average prediction acc : 0.371970\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.371970\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr3/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr3_seg100000_segment2_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr3/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr3_seg100000_segment2_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 39289 , cntTot: 100000 ; average prediction acc : 0.392890\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.392890\n","No repeat sections were recorded in the genome data.\n","Reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 200\n","Fast reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 200\n","OBS: no file containing exonic info was provided, so exonic status is set to 0 from 0 - 100000001\n","Genome data file 1st line:\n","  >ref|NC_001136| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=IV]\n","\n","Found data for this chromosome: ref|NC_001136| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=IV]\n","60 60 60\n","[['ref|NC_001136| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=IV]', 1531933]]\n","[]\n","Length of genome sequence read in:1531933\n","Length of exonic-info sequence read in:1531933\n","ACGTacgt checked 0 tokens\n","Length genome sequence, ACGT's and W's:1531933\n","Of these 0 are W's\n","Length genome sequence, only ACGT's:1531933\n","I'm doing a forATorGCbias run!\n","nrSegments:  15\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr4_seg100000_segment0_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr4_seg100000_segment0_avgRevCompl0\n","lSamples  99950 , nrSteps  99950\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38375 , cntTot: 99950 ; average prediction acc : 0.383942\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.383942\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr4_seg100000_segment1_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr4_seg100000_segment1_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38615 , cntTot: 100000 ; average prediction acc : 0.386150\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.386150\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr4_seg100000_segment2_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr4_seg100000_segment2_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38235 , cntTot: 100000 ; average prediction acc : 0.382350\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.382350\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr4_seg100000_segment3_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr4_seg100000_segment3_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38186 , cntTot: 100000 ; average prediction acc : 0.381860\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.381860\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr4_seg100000_segment4_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr4_seg100000_segment4_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37698 , cntTot: 100000 ; average prediction acc : 0.376980\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.376980\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr4_seg100000_segment5_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr4_seg100000_segment5_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37472 , cntTot: 100000 ; average prediction acc : 0.374720\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.374720\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr4_seg100000_segment6_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr4_seg100000_segment6_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37508 , cntTot: 100000 ; average prediction acc : 0.375080\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.375080\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr4_seg100000_segment7_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr4_seg100000_segment7_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37631 , cntTot: 100000 ; average prediction acc : 0.376310\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.376310\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr4_seg100000_segment8_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr4_seg100000_segment8_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37892 , cntTot: 100000 ; average prediction acc : 0.378920\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.378920\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr4_seg100000_segment9_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr4_seg100000_segment9_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37629 , cntTot: 100000 ; average prediction acc : 0.376290\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.376290\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr4_seg100000_segment10_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr4_seg100000_segment10_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37375 , cntTot: 100000 ; average prediction acc : 0.373750\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.373750\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr4_seg100000_segment11_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr4_seg100000_segment11_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37339 , cntTot: 100000 ; average prediction acc : 0.373390\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.373390\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr4_seg100000_segment12_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr4_seg100000_segment12_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38006 , cntTot: 100000 ; average prediction acc : 0.380060\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.380060\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr4_seg100000_segment13_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr4_seg100000_segment13_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38110 , cntTot: 100000 ; average prediction acc : 0.381100\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.381100\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr4_seg100000_segment14_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr4/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr4_seg100000_segment14_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38188 , cntTot: 100000 ; average prediction acc : 0.381880\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.381880\n","No repeat sections were recorded in the genome data.\n","Reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 200\n","Fast reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 200\n","OBS: no file containing exonic info was provided, so exonic status is set to 0 from 0 - 100000001\n","Genome data file 1st line:\n","  >ref|NC_001137| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=V]\n","\n","Found data for this chromosome: ref|NC_001137| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=V]\n","60 60 60\n","[['ref|NC_001137| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=V]', 576874]]\n","[]\n","Length of genome sequence read in:576874\n","Length of exonic-info sequence read in:576874\n","ACGTacgt checked 0 tokens\n","Length genome sequence, ACGT's and W's:576874\n","Of these 0 are W's\n","Length genome sequence, only ACGT's:576874\n","I'm doing a forATorGCbias run!\n","Directory /content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr5/ created. Output will be placed there.\n","nrSegments:  5\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr5/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr5_seg100000_segment0_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr5/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr5_seg100000_segment0_avgRevCompl0\n","lSamples  99950 , nrSteps  99950\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 39171 , cntTot: 99950 ; average prediction acc : 0.391906\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.391906\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr5/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr5_seg100000_segment1_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr5/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr5_seg100000_segment1_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37870 , cntTot: 100000 ; average prediction acc : 0.378700\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.378700\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr5/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr5_seg100000_segment2_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr5/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr5_seg100000_segment2_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38470 , cntTot: 100000 ; average prediction acc : 0.384700\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.384700\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr5/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr5_seg100000_segment3_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr5/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr5_seg100000_segment3_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38571 , cntTot: 100000 ; average prediction acc : 0.385710\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.385710\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr5/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr5_seg100000_segment4_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr5/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr5_seg100000_segment4_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38311 , cntTot: 100000 ; average prediction acc : 0.383110\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.383110\n","No repeat sections were recorded in the genome data.\n","Reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 200\n","Fast reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 200\n","OBS: no file containing exonic info was provided, so exonic status is set to 0 from 0 - 100000001\n","Genome data file 1st line:\n","  >ref|NC_001138| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VI]\n","\n","Found data for this chromosome: ref|NC_001138| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VI]\n","60 60 60\n","[['ref|NC_001138| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VI]', 270161]]\n","[]\n","Length of genome sequence read in:270161\n","Length of exonic-info sequence read in:270161\n","ACGTacgt checked 0 tokens\n","Length genome sequence, ACGT's and W's:270161\n","Of these 0 are W's\n","Length genome sequence, only ACGT's:270161\n","I'm doing a forATorGCbias run!\n","Directory /content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr6/ created. Output will be placed there.\n","nrSegments:  2\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr6/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr6_seg100000_segment0_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr6/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr6_seg100000_segment0_avgRevCompl0\n","lSamples  99950 , nrSteps  99950\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 39447 , cntTot: 99950 ; average prediction acc : 0.394667\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.394667\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr6/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr6_seg100000_segment1_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr6/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr6_seg100000_segment1_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37861 , cntTot: 100000 ; average prediction acc : 0.378610\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.378610\n","No repeat sections were recorded in the genome data.\n","Reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 200\n","Fast reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 200\n","OBS: no file containing exonic info was provided, so exonic status is set to 0 from 0 - 100000001\n","Genome data file 1st line:\n","  >ref|NC_001139| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VII]\n","\n","Found data for this chromosome: ref|NC_001139| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VII]\n","60 60 60\n","[['ref|NC_001139| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VII]', 1090940]]\n","[]\n","Length of genome sequence read in:1090940\n","Length of exonic-info sequence read in:1090940\n","ACGTacgt checked 0 tokens\n","Length genome sequence, ACGT's and W's:1090940\n","Of these 0 are W's\n","Length genome sequence, only ACGT's:1090940\n","I'm doing a forATorGCbias run!\n","Directory /content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr7/ created. Output will be placed there.\n","nrSegments:  10\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr7_seg100000_segment0_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr7_seg100000_segment0_avgRevCompl0\n","lSamples  99950 , nrSteps  99950\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37696 , cntTot: 99950 ; average prediction acc : 0.377149\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.377149\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr7_seg100000_segment1_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr7_seg100000_segment1_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38097 , cntTot: 100000 ; average prediction acc : 0.380970\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.380970\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr7_seg100000_segment2_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr7_seg100000_segment2_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37976 , cntTot: 100000 ; average prediction acc : 0.379760\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.379760\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr7_seg100000_segment3_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr7_seg100000_segment3_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37900 , cntTot: 100000 ; average prediction acc : 0.379000\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.379000\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr7_seg100000_segment4_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr7_seg100000_segment4_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38389 , cntTot: 100000 ; average prediction acc : 0.383890\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.383890\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr7_seg100000_segment5_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr7_seg100000_segment5_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38153 , cntTot: 100000 ; average prediction acc : 0.381530\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.381530\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr7_seg100000_segment6_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr7_seg100000_segment6_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37736 , cntTot: 100000 ; average prediction acc : 0.377360\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.377360\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr7_seg100000_segment7_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr7_seg100000_segment7_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37740 , cntTot: 100000 ; average prediction acc : 0.377400\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.377400\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr7_seg100000_segment8_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr7_seg100000_segment8_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37764 , cntTot: 100000 ; average prediction acc : 0.377640\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.377640\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr7_seg100000_segment9_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr7/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr7_seg100000_segment9_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38346 , cntTot: 100000 ; average prediction acc : 0.383460\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.383460\n","No repeat sections were recorded in the genome data.\n","Reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 200\n","Fast reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 200\n","OBS: no file containing exonic info was provided, so exonic status is set to 0 from 0 - 100000001\n","Genome data file 1st line:\n","  >ref|NC_001140| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VIII]\n","\n","Found data for this chromosome: ref|NC_001140| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VIII]\n","60 60 60\n","[['ref|NC_001140| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=VIII]', 562643]]\n","[]\n","Length of genome sequence read in:562643\n","Length of exonic-info sequence read in:562643\n","ACGTacgt checked 0 tokens\n","Length genome sequence, ACGT's and W's:562643\n","Of these 0 are W's\n","Length genome sequence, only ACGT's:562643\n","I'm doing a forATorGCbias run!\n","Directory /content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr8/ created. Output will be placed there.\n","nrSegments:  5\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr8/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr8_seg100000_segment0_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr8/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr8_seg100000_segment0_avgRevCompl0\n","lSamples  99950 , nrSteps  99950\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38692 , cntTot: 99950 ; average prediction acc : 0.387114\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.387114\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr8/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr8_seg100000_segment1_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr8/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr8_seg100000_segment1_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38678 , cntTot: 100000 ; average prediction acc : 0.386780\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.386780\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr8/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr8_seg100000_segment2_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr8/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr8_seg100000_segment2_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37969 , cntTot: 100000 ; average prediction acc : 0.379690\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.379690\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr8/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr8_seg100000_segment3_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr8/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr8_seg100000_segment3_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38309 , cntTot: 100000 ; average prediction acc : 0.383090\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.383090\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr8/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr8_seg100000_segment4_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr8/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr8_seg100000_segment4_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38438 , cntTot: 100000 ; average prediction acc : 0.384380\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.384380\n","No repeat sections were recorded in the genome data.\n","Reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 200\n","Fast reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 200\n","OBS: no file containing exonic info was provided, so exonic status is set to 0 from 0 - 100000001\n","Genome data file 1st line:\n","  >ref|NC_001141| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=IX]\n","\n","Found data for this chromosome: ref|NC_001141| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=IX]\n","60 60 60\n","[['ref|NC_001141| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=IX]', 439888]]\n","[]\n","Length of genome sequence read in:439888\n","Length of exonic-info sequence read in:439888\n","ACGTacgt checked 0 tokens\n","Length genome sequence, ACGT's and W's:439888\n","Of these 0 are W's\n","Length genome sequence, only ACGT's:439888\n","I'm doing a forATorGCbias run!\n","Directory /content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr9/ created. Output will be placed there.\n","nrSegments:  4\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr9/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr9_seg100000_segment0_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr9/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr9_seg100000_segment0_avgRevCompl0\n","lSamples  99950 , nrSteps  99950\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 39401 , cntTot: 99950 ; average prediction acc : 0.394207\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.394207\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr9/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr9_seg100000_segment1_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr9/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr9_seg100000_segment1_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38234 , cntTot: 100000 ; average prediction acc : 0.382340\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.382340\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr9/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr9_seg100000_segment2_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr9/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr9_seg100000_segment2_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38701 , cntTot: 100000 ; average prediction acc : 0.387010\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.387010\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr9/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr9_seg100000_segment3_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr9/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr9_seg100000_segment3_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 39173 , cntTot: 100000 ; average prediction acc : 0.391730\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.391730\n","No repeat sections were recorded in the genome data.\n","Reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 200\n","Fast reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 200\n","OBS: no file containing exonic info was provided, so exonic status is set to 0 from 0 - 100000001\n","Genome data file 1st line:\n","  >ref|NC_001142| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=X]\n","\n","Found data for this chromosome: ref|NC_001142| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=X]\n","60 60 60\n","[['ref|NC_001142| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=X]', 745751]]\n","[]\n","Length of genome sequence read in:745751\n","Length of exonic-info sequence read in:745751\n","ACGTacgt checked 0 tokens\n","Length genome sequence, ACGT's and W's:745751\n","Of these 0 are W's\n","Length genome sequence, only ACGT's:745751\n","I'm doing a forATorGCbias run!\n","Directory /content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr10/ created. Output will be placed there.\n","nrSegments:  7\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr10/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr10_seg100000_segment0_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr10/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr10_seg100000_segment0_avgRevCompl0\n","lSamples  99950 , nrSteps  99950\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38460 , cntTot: 99950 ; average prediction acc : 0.384792\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.384792\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr10/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr10_seg100000_segment1_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr10/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr10_seg100000_segment1_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38616 , cntTot: 100000 ; average prediction acc : 0.386160\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.386160\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr10/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr10_seg100000_segment2_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr10/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr10_seg100000_segment2_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37912 , cntTot: 100000 ; average prediction acc : 0.379120\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.379120\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr10/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr10_seg100000_segment3_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr10/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr10_seg100000_segment3_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38415 , cntTot: 100000 ; average prediction acc : 0.384150\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.384150\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr10/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr10_seg100000_segment4_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr10/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr10_seg100000_segment4_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38054 , cntTot: 100000 ; average prediction acc : 0.380540\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.380540\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr10/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr10_seg100000_segment5_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr10/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr10_seg100000_segment5_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38367 , cntTot: 100000 ; average prediction acc : 0.383670\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.383670\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr10/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr10_seg100000_segment6_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr10/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr10_seg100000_segment6_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38429 , cntTot: 100000 ; average prediction acc : 0.384290\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.384290\n","No repeat sections were recorded in the genome data.\n","Reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 200\n","Fast reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 200\n","OBS: no file containing exonic info was provided, so exonic status is set to 0 from 0 - 100000001\n","Genome data file 1st line:\n","  >ref|NC_001143| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XI]\n","\n","Found data for this chromosome: ref|NC_001143| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XI]\n","60 60 60\n","[['ref|NC_001143| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XI]', 666816]]\n","[]\n","Length of genome sequence read in:666816\n","Length of exonic-info sequence read in:666816\n","ACGTacgt checked 0 tokens\n","Length genome sequence, ACGT's and W's:666816\n","Of these 0 are W's\n","Length genome sequence, only ACGT's:666816\n","I'm doing a forATorGCbias run!\n","Directory /content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr11/ created. Output will be placed there.\n","nrSegments:  6\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr11/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr11_seg100000_segment0_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr11/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr11_seg100000_segment0_avgRevCompl0\n","lSamples  99950 , nrSteps  99950\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38333 , cntTot: 99950 ; average prediction acc : 0.383522\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.383522\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr11/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr11_seg100000_segment1_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr11/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr11_seg100000_segment1_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38018 , cntTot: 100000 ; average prediction acc : 0.380180\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.380180\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr11/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr11_seg100000_segment2_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr11/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr11_seg100000_segment2_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38136 , cntTot: 100000 ; average prediction acc : 0.381360\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.381360\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr11/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr11_seg100000_segment3_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr11/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr11_seg100000_segment3_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38240 , cntTot: 100000 ; average prediction acc : 0.382400\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.382400\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr11/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr11_seg100000_segment4_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr11/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr11_seg100000_segment4_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37276 , cntTot: 100000 ; average prediction acc : 0.372760\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.372760\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr11/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr11_seg100000_segment5_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr11/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr11_seg100000_segment5_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38673 , cntTot: 100000 ; average prediction acc : 0.386730\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.386730\n","No repeat sections were recorded in the genome data.\n","Reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 200\n","Fast reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 200\n","OBS: no file containing exonic info was provided, so exonic status is set to 0 from 0 - 100000001\n","Genome data file 1st line:\n","  >ref|NC_001144| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XII]\n","\n","Found data for this chromosome: ref|NC_001144| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XII]\n","60 60 60\n","[['ref|NC_001144| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XII]', 1078177]]\n","[]\n","Length of genome sequence read in:1078177\n","Length of exonic-info sequence read in:1078177\n","ACGTacgt checked 0 tokens\n","Length genome sequence, ACGT's and W's:1078177\n","Of these 0 are W's\n","Length genome sequence, only ACGT's:1078177\n","I'm doing a forATorGCbias run!\n","Directory /content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr12/ created. Output will be placed there.\n","nrSegments:  10\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr12_seg100000_segment0_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr12_seg100000_segment0_avgRevCompl0\n","lSamples  99950 , nrSteps  99950\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 39232 , cntTot: 99950 ; average prediction acc : 0.392516\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.392516\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr12_seg100000_segment1_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr12_seg100000_segment1_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37789 , cntTot: 100000 ; average prediction acc : 0.377890\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.377890\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr12_seg100000_segment2_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr12_seg100000_segment2_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38104 , cntTot: 100000 ; average prediction acc : 0.381040\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.381040\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr12_seg100000_segment3_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr12_seg100000_segment3_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38704 , cntTot: 100000 ; average prediction acc : 0.387040\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.387040\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr12_seg100000_segment4_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr12_seg100000_segment4_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 39938 , cntTot: 100000 ; average prediction acc : 0.399380\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.399380\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr12_seg100000_segment5_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr12_seg100000_segment5_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38734 , cntTot: 100000 ; average prediction acc : 0.387340\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.387340\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr12_seg100000_segment6_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr12_seg100000_segment6_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37776 , cntTot: 100000 ; average prediction acc : 0.377760\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.377760\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr12_seg100000_segment7_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr12_seg100000_segment7_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37685 , cntTot: 100000 ; average prediction acc : 0.376850\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.376850\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr12_seg100000_segment8_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr12_seg100000_segment8_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38739 , cntTot: 100000 ; average prediction acc : 0.387390\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.387390\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr12_seg100000_segment9_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr12/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr12_seg100000_segment9_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37978 , cntTot: 100000 ; average prediction acc : 0.379780\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.379780\n","No repeat sections were recorded in the genome data.\n","Reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 200\n","Fast reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 200\n","OBS: no file containing exonic info was provided, so exonic status is set to 0 from 0 - 100000001\n","Genome data file 1st line:\n","  >ref|NC_001145| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XIII]\n","\n","Found data for this chromosome: ref|NC_001145| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XIII]\n","60 60 60\n","[['ref|NC_001145| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XIII]', 924431]]\n","[]\n","Length of genome sequence read in:924431\n","Length of exonic-info sequence read in:924431\n","ACGTacgt checked 0 tokens\n","Length genome sequence, ACGT's and W's:924431\n","Of these 0 are W's\n","Length genome sequence, only ACGT's:924431\n","I'm doing a forATorGCbias run!\n","Directory /content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr13/ created. Output will be placed there.\n","nrSegments:  9\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr13_seg100000_segment0_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr13_seg100000_segment0_avgRevCompl0\n","lSamples  99950 , nrSteps  99950\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38751 , cntTot: 99950 ; average prediction acc : 0.387704\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.387704\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr13_seg100000_segment1_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr13_seg100000_segment1_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38215 , cntTot: 100000 ; average prediction acc : 0.382150\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.382150\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr13_seg100000_segment2_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr13_seg100000_segment2_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38026 , cntTot: 100000 ; average prediction acc : 0.380260\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.380260\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr13_seg100000_segment3_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr13_seg100000_segment3_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38352 , cntTot: 100000 ; average prediction acc : 0.383520\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.383520\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr13_seg100000_segment4_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr13_seg100000_segment4_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37992 , cntTot: 100000 ; average prediction acc : 0.379920\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.379920\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr13_seg100000_segment5_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr13_seg100000_segment5_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37532 , cntTot: 100000 ; average prediction acc : 0.375320\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.375320\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr13_seg100000_segment6_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr13_seg100000_segment6_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38263 , cntTot: 100000 ; average prediction acc : 0.382630\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.382630\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr13_seg100000_segment7_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr13_seg100000_segment7_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37848 , cntTot: 100000 ; average prediction acc : 0.378480\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.378480\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr13_seg100000_segment8_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr13/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr13_seg100000_segment8_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38462 , cntTot: 100000 ; average prediction acc : 0.384620\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.384620\n","No repeat sections were recorded in the genome data.\n","Reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 200\n","Fast reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 200\n","OBS: no file containing exonic info was provided, so exonic status is set to 0 from 0 - 100000001\n","Genome data file 1st line:\n","  >ref|NC_001146| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XIV]\n","\n","Found data for this chromosome: ref|NC_001146| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XIV]\n","60 60 60\n","[['ref|NC_001146| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XIV]', 784333]]\n","[]\n","Length of genome sequence read in:784333\n","Length of exonic-info sequence read in:784333\n","ACGTacgt checked 0 tokens\n","Length genome sequence, ACGT's and W's:784333\n","Of these 0 are W's\n","Length genome sequence, only ACGT's:784333\n","I'm doing a forATorGCbias run!\n","Directory /content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr14/ created. Output will be placed there.\n","nrSegments:  7\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr14/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr14_seg100000_segment0_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr14/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr14_seg100000_segment0_avgRevCompl0\n","lSamples  99950 , nrSteps  99950\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 39249 , cntTot: 99950 ; average prediction acc : 0.392686\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.392686\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr14/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr14_seg100000_segment1_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr14/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr14_seg100000_segment1_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38842 , cntTot: 100000 ; average prediction acc : 0.388420\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.388420\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr14/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr14_seg100000_segment2_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr14/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr14_seg100000_segment2_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38162 , cntTot: 100000 ; average prediction acc : 0.381620\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.381620\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr14/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr14_seg100000_segment3_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr14/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr14_seg100000_segment3_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38393 , cntTot: 100000 ; average prediction acc : 0.383930\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.383930\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr14/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr14_seg100000_segment4_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr14/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr14_seg100000_segment4_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38514 , cntTot: 100000 ; average prediction acc : 0.385140\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.385140\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr14/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr14_seg100000_segment5_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr14/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr14_seg100000_segment5_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38288 , cntTot: 100000 ; average prediction acc : 0.382880\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.382880\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr14/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr14_seg100000_segment6_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr14/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr14_seg100000_segment6_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 39000 , cntTot: 100000 ; average prediction acc : 0.390000\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.390000\n","No repeat sections were recorded in the genome data.\n","Reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 200\n","Fast reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 200\n","OBS: no file containing exonic info was provided, so exonic status is set to 0 from 0 - 100000001\n","Genome data file 1st line:\n","  >ref|NC_001147| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XV]\n","\n","Found data for this chromosome: ref|NC_001147| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XV]\n","60 60 60\n","[['ref|NC_001147| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XV]', 1091291]]\n","[]\n","Length of genome sequence read in:1091291\n","Length of exonic-info sequence read in:1091291\n","ACGTacgt checked 0 tokens\n","Length genome sequence, ACGT's and W's:1091291\n","Of these 0 are W's\n","Length genome sequence, only ACGT's:1091291\n","I'm doing a forATorGCbias run!\n","Directory /content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr15/ created. Output will be placed there.\n","nrSegments:  10\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr15_seg100000_segment0_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr15_seg100000_segment0_avgRevCompl0\n","lSamples  99950 , nrSteps  99950\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38265 , cntTot: 99950 ; average prediction acc : 0.382841\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.382841\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr15_seg100000_segment1_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr15_seg100000_segment1_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37846 , cntTot: 100000 ; average prediction acc : 0.378460\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.378460\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr15_seg100000_segment2_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr15_seg100000_segment2_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38721 , cntTot: 100000 ; average prediction acc : 0.387210\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.387210\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr15_seg100000_segment3_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr15_seg100000_segment3_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37907 , cntTot: 100000 ; average prediction acc : 0.379070\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.379070\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr15_seg100000_segment4_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr15_seg100000_segment4_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38306 , cntTot: 100000 ; average prediction acc : 0.383060\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.383060\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr15_seg100000_segment5_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr15_seg100000_segment5_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38445 , cntTot: 100000 ; average prediction acc : 0.384450\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.384450\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr15_seg100000_segment6_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr15_seg100000_segment6_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37521 , cntTot: 100000 ; average prediction acc : 0.375210\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.375210\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr15_seg100000_segment7_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr15_seg100000_segment7_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37838 , cntTot: 100000 ; average prediction acc : 0.378380\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.378380\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr15_seg100000_segment8_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr15_seg100000_segment8_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37807 , cntTot: 100000 ; average prediction acc : 0.378070\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.378070\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr15_seg100000_segment9_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr15/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr15_seg100000_segment9_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38331 , cntTot: 100000 ; average prediction acc : 0.383310\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.383310\n","No repeat sections were recorded in the genome data.\n","Reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 200\n","Fast reading in genome data ... \n","Only considering data following fasta header lines (: chromo names \n"," for eucaryots) of length < 200\n","OBS: no file containing exonic info was provided, so exonic status is set to 0 from 0 - 100000001\n","Genome data file 1st line:\n","  >ref|NC_001148| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XVI]\n","\n","Found data for this chromosome: ref|NC_001148| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XVI]\n","60 60 60\n","[['ref|NC_001148| [org=Saccharomyces cerevisiae] [strain=S288C] [moltype=genomic] [chromosome=XVI]', 948066]]\n","[]\n","Length of genome sequence read in:948066\n","Length of exonic-info sequence read in:948066\n","ACGTacgt checked 0 tokens\n","Length genome sequence, ACGT's and W's:948066\n","Of these 0 are W's\n","Length genome sequence, only ACGT's:948066\n","I'm doing a forATorGCbias run!\n","Directory /content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr16/ created. Output will be placed there.\n","nrSegments:  9\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr16_seg100000_segment0_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr16_seg100000_segment0_avgRevCompl0\n","lSamples  99950 , nrSteps  99950\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38689 , cntTot: 99950 ; average prediction acc : 0.387084\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.387084\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr16_seg100000_segment1_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr16_seg100000_segment1_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38260 , cntTot: 100000 ; average prediction acc : 0.382600\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.382600\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr16_seg100000_segment2_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr16_seg100000_segment2_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37720 , cntTot: 100000 ; average prediction acc : 0.377200\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.377200\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr16_seg100000_segment3_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr16_seg100000_segment3_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38117 , cntTot: 100000 ; average prediction acc : 0.381170\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.381170\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr16_seg100000_segment4_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr16_seg100000_segment4_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37994 , cntTot: 100000 ; average prediction acc : 0.379940\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.379940\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr16_seg100000_segment5_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr16_seg100000_segment5_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37816 , cntTot: 100000 ; average prediction acc : 0.378160\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.378160\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr16_seg100000_segment6_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr16_seg100000_segment6_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38080 , cntTot: 100000 ; average prediction acc : 0.380800\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.380800\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr16_seg100000_segment7_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr16_seg100000_segment7_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 37524 , cntTot: 100000 ; average prediction acc : 0.375240\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.375240\n","No repeat sections were recorded in the genome data.\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_labelArray_R64_chr16_seg100000_segment8_avgRevCompl0\n","/content/drive/MyDrive/DNA_proj/results/predictions/R64_chr16/modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr19_predArray_R64_chr16_seg100000_segment8_avgRevCompl0\n","lSamples  100000 , nrSteps  100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:5081: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  plt.figure()\n"]},{"output_type":"stream","name":"stdout","text":["cntCorr: 38126 , cntTot: 100000 ; average prediction acc : 0.381260\n","corrDisq: 0 , cntDisqTot: 0 ; average prediction acc no disq: 0.381260\n","No repeat sections were recorded in the genome data.\n"]}]},{"cell_type":"markdown","source":["Then run the Fourier transforms again on new settings (this is even faster than above):"],"metadata":{"id":"KRKPCDO_3F2Z"}},{"cell_type":"code","source":["#GC bias\n","#To get the qual arrays from the model-pred run we need:\n","rootOutput = r\"/content/drive/MyDrive/DNA_proj/results/predictions/\"\n","lastRepeatNr = nrOfRepeats -1 \n","modelFileNameNN = \"modelLSTM_1LayerConv2LayerLstm1LayerDense50_flanks50_win4_filters256_stride1_overlap0_dropout00_bigLoopIter0_repeatNr\" + str(lastRepeatNr)\n","#To get the predReturs fo the bias:\n","modelFileName_forATorGCbias = 'GCbias'\n","rootOutput_forATorGCbias = r\"/content/drive/MyDrive/DNA_proj/results/GCbias/\"  \n","rootInput_forATorGCbias =rootOutput_forATorGCbias\n","forATorGCbias_b = 1 #!\n","\n"],"metadata":{"id":"oBo2RWSV3FKI","executionInfo":{"status":"ok","timestamp":1650143452346,"user_tz":-120,"elapsed":260,"user":{"displayName":"Christian Grønbæk","userId":"14909394381660051993"}}},"execution_count":29,"outputs":[]},{"cell_type":"markdown","source":["These settings are unchanged, except inputArrayType which must be set to 0!:"],"metadata":{"id":"4PXMNg3vJ1l4"}},{"cell_type":"code","source":["\n","#We reuse some of the settings from above\n","modelFileNameNN = modelFileNameNN\n","chromosomeOrderList = chromosomeOrderList\n","chromosomeDict = chromosomeDict \n","\n","#General settings:\n","segmentLength = 100000\n","\n","augmentWithRevComplementary_b = 0  #!\n","\n","#window lgth and stepsize used in generating the avg prediction\n","windowLength = 1\n","stepSize = 1\n","                      \n","#Param's for Fourier plots:\n","shuffle_b = 0 #!\n","randomizeDisqualified_b = 0 #!\n","randomizingByShuffle_b = 0 #!\n","fullGC_b = 0 #!\n","\n","#Which input to use:\n","inputArrayType = 0 # 1: ref base prob's; 0: pred returns (GCbias)\n","plotOnlyNorm_b = 1 #default:1 \n","\n","#Here we only plot the 20-2000 range; extend the list with more starts/stops to cover several ranges\n","fourierStep = 10\n","fourierRawPlotFrq = 5\n","fourierStartList = [20]\n","fourierStopList = [2000]\n","fourierWindowLengthList = [100]\n","\n","ratioQcutoff = 0.9 #0.7 w hg18\n","dumpFourier_b = 0\n","dumpFileNamePrefix = ''"],"metadata":{"id":"h0Xch6eG8dF5","executionInfo":{"status":"ok","timestamp":1650143456615,"user_tz":-120,"elapsed":245,"user":{"displayName":"Christian Grønbæk","userId":"14909394381660051993"}}},"execution_count":30,"outputs":[]},{"cell_type":"code","source":["#Then call Fourier-wrapper:\n","for i in range(len(fourierStartList)):\n","    fourierStart = fourierStartList[i]\n","    fourierStop = fourierStopList[i]\n","    fourierWindowLength = fourierWindowLengthList[i]\n","    stats.computeFourierChromosomes(chromosomeOrderList = chromosomeOrderList,\n","                                    rootOutput = rootOutput,\n","                                    modelFileName = modelFileNameNN,  \n","                                    segmentLength = segmentLength,\n","                                    inputArrayType = inputArrayType,\n","                                    averageRevComplementary_b = augmentWithRevComplementary_b,\n","                                    ratioQcutoff = ratioQcutoff,\n","                                    windowLength = windowLength,\n","                                    stepSize = stepSize,\n","                                    plotOnlyNorm_b = plotOnlyNorm_b,\n","                                    fourierWindowLength = fourierWindowLength,\n","                                    fourierStart = fourierStart,\n","                                    fourierStop = fourierStop,\n","                                    fourierStep = fourierStep, \n","                                    fourierRawPlotFrq = fourierRawPlotFrq,\n","                                    shuffle_b = shuffle_b,\n","                                    randomizeDisqualified_b =randomizeDisqualified_b,\n","                                    randomizingByShuffle_b = randomizingByShuffle_b,\n","                                    forATorGCbias_b = forATorGCbias_b, \n","                                    rootOutput_forATorGCbias= rootOutput_forATorGCbias,\n","                                    rootInput_forATorGCbias = rootInput_forATorGCbias,\n","                                    fullGC_b = fullGC_b,\n","                                    dumpFourier_b = dumpFourier_b,\n","                                    dumpFileNamePrefix = dumpFileNamePrefix, \n","                                    modelFileName_forATorGCbias = modelFileName_forATorGCbias)\n"],"metadata":{"id":"8sb2z_r38s_t","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1650143496441,"user_tz":-120,"elapsed":32989,"user":{"displayName":"Christian Grønbæk","userId":"14909394381660051993"}},"outputId":"8759860e-4c1b-428a-f04a-7fd74e35d272"},"execution_count":31,"outputs":[{"output_type":"stream","name":"stdout","text":["Directory /content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr1/FourierOnPredReturn/20to2000/ created. Output will be placed there.\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr1/GCbias_predReturn_R64_chr1_seg100000_segment0_avgRevCompl0_win1_step1\n","0 (99950,)\n","99950\n","avgPredArray size:  99950\n","ratioQ:  1.0\n","Fourier transf lenght:  99950\n","frqs  [ 0.00000000e+00  1.00050025e-05  2.00100050e-05 ... -3.00150075e-05\n"," -2.00100050e-05 -1.00050025e-05]\n","First 10 (positive) frqs:  [0.00000000e+00 1.00050025e-05 2.00100050e-05 3.00150075e-05\n"," 4.00200100e-05 5.00250125e-05 6.00300150e-05 7.00350175e-05\n"," 8.00400200e-05 9.00450225e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.00050025e-05 -2.00100050e-05 -3.00150075e-05 -4.00200100e-05\n"," -5.00250125e-05 -6.00300150e-05 -7.00350175e-05 -8.00400200e-05\n"," -9.00450225e-05 -1.00050025e-04]\n","Up to stop (2000) with frq 0.020010, the max (min) modulus of Fourier coeff is: 39135.000000 (3.855938)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 950.104187 (0.896689)\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6875: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6876: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6877: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1, ax1 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6878: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6879: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2, ax2 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6880: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr1/GCbias_predReturn_R64_chr1_seg100000_segment1_avgRevCompl0_win1_step1\n","1 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 39455.000000 (3.488686)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 812.552551 (0.740524)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr1/GCbias_predReturn_R64_chr1_seg100000_segment2_avgRevCompl0_win1_step1\n","Directory /content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr2/FourierOnPredReturn/20to2000/ created. Output will be placed there.\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr2/GCbias_predReturn_R64_chr2_seg100000_segment0_avgRevCompl0_win1_step1\n","0 (99950,)\n","99950\n","avgPredArray size:  99950\n","ratioQ:  1.0\n","Fourier transf lenght:  99950\n","frqs  [ 0.00000000e+00  1.00050025e-05  2.00100050e-05 ... -3.00150075e-05\n"," -2.00100050e-05 -1.00050025e-05]\n","First 10 (positive) frqs:  [0.00000000e+00 1.00050025e-05 2.00100050e-05 3.00150075e-05\n"," 4.00200100e-05 5.00250125e-05 6.00300150e-05 7.00350175e-05\n"," 8.00400200e-05 9.00450225e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.00050025e-05 -2.00100050e-05 -3.00150075e-05 -4.00200100e-05\n"," -5.00250125e-05 -6.00300150e-05 -7.00350175e-05 -8.00400200e-05\n"," -9.00450225e-05 -1.00050025e-04]\n","Up to stop (2000) with frq 0.020010, the max (min) modulus of Fourier coeff is: 38526.000000 (5.399194)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 991.838562 (0.498854)\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6875: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6876: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6877: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1, ax1 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6878: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6879: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2, ax2 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6880: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr2/GCbias_predReturn_R64_chr2_seg100000_segment1_avgRevCompl0_win1_step1\n","1 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38442.000000 (1.694917)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 899.119202 (0.852205)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr2/GCbias_predReturn_R64_chr2_seg100000_segment2_avgRevCompl0_win1_step1\n","2 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38137.000000 (6.597044)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 1018.014709 (0.678377)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr2/GCbias_predReturn_R64_chr2_seg100000_segment3_avgRevCompl0_win1_step1\n","3 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37965.000000 (2.864600)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 708.887146 (0.735072)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr2/GCbias_predReturn_R64_chr2_seg100000_segment4_avgRevCompl0_win1_step1\n","4 (100000,)\n","100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7004: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7005: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7006: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7007: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38670.000000 (5.712512)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 947.602356 (0.590547)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr2/GCbias_predReturn_R64_chr2_seg100000_segment5_avgRevCompl0_win1_step1\n","5 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37394.000000 (3.138423)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 804.118591 (0.482632)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr2/GCbias_predReturn_R64_chr2_seg100000_segment6_avgRevCompl0_win1_step1\n","6 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 39083.000000 (3.007717)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 918.461243 (0.892669)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr2/GCbias_predReturn_R64_chr2_seg100000_segment7_avgRevCompl0_win1_step1\n","7 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38455.000000 (6.049281)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 886.269226 (0.578526)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr2/GCbias_predReturn_R64_chr2_seg100000_segment8_avgRevCompl0_win1_step1\n","Directory /content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr3/FourierOnPredReturn/20to2000/ created. Output will be placed there.\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr3/GCbias_predReturn_R64_chr3_seg100000_segment0_avgRevCompl0_win1_step1\n","0 (99950,)\n","99950\n","avgPredArray size:  99950\n","ratioQ:  1.0\n","Fourier transf lenght:  99950\n","frqs  [ 0.00000000e+00  1.00050025e-05  2.00100050e-05 ... -3.00150075e-05\n"," -2.00100050e-05 -1.00050025e-05]\n","First 10 (positive) frqs:  [0.00000000e+00 1.00050025e-05 2.00100050e-05 3.00150075e-05\n"," 4.00200100e-05 5.00250125e-05 6.00300150e-05 7.00350175e-05\n"," 8.00400200e-05 9.00450225e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.00050025e-05 -2.00100050e-05 -3.00150075e-05 -4.00200100e-05\n"," -5.00250125e-05 -6.00300150e-05 -7.00350175e-05 -8.00400200e-05\n"," -9.00450225e-05 -1.00050025e-04]\n","Up to stop (2000) with frq 0.020010, the max (min) modulus of Fourier coeff is: 39102.000000 (2.061519)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 740.832458 (0.441258)\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6875: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6876: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6877: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1, ax1 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6878: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6879: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2, ax2 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6880: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr3/GCbias_predReturn_R64_chr3_seg100000_segment1_avgRevCompl0_win1_step1\n","1 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37197.000000 (3.212271)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 598.906738 (0.452759)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr3/GCbias_predReturn_R64_chr3_seg100000_segment2_avgRevCompl0_win1_step1\n","2 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 39289.000000 (1.296322)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 1063.220215 (0.936816)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr3/GCbias_predReturn_R64_chr3_seg100000_segment3_avgRevCompl0_win1_step1\n","Directory /content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr4/FourierOnPredReturn/20to2000/ created. Output will be placed there.\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr4/GCbias_predReturn_R64_chr4_seg100000_segment0_avgRevCompl0_win1_step1\n","0 (99950,)\n","99950\n","avgPredArray size:  99950\n","ratioQ:  1.0\n","Fourier transf lenght:  99950\n","frqs  [ 0.00000000e+00  1.00050025e-05  2.00100050e-05 ... -3.00150075e-05\n"," -2.00100050e-05 -1.00050025e-05]\n","First 10 (positive) frqs:  [0.00000000e+00 1.00050025e-05 2.00100050e-05 3.00150075e-05\n"," 4.00200100e-05 5.00250125e-05 6.00300150e-05 7.00350175e-05\n"," 8.00400200e-05 9.00450225e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.00050025e-05 -2.00100050e-05 -3.00150075e-05 -4.00200100e-05\n"," -5.00250125e-05 -6.00300150e-05 -7.00350175e-05 -8.00400200e-05\n"," -9.00450225e-05 -1.00050025e-04]\n","Up to stop (2000) with frq 0.020010, the max (min) modulus of Fourier coeff is: 38375.000000 (1.257565)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 866.031555 (0.367266)\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6875: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6876: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6877: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1, ax1 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6878: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6879: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2, ax2 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6880: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr4/GCbias_predReturn_R64_chr4_seg100000_segment1_avgRevCompl0_win1_step1\n","1 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38615.000000 (1.860952)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 854.235840 (0.686210)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr4/GCbias_predReturn_R64_chr4_seg100000_segment2_avgRevCompl0_win1_step1\n","2 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38235.000000 (0.962727)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 948.962097 (0.620680)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr4/GCbias_predReturn_R64_chr4_seg100000_segment3_avgRevCompl0_win1_step1\n","3 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38186.000000 (4.063469)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 1014.057739 (0.523170)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr4/GCbias_predReturn_R64_chr4_seg100000_segment4_avgRevCompl0_win1_step1\n","4 (100000,)\n","100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7004: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7005: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7006: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7007: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37698.000000 (4.765543)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 706.232666 (0.813331)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr4/GCbias_predReturn_R64_chr4_seg100000_segment5_avgRevCompl0_win1_step1\n","5 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37472.000000 (2.361251)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 542.990051 (0.082026)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr4/GCbias_predReturn_R64_chr4_seg100000_segment6_avgRevCompl0_win1_step1\n","6 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37508.000000 (4.288541)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 904.683228 (0.726819)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr4/GCbias_predReturn_R64_chr4_seg100000_segment7_avgRevCompl0_win1_step1\n","7 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37631.000000 (3.301439)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 1074.510132 (0.768834)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr4/GCbias_predReturn_R64_chr4_seg100000_segment8_avgRevCompl0_win1_step1\n","8 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37892.000000 (0.864101)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 862.849915 (0.196460)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr4/GCbias_predReturn_R64_chr4_seg100000_segment9_avgRevCompl0_win1_step1\n","9 (100000,)\n","100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7004: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7005: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7006: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7007: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37629.000000 (1.740185)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 888.740234 (0.529046)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr4/GCbias_predReturn_R64_chr4_seg100000_segment10_avgRevCompl0_win1_step1\n","10 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37375.000000 (4.920184)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 720.126709 (0.741546)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr4/GCbias_predReturn_R64_chr4_seg100000_segment11_avgRevCompl0_win1_step1\n","11 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37339.000000 (3.271866)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 1024.765259 (0.290008)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr4/GCbias_predReturn_R64_chr4_seg100000_segment12_avgRevCompl0_win1_step1\n","12 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38006.000000 (5.178987)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 670.478638 (0.220259)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr4/GCbias_predReturn_R64_chr4_seg100000_segment13_avgRevCompl0_win1_step1\n","13 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38110.000000 (3.837377)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 982.680603 (0.325212)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr4/GCbias_predReturn_R64_chr4_seg100000_segment14_avgRevCompl0_win1_step1\n","14 (100000,)\n","100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7004: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7005: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7006: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7007: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38188.000000 (3.895235)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 741.861450 (0.847426)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr4/GCbias_predReturn_R64_chr4_seg100000_segment15_avgRevCompl0_win1_step1\n","Directory /content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr5/FourierOnPredReturn/20to2000/ created. Output will be placed there.\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr5/GCbias_predReturn_R64_chr5_seg100000_segment0_avgRevCompl0_win1_step1\n","0 (99950,)\n","99950\n","avgPredArray size:  99950\n","ratioQ:  1.0\n","Fourier transf lenght:  99950\n","frqs  [ 0.00000000e+00  1.00050025e-05  2.00100050e-05 ... -3.00150075e-05\n"," -2.00100050e-05 -1.00050025e-05]\n","First 10 (positive) frqs:  [0.00000000e+00 1.00050025e-05 2.00100050e-05 3.00150075e-05\n"," 4.00200100e-05 5.00250125e-05 6.00300150e-05 7.00350175e-05\n"," 8.00400200e-05 9.00450225e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.00050025e-05 -2.00100050e-05 -3.00150075e-05 -4.00200100e-05\n"," -5.00250125e-05 -6.00300150e-05 -7.00350175e-05 -8.00400200e-05\n"," -9.00450225e-05 -1.00050025e-04]\n","Up to stop (2000) with frq 0.020010, the max (min) modulus of Fourier coeff is: 39171.003906 (0.558704)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 724.364502 (0.134348)\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6875: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6876: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6877: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1, ax1 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6878: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6879: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2, ax2 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6880: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr5/GCbias_predReturn_R64_chr5_seg100000_segment1_avgRevCompl0_win1_step1\n","1 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37870.000000 (1.001016)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 697.037964 (1.302505)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr5/GCbias_predReturn_R64_chr5_seg100000_segment2_avgRevCompl0_win1_step1\n","2 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38470.000000 (5.784386)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 726.994995 (0.508579)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr5/GCbias_predReturn_R64_chr5_seg100000_segment3_avgRevCompl0_win1_step1\n","3 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38571.000000 (7.755730)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 737.434021 (0.447095)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr5/GCbias_predReturn_R64_chr5_seg100000_segment4_avgRevCompl0_win1_step1\n","4 (100000,)\n","100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7004: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7005: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7006: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7007: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38311.000000 (3.292802)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 778.410522 (0.149960)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr5/GCbias_predReturn_R64_chr5_seg100000_segment5_avgRevCompl0_win1_step1\n","Directory /content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr6/FourierOnPredReturn/20to2000/ created. Output will be placed there.\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr6/GCbias_predReturn_R64_chr6_seg100000_segment0_avgRevCompl0_win1_step1\n","0 (99950,)\n","99950\n","avgPredArray size:  99950\n","ratioQ:  1.0\n","Fourier transf lenght:  99950\n","frqs  [ 0.00000000e+00  1.00050025e-05  2.00100050e-05 ... -3.00150075e-05\n"," -2.00100050e-05 -1.00050025e-05]\n","First 10 (positive) frqs:  [0.00000000e+00 1.00050025e-05 2.00100050e-05 3.00150075e-05\n"," 4.00200100e-05 5.00250125e-05 6.00300150e-05 7.00350175e-05\n"," 8.00400200e-05 9.00450225e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.00050025e-05 -2.00100050e-05 -3.00150075e-05 -4.00200100e-05\n"," -5.00250125e-05 -6.00300150e-05 -7.00350175e-05 -8.00400200e-05\n"," -9.00450225e-05 -1.00050025e-04]\n","Up to stop (2000) with frq 0.020010, the max (min) modulus of Fourier coeff is: 39447.000000 (3.155544)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 748.699524 (0.390636)\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6875: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6876: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6877: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1, ax1 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6878: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6879: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2, ax2 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6880: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr6/GCbias_predReturn_R64_chr6_seg100000_segment1_avgRevCompl0_win1_step1\n","1 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37861.000000 (4.244049)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 803.977234 (0.678442)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr6/GCbias_predReturn_R64_chr6_seg100000_segment2_avgRevCompl0_win1_step1\n","Directory /content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr7/FourierOnPredReturn/20to2000/ created. Output will be placed there.\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr7/GCbias_predReturn_R64_chr7_seg100000_segment0_avgRevCompl0_win1_step1\n","0 (99950,)\n","99950\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6875: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6876: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6877: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1, ax1 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6878: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6879: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2, ax2 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6880: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["avgPredArray size:  99950\n","ratioQ:  1.0\n","Fourier transf lenght:  99950\n","frqs  [ 0.00000000e+00  1.00050025e-05  2.00100050e-05 ... -3.00150075e-05\n"," -2.00100050e-05 -1.00050025e-05]\n","First 10 (positive) frqs:  [0.00000000e+00 1.00050025e-05 2.00100050e-05 3.00150075e-05\n"," 4.00200100e-05 5.00250125e-05 6.00300150e-05 7.00350175e-05\n"," 8.00400200e-05 9.00450225e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.00050025e-05 -2.00100050e-05 -3.00150075e-05 -4.00200100e-05\n"," -5.00250125e-05 -6.00300150e-05 -7.00350175e-05 -8.00400200e-05\n"," -9.00450225e-05 -1.00050025e-04]\n","Up to stop (2000) with frq 0.020010, the max (min) modulus of Fourier coeff is: 37696.000000 (2.964901)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 887.371582 (0.774630)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr7/GCbias_predReturn_R64_chr7_seg100000_segment1_avgRevCompl0_win1_step1\n","1 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38097.000000 (3.365549)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 1051.535034 (0.689957)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr7/GCbias_predReturn_R64_chr7_seg100000_segment2_avgRevCompl0_win1_step1\n","2 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37976.000000 (4.304747)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 825.132019 (0.181377)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr7/GCbias_predReturn_R64_chr7_seg100000_segment3_avgRevCompl0_win1_step1\n","3 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37900.000000 (1.026866)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 677.136658 (1.103743)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr7/GCbias_predReturn_R64_chr7_seg100000_segment4_avgRevCompl0_win1_step1\n","4 (100000,)\n","100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7004: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7005: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7006: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7007: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38389.000000 (1.924333)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 728.271545 (0.542412)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr7/GCbias_predReturn_R64_chr7_seg100000_segment5_avgRevCompl0_win1_step1\n","5 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38153.000000 (1.699000)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 963.130493 (0.734438)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr7/GCbias_predReturn_R64_chr7_seg100000_segment6_avgRevCompl0_win1_step1\n","6 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37736.000000 (3.407702)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 662.903931 (0.085869)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr7/GCbias_predReturn_R64_chr7_seg100000_segment7_avgRevCompl0_win1_step1\n","7 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37740.000000 (4.441279)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 782.580322 (0.734931)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr7/GCbias_predReturn_R64_chr7_seg100000_segment8_avgRevCompl0_win1_step1\n","8 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37764.000000 (3.438229)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 738.992493 (0.636311)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr7/GCbias_predReturn_R64_chr7_seg100000_segment9_avgRevCompl0_win1_step1\n","9 (100000,)\n","100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7004: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7005: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7006: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7007: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38346.000000 (2.088079)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 854.756409 (0.676943)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr7/GCbias_predReturn_R64_chr7_seg100000_segment10_avgRevCompl0_win1_step1\n","Directory /content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr8/FourierOnPredReturn/20to2000/ created. Output will be placed there.\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr8/GCbias_predReturn_R64_chr8_seg100000_segment0_avgRevCompl0_win1_step1\n","0 (99950,)\n","99950\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6875: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6876: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6877: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1, ax1 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6878: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6879: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2, ax2 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6880: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["avgPredArray size:  99950\n","ratioQ:  1.0\n","Fourier transf lenght:  99950\n","frqs  [ 0.00000000e+00  1.00050025e-05  2.00100050e-05 ... -3.00150075e-05\n"," -2.00100050e-05 -1.00050025e-05]\n","First 10 (positive) frqs:  [0.00000000e+00 1.00050025e-05 2.00100050e-05 3.00150075e-05\n"," 4.00200100e-05 5.00250125e-05 6.00300150e-05 7.00350175e-05\n"," 8.00400200e-05 9.00450225e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.00050025e-05 -2.00100050e-05 -3.00150075e-05 -4.00200100e-05\n"," -5.00250125e-05 -6.00300150e-05 -7.00350175e-05 -8.00400200e-05\n"," -9.00450225e-05 -1.00050025e-04]\n","Up to stop (2000) with frq 0.020010, the max (min) modulus of Fourier coeff is: 38692.000000 (2.146831)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 826.818970 (0.454288)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr8/GCbias_predReturn_R64_chr8_seg100000_segment1_avgRevCompl0_win1_step1\n","1 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38678.000000 (3.073677)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 997.092834 (0.983115)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr8/GCbias_predReturn_R64_chr8_seg100000_segment2_avgRevCompl0_win1_step1\n","2 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37969.000000 (1.709520)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 651.773193 (0.043760)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr8/GCbias_predReturn_R64_chr8_seg100000_segment3_avgRevCompl0_win1_step1\n","3 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38309.000000 (2.118328)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 625.954163 (1.041478)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr8/GCbias_predReturn_R64_chr8_seg100000_segment4_avgRevCompl0_win1_step1\n","4 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38438.000000 (7.333632)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 900.470215 (0.943194)\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7004: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7005: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7006: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7007: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr8/GCbias_predReturn_R64_chr8_seg100000_segment5_avgRevCompl0_win1_step1\n","Directory /content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr9/FourierOnPredReturn/20to2000/ created. Output will be placed there.\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr9/GCbias_predReturn_R64_chr9_seg100000_segment0_avgRevCompl0_win1_step1\n","0 (99950,)\n","99950\n","avgPredArray size:  99950\n","ratioQ:  1.0\n","Fourier transf lenght:  99950\n","frqs  [ 0.00000000e+00  1.00050025e-05  2.00100050e-05 ... -3.00150075e-05\n"," -2.00100050e-05 -1.00050025e-05]\n","First 10 (positive) frqs:  [0.00000000e+00 1.00050025e-05 2.00100050e-05 3.00150075e-05\n"," 4.00200100e-05 5.00250125e-05 6.00300150e-05 7.00350175e-05\n"," 8.00400200e-05 9.00450225e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.00050025e-05 -2.00100050e-05 -3.00150075e-05 -4.00200100e-05\n"," -5.00250125e-05 -6.00300150e-05 -7.00350175e-05 -8.00400200e-05\n"," -9.00450225e-05 -1.00050025e-04]\n","Up to stop (2000) with frq 0.020010, the max (min) modulus of Fourier coeff is: 39401.003906 (5.490741)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 912.172363 (0.778119)\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6875: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6876: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6877: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1, ax1 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6878: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6879: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2, ax2 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6880: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr9/GCbias_predReturn_R64_chr9_seg100000_segment1_avgRevCompl0_win1_step1\n","1 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38234.000000 (3.469328)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 755.936829 (0.648541)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr9/GCbias_predReturn_R64_chr9_seg100000_segment2_avgRevCompl0_win1_step1\n","2 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38701.000000 (3.187570)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 838.946228 (0.528692)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr9/GCbias_predReturn_R64_chr9_seg100000_segment3_avgRevCompl0_win1_step1\n","3 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 39173.000000 (2.738947)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 1108.503418 (0.399133)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr9/GCbias_predReturn_R64_chr9_seg100000_segment4_avgRevCompl0_win1_step1\n","Directory /content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr10/FourierOnPredReturn/20to2000/ created. Output will be placed there.\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr10/GCbias_predReturn_R64_chr10_seg100000_segment0_avgRevCompl0_win1_step1\n","0 (99950,)\n","99950\n","avgPredArray size:  99950\n","ratioQ:  1.0\n","Fourier transf lenght:  99950\n","frqs  [ 0.00000000e+00  1.00050025e-05  2.00100050e-05 ... -3.00150075e-05\n"," -2.00100050e-05 -1.00050025e-05]\n","First 10 (positive) frqs:  [0.00000000e+00 1.00050025e-05 2.00100050e-05 3.00150075e-05\n"," 4.00200100e-05 5.00250125e-05 6.00300150e-05 7.00350175e-05\n"," 8.00400200e-05 9.00450225e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.00050025e-05 -2.00100050e-05 -3.00150075e-05 -4.00200100e-05\n"," -5.00250125e-05 -6.00300150e-05 -7.00350175e-05 -8.00400200e-05\n"," -9.00450225e-05 -1.00050025e-04]\n","Up to stop (2000) with frq 0.020010, the max (min) modulus of Fourier coeff is: 38460.000000 (3.450973)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 792.104858 (0.518334)\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6875: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6876: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6877: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1, ax1 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6878: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6879: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2, ax2 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6880: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr10/GCbias_predReturn_R64_chr10_seg100000_segment1_avgRevCompl0_win1_step1\n","1 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38616.000000 (1.772816)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 725.846252 (0.931791)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr10/GCbias_predReturn_R64_chr10_seg100000_segment2_avgRevCompl0_win1_step1\n","2 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37912.000000 (2.891898)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 809.183350 (0.962082)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr10/GCbias_predReturn_R64_chr10_seg100000_segment3_avgRevCompl0_win1_step1\n","3 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38415.000000 (1.440082)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 760.127625 (0.374168)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr10/GCbias_predReturn_R64_chr10_seg100000_segment4_avgRevCompl0_win1_step1\n","4 (100000,)\n","100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7004: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7005: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7006: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7007: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38054.000000 (3.885078)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 910.987732 (0.826151)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr10/GCbias_predReturn_R64_chr10_seg100000_segment5_avgRevCompl0_win1_step1\n","5 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38367.000000 (4.707547)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 903.843689 (0.314351)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr10/GCbias_predReturn_R64_chr10_seg100000_segment6_avgRevCompl0_win1_step1\n","6 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38429.000000 (2.926780)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 952.712341 (0.470634)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr10/GCbias_predReturn_R64_chr10_seg100000_segment7_avgRevCompl0_win1_step1\n","Directory /content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr11/FourierOnPredReturn/20to2000/ created. Output will be placed there.\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr11/GCbias_predReturn_R64_chr11_seg100000_segment0_avgRevCompl0_win1_step1\n","0 (99950,)\n","99950\n","avgPredArray size:  99950\n","ratioQ:  1.0\n","Fourier transf lenght:  99950\n","frqs  [ 0.00000000e+00  1.00050025e-05  2.00100050e-05 ... -3.00150075e-05\n"," -2.00100050e-05 -1.00050025e-05]\n","First 10 (positive) frqs:  [0.00000000e+00 1.00050025e-05 2.00100050e-05 3.00150075e-05\n"," 4.00200100e-05 5.00250125e-05 6.00300150e-05 7.00350175e-05\n"," 8.00400200e-05 9.00450225e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.00050025e-05 -2.00100050e-05 -3.00150075e-05 -4.00200100e-05\n"," -5.00250125e-05 -6.00300150e-05 -7.00350175e-05 -8.00400200e-05\n"," -9.00450225e-05 -1.00050025e-04]\n","Up to stop (2000) with frq 0.020010, the max (min) modulus of Fourier coeff is: 38333.000000 (3.280200)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 895.847778 (1.045458)\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6875: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6876: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6877: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1, ax1 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6878: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6879: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2, ax2 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6880: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr11/GCbias_predReturn_R64_chr11_seg100000_segment1_avgRevCompl0_win1_step1\n","1 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38018.000000 (2.839303)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 795.839050 (1.062823)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr11/GCbias_predReturn_R64_chr11_seg100000_segment2_avgRevCompl0_win1_step1\n","2 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38136.000000 (5.101604)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 588.971313 (0.338734)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr11/GCbias_predReturn_R64_chr11_seg100000_segment3_avgRevCompl0_win1_step1\n","3 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38240.000000 (2.809426)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 864.116699 (0.490214)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr11/GCbias_predReturn_R64_chr11_seg100000_segment4_avgRevCompl0_win1_step1\n","4 (100000,)\n","100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7004: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7005: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7006: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7007: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37276.000000 (7.150682)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 774.341125 (0.835015)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr11/GCbias_predReturn_R64_chr11_seg100000_segment5_avgRevCompl0_win1_step1\n","5 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38673.000000 (4.930295)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 1046.996826 (0.227696)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr11/GCbias_predReturn_R64_chr11_seg100000_segment6_avgRevCompl0_win1_step1\n","Directory /content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr12/FourierOnPredReturn/20to2000/ created. Output will be placed there.\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr12/GCbias_predReturn_R64_chr12_seg100000_segment0_avgRevCompl0_win1_step1\n","0 (99950,)\n","99950\n","avgPredArray size:  99950\n","ratioQ:  1.0\n","Fourier transf lenght:  99950\n","frqs  [ 0.00000000e+00  1.00050025e-05  2.00100050e-05 ... -3.00150075e-05\n"," -2.00100050e-05 -1.00050025e-05]\n","First 10 (positive) frqs:  [0.00000000e+00 1.00050025e-05 2.00100050e-05 3.00150075e-05\n"," 4.00200100e-05 5.00250125e-05 6.00300150e-05 7.00350175e-05\n"," 8.00400200e-05 9.00450225e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.00050025e-05 -2.00100050e-05 -3.00150075e-05 -4.00200100e-05\n"," -5.00250125e-05 -6.00300150e-05 -7.00350175e-05 -8.00400200e-05\n"," -9.00450225e-05 -1.00050025e-04]\n","Up to stop (2000) with frq 0.020010, the max (min) modulus of Fourier coeff is: 39232.000000 (3.170581)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 712.352051 (0.105585)\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6875: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6876: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6877: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1, ax1 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6878: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6879: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2, ax2 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6880: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr12/GCbias_predReturn_R64_chr12_seg100000_segment1_avgRevCompl0_win1_step1\n","1 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37789.000000 (2.669303)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 835.425293 (0.970405)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr12/GCbias_predReturn_R64_chr12_seg100000_segment2_avgRevCompl0_win1_step1\n","2 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38104.000000 (4.943389)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 933.570374 (0.509410)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr12/GCbias_predReturn_R64_chr12_seg100000_segment3_avgRevCompl0_win1_step1\n","3 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38704.000000 (2.942016)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 1292.360352 (1.168372)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr12/GCbias_predReturn_R64_chr12_seg100000_segment4_avgRevCompl0_win1_step1\n","4 (100000,)\n","100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7004: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7005: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7006: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7007: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 39938.000000 (6.128075)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 594.277039 (0.465773)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr12/GCbias_predReturn_R64_chr12_seg100000_segment5_avgRevCompl0_win1_step1\n","5 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38734.000000 (2.386147)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 922.256470 (0.663458)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr12/GCbias_predReturn_R64_chr12_seg100000_segment6_avgRevCompl0_win1_step1\n","6 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37776.000000 (0.175470)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 721.221252 (0.512641)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr12/GCbias_predReturn_R64_chr12_seg100000_segment7_avgRevCompl0_win1_step1\n","7 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37685.000000 (3.640028)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 728.873901 (0.589057)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr12/GCbias_predReturn_R64_chr12_seg100000_segment8_avgRevCompl0_win1_step1\n","8 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38739.000000 (5.936641)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 903.042114 (0.455218)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr12/GCbias_predReturn_R64_chr12_seg100000_segment9_avgRevCompl0_win1_step1\n","9 (100000,)\n","100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7004: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7005: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7006: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7007: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37978.000000 (3.925025)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 1114.926636 (0.427163)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr12/GCbias_predReturn_R64_chr12_seg100000_segment10_avgRevCompl0_win1_step1\n","Directory /content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr13/FourierOnPredReturn/20to2000/ created. Output will be placed there.\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr13/GCbias_predReturn_R64_chr13_seg100000_segment0_avgRevCompl0_win1_step1\n","0 (99950,)\n","99950\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6875: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6876: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6877: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1, ax1 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6878: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6879: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2, ax2 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6880: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["avgPredArray size:  99950\n","ratioQ:  1.0\n","Fourier transf lenght:  99950\n","frqs  [ 0.00000000e+00  1.00050025e-05  2.00100050e-05 ... -3.00150075e-05\n"," -2.00100050e-05 -1.00050025e-05]\n","First 10 (positive) frqs:  [0.00000000e+00 1.00050025e-05 2.00100050e-05 3.00150075e-05\n"," 4.00200100e-05 5.00250125e-05 6.00300150e-05 7.00350175e-05\n"," 8.00400200e-05 9.00450225e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.00050025e-05 -2.00100050e-05 -3.00150075e-05 -4.00200100e-05\n"," -5.00250125e-05 -6.00300150e-05 -7.00350175e-05 -8.00400200e-05\n"," -9.00450225e-05 -1.00050025e-04]\n","Up to stop (2000) with frq 0.020010, the max (min) modulus of Fourier coeff is: 38751.000000 (2.401865)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 819.407166 (0.658489)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr13/GCbias_predReturn_R64_chr13_seg100000_segment1_avgRevCompl0_win1_step1\n","1 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38215.000000 (1.030676)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 863.363281 (0.842250)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr13/GCbias_predReturn_R64_chr13_seg100000_segment2_avgRevCompl0_win1_step1\n","2 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38026.000000 (0.779880)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 807.149475 (1.352836)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr13/GCbias_predReturn_R64_chr13_seg100000_segment3_avgRevCompl0_win1_step1\n","3 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38352.000000 (3.411785)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 786.222412 (0.746424)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr13/GCbias_predReturn_R64_chr13_seg100000_segment4_avgRevCompl0_win1_step1\n","4 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37992.000000 (2.635643)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 708.540161 (0.237193)\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7004: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7005: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7006: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7007: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr13/GCbias_predReturn_R64_chr13_seg100000_segment5_avgRevCompl0_win1_step1\n","5 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37532.000000 (3.961549)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 669.964661 (0.083615)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr13/GCbias_predReturn_R64_chr13_seg100000_segment6_avgRevCompl0_win1_step1\n","6 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38263.000000 (2.025059)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 974.887817 (0.304351)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr13/GCbias_predReturn_R64_chr13_seg100000_segment7_avgRevCompl0_win1_step1\n","7 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37848.000000 (3.255530)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 1180.512207 (0.222320)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr13/GCbias_predReturn_R64_chr13_seg100000_segment8_avgRevCompl0_win1_step1\n","8 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38462.000000 (6.900657)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 875.096924 (0.755974)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr13/GCbias_predReturn_R64_chr13_seg100000_segment9_avgRevCompl0_win1_step1\n","Directory /content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr14/FourierOnPredReturn/20to2000/ created. Output will be placed there.\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr14/GCbias_predReturn_R64_chr14_seg100000_segment0_avgRevCompl0_win1_step1\n","0 (99950,)\n","99950\n","avgPredArray size:  99950\n","ratioQ:  1.0\n","Fourier transf lenght:  99950\n","frqs  [ 0.00000000e+00  1.00050025e-05  2.00100050e-05 ... -3.00150075e-05\n"," -2.00100050e-05 -1.00050025e-05]\n","First 10 (positive) frqs:  [0.00000000e+00 1.00050025e-05 2.00100050e-05 3.00150075e-05\n"," 4.00200100e-05 5.00250125e-05 6.00300150e-05 7.00350175e-05\n"," 8.00400200e-05 9.00450225e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.00050025e-05 -2.00100050e-05 -3.00150075e-05 -4.00200100e-05\n"," -5.00250125e-05 -6.00300150e-05 -7.00350175e-05 -8.00400200e-05\n"," -9.00450225e-05 -1.00050025e-04]\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6875: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6876: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6877: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1, ax1 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6878: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6879: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2, ax2 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6880: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["Up to stop (2000) with frq 0.020010, the max (min) modulus of Fourier coeff is: 39249.000000 (1.739794)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 963.375427 (0.950406)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr14/GCbias_predReturn_R64_chr14_seg100000_segment1_avgRevCompl0_win1_step1\n","1 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38842.000000 (1.533661)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 1372.126709 (0.280411)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr14/GCbias_predReturn_R64_chr14_seg100000_segment2_avgRevCompl0_win1_step1\n","2 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38162.000000 (1.556250)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 984.528809 (0.234910)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr14/GCbias_predReturn_R64_chr14_seg100000_segment3_avgRevCompl0_win1_step1\n","3 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38393.000000 (0.374439)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 1200.681885 (0.475324)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr14/GCbias_predReturn_R64_chr14_seg100000_segment4_avgRevCompl0_win1_step1\n","4 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38514.000000 (1.273655)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 736.838684 (0.941530)\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7004: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7005: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7006: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7007: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr14/GCbias_predReturn_R64_chr14_seg100000_segment5_avgRevCompl0_win1_step1\n","5 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38288.000000 (2.974340)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 741.596252 (0.499196)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr14/GCbias_predReturn_R64_chr14_seg100000_segment6_avgRevCompl0_win1_step1\n","6 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 39000.000000 (2.300987)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 787.869690 (0.869069)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr14/GCbias_predReturn_R64_chr14_seg100000_segment7_avgRevCompl0_win1_step1\n","Directory /content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr15/FourierOnPredReturn/20to2000/ created. Output will be placed there.\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr15/GCbias_predReturn_R64_chr15_seg100000_segment0_avgRevCompl0_win1_step1\n","0 (99950,)\n","99950\n","avgPredArray size:  99950\n","ratioQ:  1.0\n","Fourier transf lenght:  99950\n","frqs  [ 0.00000000e+00  1.00050025e-05  2.00100050e-05 ... -3.00150075e-05\n"," -2.00100050e-05 -1.00050025e-05]\n","First 10 (positive) frqs:  [0.00000000e+00 1.00050025e-05 2.00100050e-05 3.00150075e-05\n"," 4.00200100e-05 5.00250125e-05 6.00300150e-05 7.00350175e-05\n"," 8.00400200e-05 9.00450225e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.00050025e-05 -2.00100050e-05 -3.00150075e-05 -4.00200100e-05\n"," -5.00250125e-05 -6.00300150e-05 -7.00350175e-05 -8.00400200e-05\n"," -9.00450225e-05 -1.00050025e-04]\n","Up to stop (2000) with frq 0.020010, the max (min) modulus of Fourier coeff is: 38265.000000 (6.673558)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 781.373840 (0.481862)\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6875: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6876: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6877: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1, ax1 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6878: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6879: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2, ax2 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6880: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr15/GCbias_predReturn_R64_chr15_seg100000_segment1_avgRevCompl0_win1_step1\n","1 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37846.000000 (4.855501)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 724.919922 (0.499967)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr15/GCbias_predReturn_R64_chr15_seg100000_segment2_avgRevCompl0_win1_step1\n","2 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38721.000000 (2.347907)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 748.402649 (0.946146)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr15/GCbias_predReturn_R64_chr15_seg100000_segment3_avgRevCompl0_win1_step1\n","3 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37907.000000 (2.838657)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 912.324219 (0.296430)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr15/GCbias_predReturn_R64_chr15_seg100000_segment4_avgRevCompl0_win1_step1\n","4 (100000,)\n","100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7004: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7005: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7006: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7007: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38306.000000 (1.972946)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 752.990601 (0.386541)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr15/GCbias_predReturn_R64_chr15_seg100000_segment5_avgRevCompl0_win1_step1\n","5 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38445.000000 (3.331140)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 760.962830 (0.256749)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr15/GCbias_predReturn_R64_chr15_seg100000_segment6_avgRevCompl0_win1_step1\n","6 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37521.000000 (2.689893)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 711.439270 (1.315522)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr15/GCbias_predReturn_R64_chr15_seg100000_segment7_avgRevCompl0_win1_step1\n","7 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37838.000000 (6.334293)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 802.563293 (0.587268)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr15/GCbias_predReturn_R64_chr15_seg100000_segment8_avgRevCompl0_win1_step1\n","8 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37807.000000 (2.591267)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 871.201538 (1.348981)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr15/GCbias_predReturn_R64_chr15_seg100000_segment9_avgRevCompl0_win1_step1\n","9 (100000,)\n","100000\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7004: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7005: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7006: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7007: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38331.000000 (0.763626)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 1028.242310 (0.629951)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr15/GCbias_predReturn_R64_chr15_seg100000_segment10_avgRevCompl0_win1_step1\n","Directory /content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr16/FourierOnPredReturn/20to2000/ created. Output will be placed there.\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr16/GCbias_predReturn_R64_chr16_seg100000_segment0_avgRevCompl0_win1_step1\n","0 (99950,)\n","99950\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6875: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6876: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6877: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1, ax1 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6878: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6879: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2, ax2 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:6880: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["avgPredArray size:  99950\n","ratioQ:  1.0\n","Fourier transf lenght:  99950\n","frqs  [ 0.00000000e+00  1.00050025e-05  2.00100050e-05 ... -3.00150075e-05\n"," -2.00100050e-05 -1.00050025e-05]\n","First 10 (positive) frqs:  [0.00000000e+00 1.00050025e-05 2.00100050e-05 3.00150075e-05\n"," 4.00200100e-05 5.00250125e-05 6.00300150e-05 7.00350175e-05\n"," 8.00400200e-05 9.00450225e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.00050025e-05 -2.00100050e-05 -3.00150075e-05 -4.00200100e-05\n"," -5.00250125e-05 -6.00300150e-05 -7.00350175e-05 -8.00400200e-05\n"," -9.00450225e-05 -1.00050025e-04]\n","Up to stop (2000) with frq 0.020010, the max (min) modulus of Fourier coeff is: 38689.003906 (1.614296)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 723.897400 (0.540813)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr16/GCbias_predReturn_R64_chr16_seg100000_segment1_avgRevCompl0_win1_step1\n","1 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38260.000000 (2.573526)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 1055.551636 (0.469753)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr16/GCbias_predReturn_R64_chr16_seg100000_segment2_avgRevCompl0_win1_step1\n","2 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37720.000000 (2.430522)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 671.109314 (0.585198)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr16/GCbias_predReturn_R64_chr16_seg100000_segment3_avgRevCompl0_win1_step1\n","3 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38117.000000 (4.540146)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 812.011169 (0.441618)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr16/GCbias_predReturn_R64_chr16_seg100000_segment4_avgRevCompl0_win1_step1\n","4 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37994.000000 (1.448889)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 706.109985 (0.292580)\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7004: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig0, ax0 = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7005: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig1_loc, ax1_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7006: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig2_loc, ax2_loc = plt.subplots()\n","/content/drive/MyDrive/DNA_proj/development/dnaNet_stats_py3.py:7007: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n","  fig3, ax3 = plt.subplots()\n"]},{"output_type":"stream","name":"stdout","text":["/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr16/GCbias_predReturn_R64_chr16_seg100000_segment5_avgRevCompl0_win1_step1\n","5 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37816.000000 (2.446056)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 1092.704956 (0.352133)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr16/GCbias_predReturn_R64_chr16_seg100000_segment6_avgRevCompl0_win1_step1\n","6 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38080.000000 (1.625258)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 764.603149 (0.274677)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr16/GCbias_predReturn_R64_chr16_seg100000_segment7_avgRevCompl0_win1_step1\n","7 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 37524.000000 (1.535277)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 695.751648 (1.093943)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr16/GCbias_predReturn_R64_chr16_seg100000_segment8_avgRevCompl0_win1_step1\n","8 (100000,)\n","100000\n","ratioQ:  1.0\n","Fourier transf lenght:  100000\n","frqs  [ 0.e+00  1.e-05  2.e-05 ... -3.e-05 -2.e-05 -1.e-05]\n","First 10 (positive) frqs:  [0.e+00 1.e-05 2.e-05 3.e-05 4.e-05 5.e-05 6.e-05 7.e-05 8.e-05 9.e-05]\n","Largest (positive) frq: -0.500000\n","First 10 (negative) frqs:  [-1.e-05 -2.e-05 -3.e-05 -4.e-05 -5.e-05 -6.e-05 -7.e-05 -8.e-05 -9.e-05\n"," -1.e-04]\n","Up to stop (2000) with frq 0.020000, the max (min) modulus of Fourier coeff is: 38126.000000 (3.540286)\n","After stop (2000) the max (min) modulus of Fourier coeff is: 699.919067 (0.580020)\n","/content/drive/MyDrive/DNA_proj/results/GCbias/R64_chr16/GCbias_predReturn_R64_chr16_seg100000_segment9_avgRevCompl0_win1_step1\n"]}]},{"cell_type":"markdown","source":["The output plots should now sit in FourierOnPredReturn/20To2000 subfolders of the GCbias/chrNN folders."],"metadata":{"id":"Rs8CJmOKKnPg"}}],"metadata":{"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"name":"python"},"colab":{"name":"LSTM50onYeast.ipynb","provenance":[],"collapsed_sections":[]},"accelerator":"GPU"},"nbformat":4,"nbformat_minor":0}